// @aiot-toolkit/aiotpack/lib/utils/ux/android/AndroidUx.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parser = require("@aiot-toolkit/parser");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _UxToTypescript = _interopRequireDefault(require("@aiot-toolkit/parser/lib/ux/translate/android/UxToTypescript"));
var _generator = require("@aiot-toolkit/generator");
var _path = _interopRequireDefault(require("path"));
var _QuickAppDocHelp = require("@aiot-toolkit/shared-utils/lib/utils/QuickAppDocHelp");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * AndroidUx
 */
class AndroidUx {
  /**
   *
   * 编译 ux 文件
   * 1. 获取文件类型：app  page  普通ux
   * 2. 使用`UxParse`解析
   * 3. 使用`UxToTypeScript`转换
   * 4. 使用`TypeScriptGenerator`生成
   * @param params
   */
  async compileUx(params) {
    const {
      context,
      file,
      compilerOption
    } = params;
    const {
      path: filePath,
      content
    } = file;
    const {
      projectPath
    } = context;
    const fileType = this.getFileType(filePath, context);

    // 无内容的文件，添加空内容警告
    if (!content) {
      return {
        files: [],
        logs: [{
          level: _sharedUtils.Loglevel.WARN,
          message: `${filePath} is empty`,
          filePath
        }]
      };
    }
    const logs = [];
    const option = {
      filePath,
      projectPath,
      projectType: _sharedUtils.ProjectType.ANDDROID_UX,
      fileType,
      onLog: function (log) {
        logs.push(log);
      },
      content: content.toString(),
      docHelp: new _QuickAppDocHelp.QuickAppDocHelper()
    };
    const uxAst = await new _parser.UxParser(option, compilerOption, {}, path => path).parser();
    const tsAst = await new _UxToTypescript.default(option, compilerOption, context).translate(uxAst.ast, uxAst.offsetList || []);
    const filePathParse = _path.default.parse(filePath);
    const code = new _generator.TypescriptGenerator().generate({
      sourceFilePath: filePathParse.base,
      targetFilePath: `${filePathParse.name}.js`,
      ast: tsAst.targetTree,
      mapList: tsAst.mapList
    });
    return {
      files: [{
        path: filePath,
        content: code.code
      }, {
        path: `${filePathParse.dir}/${filePathParse.base}.map`,
        content: tsAst.sourceMap.toString()
      }
      // resource todo
      ],
      logs
    };
  }
  getFileType(filePath, context) {
    const {
      entries
    } = context;
    if (_path.default.basename(filePath) === 'app.ux') {
      return 'app';
    } else if (entries.includes(filePath)) {
      return 'page';
    }
  }
}
var _default = exports.default = AndroidUx;
