// aiot-toolkit/lib/starter/androidRouter/PackageRouter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _qrImage = _interopRequireDefault(require("qr-image"));
var _LinkMode = _interopRequireDefault(require("./LinkMode"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const CLIENT_PORT = 39517;
/**
 * 打包相关的路由配置
 * 用于开发机的httpServer
 */
class PackageRouter {
  constructor(param) {
    this.param = param;
  }
  routeList = [{
    path: '/qrcode',
    // 获取二维码图片
    handler: (ctx, next) => {
      return this.handlerQrCode(ctx, next);
    }
  }, {
    path: '/bundle',
    // 获取 rpk 包
    handler: (ctx, next) => {
      this.recordDevice(ctx);
      return this.handlerDownLoadRpk(ctx, next);
    }
  }];

  /**
   * 下载 rpk 包
   * @param ctx
   * @param next
   */
  handlerDownLoadRpk = (ctx, next) => {
    const {
      projectPath,
      options
    } = this.param;
    const dist = _path.default.join(projectPath, options.releasePath);
    const files = _fsExtra.default.readdirSync(dist).filter(item => _path.default.extname(item) === '.rpk');
    const filePath = files.length ? _path.default.join(dist, files[0]) : '';
    this.download(filePath, ctx);
    return next();
  };
  handlerQrCode = (ctx, next) => {
    const qrText = ctx.origin;
    const image = _qrImage.default.image(qrText, {
      size: 9
    });
    ctx.type = 'image/png';
    ctx.body = image;
    return next();
  };

  /**
   * 记录设备信息
   *
   * 获取请求的 ip 和端口，记录到本机中，最多记录5个
   *
   * 文件内容的格式为：
   * ```
   * records: {
   *    "快应用项目路径":[
   *        {ip:"", port:""},
   *        {ip:"", port:""},
   *        {ip:"", port:""}
   *    ]
   * }
   * ```
   *
   *
   * 此记录的作用是：当重新打包后，获取已连接的设备列表
   * @param ctx
   * @param next
   */
  recordDevice = async ctx => {
    const clientInfo = this.getClientFromRequest(ctx);
    if (!clientInfo?.clientIp) {
      return;
    }
    if (clientInfo.linkMode === _LinkMode.default.WIFI) {
      const max = 5;
      const filePath = this.param.options.clientRecordPath;
      const projectPath = this.param.projectPath;
      const data = {
        sn: clientInfo.sn,
        ip: clientInfo.clientIp,
        port: CLIENT_PORT
      };
      const json = _fsExtra.default.readJSONSync(filePath, {
        throws: false
      }) || {};
      if (!json.records) {
        json.records = {};
      }
      if (!json.records[projectPath]) {
        json.records[projectPath] = [];
      }
      const deviceList = json.records[projectPath];
      if (deviceList.findIndex(item => item.ip === data.ip) < 0) {
        deviceList.push(data);
        if (deviceList.length > max) {
          deviceList.splice(0, deviceList.length - max);
        }
        _fsExtra.default.ensureDirSync(_path.default.dirname(filePath));
        _fsExtra.default.writeJSONSync(filePath, json, {
          spaces: 2
        });
      }
    }
  };
  getClientFromRequest(ctx) {
    const clientIp = _sharedUtils.NetworkUtil.getClientIp(ctx);
    const serverIp = _sharedUtils.NetworkUtil.getIPv4IPAddress();
    const sn = ctx.request.header['device-serial-number'];
    let linkMode = _LinkMode.default.NULL;
    if (clientIp === '127.0.0.1' && sn) {
      linkMode = _LinkMode.default.ADB;
    } else if (clientIp !== '127.0.0.1' && clientIp !== serverIp) {
      linkMode = _LinkMode.default.WIFI;
    }
    return {
      clientIp,
      sn,
      linkMode
    };
  }

  /**
   * http 请求中，用于返回下载文件
   *
   * 调用此方法后，仍需调用 next()
   *
   * @param filePath 文件绝对路径
   * @param ctx 请求上下文
   */
  download(filePath, ctx) {
    if (filePath && _fsExtra.default.existsSync(filePath)) {
      ctx.set('Content-Type', 'application/octet-stream');
      ctx.set('Content-Transfer-Encoding', 'binary');
      ctx.set('Content-Disposition', `attachment; filename=${_path.default.basename(filePath)}`);
      ctx.body = _fsExtra.default.createReadStream(filePath);
    } else {
      ctx.throw('404', 404);
    }
  }
}
var _default = exports.default = PackageRouter;
