// @aiot-toolkit/parser/lib/ux/translate/vela/ScriptToTypescript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var babel = _interopRequireWildcard(require("@babel/core"));
var _path = _interopRequireDefault(require("path"));
var _ParserUtil = _interopRequireDefault(require("../../../utils/ParserUtil"));
var _FeatureConfig = _interopRequireDefault(require("../../config/FeatureConfig"));
var _eslint = require("eslint");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
const BabelPreset = require('@babel/preset-env');

/**
 * ScriptToTypescript
 */
class ScriptToTypescript {
  _systemFeatures = (() => new Set())();
  constructor(options, compilerOption, context) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.context = context;
  }
  async translate(sourceTree,
  // eslint-disable-next-line
  offsetList) {
    // 检查代码错误
    const logs = await this.checkCode(sourceTree);
    if (logs) {
      logs.forEach(log => {
        this.options.onLog(log);
      });
    }
    // 使用babel实现替换，由此实现自动更新source map
    const content = this.babelScript(sourceTree);
    const fileName = this.options.filePath + '.js';
    const targetTree = _ParserUtil.default.createSourceFile(content, fileName);
    this.cacheAndCheckFeature();
    return {
      targetTree,
      mapList: []
    };
  }
  /**
   * babel自定义插件实现替换
   * @param sourceCode
   * @returns
   */
  babelScript(sourceTree) {
    const sourceCode = sourceTree.content ? sourceTree.content : `""`;
    const relativePath = _path.default.relative(this.options.projectPath, this.options.filePath);
    const plugins = [
    // 替换所有的require节点
    [_path.default.resolve(__dirname, './plugins/require.js'), {
      options: this.options,
      systemFeatures: this._systemFeatures,
      compilerOption: this.compilerOption
    }]];
    const pluginList = [{
      enable: this.compilerOption.enableE2e,
      path: _path.default.resolve(__dirname, './plugins/e2e.js')
    }, {
      enable: this.compilerOption.startPage && this.options.fileType === 'app',
      path: _path.default.resolve(__dirname, './plugins/startPage.js')
    }];
    pluginList.forEach(item => {
      if (item.enable) {
        plugins.push([item.path, {
          options: this.options,
          compilerOption: this.compilerOption
        }]);
      }
    });
    try {
      const babelResult = babel.transformSync(sourceCode, {
        sourceMaps: 'inline',
        filename: relativePath,
        sourceFileName: relativePath,
        babelrc: false,
        presets: [[BabelPreset, {
          targets: {
            node: 8
          },
          modules: 'commonjs'
        }]],
        plugins
      });
      return babelResult?.code || '';
    } catch (error) {
      const loc = error.loc;
      const babelError = error;
      let position = undefined;
      let message = babelError.message;
      if (loc) {
        const offsetLine = (sourceTree.sourceCodeLocation?.startLine || 0) - 1;
        const startLine = loc.line + offsetLine;
        const startColumn = loc.column;
        // babel 错误信息中包含行列信息，需要替换
        // 示例：Missing semicolon(123:6), 去掉(123:6)，因为在log中会显示行列信息
        message = message.replace(/\(\d+:\d+\)/, ``);

        // 示例：替换下图中的1 替换为 5， 2 替换为 6， 3 替换为 7
        // ```js
        // 1 | import { a } from '@aiot-toolkit/shared-utils'
        // 2 |
        // 3 | console.log(a)
        // ```
        message = message.replace(/(\d+) \|/g, (match, value) => {
          let newValue = (offsetLine + Number(value)).toString();
          return match.replace(value, newValue);
        });
        position = {
          startLine,
          startColumn,
          pos: loc.index,
          endColumn: startColumn,
          endLine: startLine,
          end: loc.index
        };
      }
      this.options.onLog({
        filePath: relativePath,
        message: message,
        level: _sharedUtils.Loglevel.ERROR,
        position
      });
    }
    return '';
  }

  /**
   * 缓存并检查 feature
   */
  cacheAndCheckFeature() {
    const {
      onLog
    } = this.options;
    const errorFeatures = [];
    const translateCache = this.context.translateCache;
    this._systemFeatures.forEach(item => {
      if (!_FeatureConfig.default.isEffectSystemModule(item)) {
        errorFeatures.push(item);
      }
      if (translateCache) {
        translateCache.featureCache.add(item);
      }
    });
    if (errorFeatures.length) {
      onLog({
        filePath: this.options.filePath,
        level: _sharedUtils.Loglevel.WARN,
        message: [`unknown feature：\r\n`, {
          word: errorFeatures.map((item, index) => `${index + 1}. ${item}`).join('\r\n')
        }, '\r\n']
      });
    }
  }

  /**
   * 检查代码
   */
  async checkCode(sourceTree) {
    const code = sourceTree.content;
    if (!code) {
      return;
    }
    const lintResult = await new _eslint.ESLint({
      useEslintrc: false,
      baseConfig: {
        env: {
          browser: true,
          node: true,
          es2022: true
        },
        parserOptions: {
          sourceType: 'module',
          ecmaVersion: 'latest'
        },
        rules: {
          '*': 'off',
          'no-undef': 'error',
          'no-unused-vars': 'warn',
          'max-lines-per-function': ['warn', 500]
        }
      }
    }).lintText(code);
    if (lintResult?.length) {
      const result = [];
      const offsetLine = (sourceTree.sourceCodeLocation?.startLine || 0) - 1;
      lintResult.forEach(item => {
        item.messages.forEach(message => {
          result.push({
            filePath: this.options.filePath,
            level: _sharedUtils.Loglevel.WARN,
            message: message.message,
            position: {
              startLine: message.line + offsetLine,
              startColumn: message.column
            }
          });
        });
      });
      return result;
    }
  }
}
var _default = exports.default = ScriptToTypescript;
