// @aiot-toolkit/parser/lib/ux/translate/vela/UxToTypescript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ScriptToTypescript = _interopRequireDefault(require("./ScriptToTypescript"));
var _StyleToTypescript = _interopRequireDefault(require("./StyleToTypescript"));
var _TemplateToTypescript = _interopRequireDefault(require("./TemplateToTypescript"));
var _SourceMapUtil = _interopRequireDefault(require("./utils/SourceMapUtil"));
var _UxUtil = _interopRequireDefault(require("../../utils/UxUtil"));
var _ParserUtil = _interopRequireDefault(require("../../../utils/ParserUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * UxToTypescript
 */
class UxToTypescript {
  constructor(options, integratedFunction, compilerOption, context) {
    this.options = options;
    this.integratedFunction = integratedFunction;
    this.compilerOption = compilerOption;
    this.context = context;
  }
  async translate(sourceTree,
  // eslint-disable-next-line
  offsetList) {
    const appImport = this.importToTypescript(sourceTree.import);
    const appStyle = new _StyleToTypescript.default(this.options, this.compilerOption).translate(sourceTree.style, []);
    const appTemplate = await new _TemplateToTypescript.default(this.options, this.compilerOption, sourceTree.import.map(item => item.name.toLowerCase())).translate(sourceTree.template, []);
    const appScript = await new _ScriptToTypescript.default(this.options, this.compilerOption, this.context).translate(sourceTree.script, []);
    // 切割出源码和map信息
    const spliteResult = _SourceMapUtil.default.splitSourceMap(appScript.targetTree);
    const integratedList = this.integratedFunction.bind(this, appImport, appStyle.targetTree, appTemplate.targetTree, spliteResult.newScriptTree, this.options)();
    // 源码的行偏移量
    // 减1的原因：parse5解析的行数是从1开始的，而sourcemap的行数是从0开始的
    const scriptOriginOffsetLine = this.options.scriptOffsetLine ? this.options.scriptOffsetLine - 1 : 0;
    // 目标代码的行偏移量
    // 这里原本也需要减1，但是因为integratedList 最后会 join("\n"), script 前面增加了\n， 所以抵消了
    let offsetLines = _SourceMapUtil.default.countStatementLines(integratedList[0]) + _SourceMapUtil.default.countStatementLines(integratedList[1]);
    // 更新source map
    const newGenerator = await _SourceMapUtil.default.updateSourceMap(spliteResult.scriptSourceMap, {
      origin: {
        line: scriptOriginOffsetLine,
        col: 0
      },
      target: {
        line: offsetLines,
        col: 0
      }
    }, this.options);
    return {
      targetTree: _ParserUtil.default.createSourceFile(integratedList.filter(Boolean).join('\n'), this.options.filePath),
      mapList: [],
      sourceMap: newGenerator
    };
  }
  /**
   *
   * @param importList
   * @returns
   */
  importToTypescript(importList) {
    return _UxUtil.default.importToTypescript(importList, this.compilerOption);
  }
}
var _default = exports.default = UxToTypescript;
