// @aiot-toolkit/aiotpack/lib/compiler/javascript/vela/plugin/WrapPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _core = require("@rspack/core");
var _webpackSources = require("webpack-sources");
class WrapPlugin {
  constructor(compilerOption) {
    this.compilerOption = compilerOption;
  }
  apply(compiler) {
    // 给入口文件加上包裹函数
    compiler.hooks.compilation.tap('WrapPlugin', compilation => {
      compilation.hooks.processAssets.tap({
        name: 'WrapPlugin',
        stage: _core.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE
      }, () => {
        this.wrap(compilation);
      });
    });
  }
  wrap(compilation) {
    const {
      enableE2e
    } = this.compilerOption;
    // 获取入口文件
    const entrys = Object.keys(compilation.options.entry).map(item => `${item}.js`);
    // 从chunk找到所有入口文件，添加包裹函数
    entrys.forEach(entry => {
      if (compilation.assets[entry]) {
        const source = compilation.assets[entry];
        const isApp = entry === 'app.js';
        const createFuncnName = isApp ? 'createAppHandler' : 'createPageHandler';
        compilation.assets[entry] = new _webpackSources.ConcatSource(`export default function(global, globalThis, window, $app_exports$, $app_evaluate$){
          var org_app_require = $app_require$;
        
          (function(global, globalThis, window, $app_exports$, $app_evaluate$){
            var setTimeout = global.setTimeout;
            var setInterval = global.setInterval;
            var clearTimeout = global.clearTimeout;
            var clearInterval = global.clearInterval;
            var $app_require$ = global.$app_require$ || org_app_require

            ${enableE2e ? `globalThis = undefined; \n global = typeof window === "undefined" ? global.__proto__  : window;` : ''}
            var ${createFuncnName} = function() {
              return `, source, `
            }
        
            return ${createFuncnName}();
          })(global, globalThis, window, $app_exports$, $app_evaluate$)
        }`);
      }
    });
  }
  translateStyleFunc() {
    return `
var $translateStyle$ = function (value) {
  if (typeof value === 'string') {
    return Object.fromEntries(
      value
        .split(';')
        .filter((item) => Boolean(item && item.trim()))
        .map((item) => {
          const matchs = item.match(/([^:]+):(.*)/);
          if (matchs && matchs.length > 2) {
            return [
              matchs[1]
                .trim()
                .replace(/-([a-z])/g, (_, match) => match.toUpperCase()),
              matchs[2].trim(),
            ];
          }
          return [];
        })
    );
  }
  return value;
};
global.$translateStyle$ = $translateStyle$
`;
  }
}
var _default = exports.default = WrapPlugin;
