// @aiot-toolkit/aiotpack/lib/config/UxConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parser = require("@aiot-toolkit/parser");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _JsLoader = _interopRequireDefault(require("../loader/ux/JsLoader"));
var _PngLoader = _interopRequireDefault(require("../loader/ux/PngLoader"));
var _AppUxLoader = _interopRequireDefault(require("../loader/ux/vela/AppUxLoader"));
var _HmlLoader = _interopRequireDefault(require("../loader/ux/vela/HmlLoader"));
var _UxLoader = _interopRequireDefault(require("../loader/ux/vela/UxLoader"));
var _UxLoader2 = _interopRequireDefault(require("../loader/ux/android/UxLoader"));
var _UxBeforeWorks = _interopRequireDefault(require("../beforeWorks/ux/UxBeforeWorks"));
var _UxAfterWorks = _interopRequireDefault(require("../afterWorks/ux/UxAfterWorks"));
var _UxAfterCompile = _interopRequireDefault(require("../afterCompile/ux/UxAfterCompile"));
var _UxBeforeCompile = _interopRequireDefault(require("../beforeCompile/ux/UxBeforeCompile"));
var _BeforeCompileUtils = _interopRequireDefault(require("../utils/BeforeCompileUtils"));
var _IChangedFile = require("file-lane/lib/interface/IChangedFile");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class UxConfig {
  constructor(projectPath) {
    this.projectPath = projectPath;
  }
  /**
   * 1. 取项目中所有的真实文件
   *  1.1 无entryFileList时
   *  1.2 有entryFileList且文件列表中有 与其他文件相关 的文件时
   *  1.3 有entryFileList且文件列表中有文件为 删除 操作时
   * 2. 返回符合条件的文件列表
   *  2.1 有entryFileList且文件列表中 只有 不影响其他文件 的文件时
   * @param entryFileList
   * @returns
   */
  collectFile = entryFileList => {
    // 取项目中所有的真实文件
    const getProjectFiles = () => {
      const projectPath = this.projectPath;
      let files = _sharedUtils.FileUtil.readAlldirSync(projectPath, undefined, this.exclude);
      return files;
    };
    // 2.
    if (entryFileList) {
      let isGetAll = entryFileList.every(item => item.type === _IChangedFile.HandlerType.UNLINK || !/\.(ux|json|js)$/.test(item.path));
      if (!isGetAll) {
        return entryFileList.filter(filePath => _sharedUtils.FileUtil.include(filePath.path, undefined, this.exclude)).map(item => item.path);
      }
    }
    // 1.
    return getProjectFiles();
  };
  exclude = ['**/node_modules{,/**}', '**/dist{,/**}', '**/build{,/**}', '**/.git{,/**}', filePath => {
    // 如果 是 script， style文件，且同路径下存在同名 hml，则忽略
    const {
      ext,
      name
    } = _path.default.parse(filePath);
    if (![..._parser.ExtensionConfig.SCRIPTS, ..._parser.ExtensionConfig.STYLES].includes(ext)) {
      return false;
    }
    const hmlPath = _sharedUtils.FileUtil.updateFileName(filePath, `${name}${_parser.ExtensionConfig.HML}`);
    if (_fs.default.existsSync(hmlPath)) {
      return true;
    }
    return false;
  }];
  get output() {
    const name = _path.default.basename(this.projectPath);
    const result = `../.temp_${name}`;
    return result;
  }
  beforeWorks = (() => [_UxBeforeWorks.default.cleanOutput])();
  beforeCompile = (() => [_UxBeforeCompile.default.validateManifest, _UxBeforeCompile.default.validateSitemap, _BeforeCompileUtils.default.clean, _BeforeCompileUtils.default.getEntries, _BeforeCompileUtils.default.getGlobalVar])();
  afterCompile = (() => [{
    worker: _UxAfterCompile.default.writeGitIgnore
  }, {
    worker: _UxAfterCompile.default.symlinkNodeModule,
    workerDescribe: 'Create a soft link to the node_modules folder'
  }, {
    worker: _UxAfterCompile.default.compileJavascript,
    workerDescribe: 'Compile javascript project'
  }, {
    worker: _UxAfterCompile.default.copyResource,
    workerDescribe: 'Copy resource files'
  }, {
    worker: _UxAfterCompile.default.compressResource,
    workerDescribe: 'Compress resource files'
  }, {
    worker: _UxAfterCompile.default.jsc,
    workerDescribe: 'Generate jsc bytecode'
  }, {
    worker: _UxAfterCompile.default.protobuf,
    workerDescribe: 'Generate protobuf json'
  }, {
    worker: _UxAfterCompile.default.compileLiteCard,
    workerDescribe: 'compile lite card'
  }, {
    worker: _UxAfterCompile.default.toRpk,
    workerDescribe: 'Package the project into an RPK file'
  }, {
    worker: _UxAfterCompile.default.generateDiff,
    workerDescribe: 'Generate diff json'
  }, {
    worker: _UxAfterCompile.default.moveBackResult,
    workerDescribe: 'Migrate temporary project'
  }, {
    worker: _UxAfterCompile.default.resourceCheck,
    workerDescribe: 'Check resource'
  }])();
  afterWorks = (() => [_UxAfterWorks.default.cleanOutput])();
  watchIgnores = [/node_modules/, /build/, /dist/];

  /**
   * 通过项目类型，返回模块配置
   */
  get module() {
    const isVela = this.getProjectType() === _sharedUtils.ProjectType.VELA_UX;
    return isVela ? {
      rules: [{
        test: ['app.ux'],
        loader: [_AppUxLoader.default]
      }, {
        test: [/.+\.ux$/],
        exclude: [/app\.ux/],
        loader: [_UxLoader.default]
      }, {
        test: [/.+\.hml$/],
        loader: [_HmlLoader.default, _UxLoader.default]
      }, {
        test: [/.+\.js$/],
        loader: [_JsLoader.default]
      }, {
        test: [/.+\.9.png/],
        loader: [_PngLoader.default]
      }]
    } : {
      rules: [{
        test: [/.+\.ux$/],
        loader: [_UxLoader2.default]
      }, {
        test: [/.+\.js$/],
        loader: [_JsLoader.default]
      }, {
        test: [/.+\.9.png/],
        loader: [_PngLoader.default]
      }]
    };
  }

  /**
   * 判断项目类型
   *
   * 目前的办法是通过 manifest.json 的 deviceTypeList 是否有 watch 判断
   */
  getProjectType() {
    return _sharedUtils.ProjectType.getProjectType(this.projectPath);
  }
}
var _default = exports.default = UxConfig;
