// @aiot-toolkit/parser/lib/ux/translate/android/UxToTypescript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _ParserUtil = _interopRequireDefault(require("../../../utils/ParserUtil"));
var _ScriptToTypescript = _interopRequireDefault(require("../vela/ScriptToTypescript"));
var _SourceMapUtil = _interopRequireDefault(require("../vela/utils/SourceMapUtil"));
var _StyleToTypescript = _interopRequireDefault(require("./StyleToTypescript"));
var _TemplateToTypescript = _interopRequireDefault(require("./TemplateToTypescript"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * UxToTypescript
 */
class UxToTypescript {
  contenAccess = `\n
  const moduleOwn = exports.default || module.exports
  const accessors = ['public', 'protected', 'private']

  if (moduleOwn.data && accessors.some(function (acc) { return moduleOwn[acc] })) {
    throw new Error('页面VM对象中的属性data不可与"' + accessors.join(',') + '"同时存在，请使用private替换data名称')
  }
  else if (!moduleOwn.data) {
    moduleOwn.data = {}
    moduleOwn._descriptor = {}
    accessors.forEach(function (acc) {
      const accType = typeof moduleOwn[acc]
      if (accType === 'object') {
        moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc])
        for (const name in moduleOwn[acc]) {
          moduleOwn._descriptor[name] = { access: acc }
        }
      }
      else if (accType === 'function') {
        console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象')
      }
    })
  }`;
  constructor(options, compilerOption, contxt) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.contxt = contxt;
  }
  /**
   *
   * 转换过程
   * 1. 源树分为 import  template  script  style 4部分
   * 2. 转换import: 此步骤使用 import
   * 3. 转换template: 此步骤使用 template
   * 4. 转换script: 此步骤使用 import  script
   * 5. 转换style: 此步骤使用 style
   * 6. 添加包裹函数
   * @param sourceTree
   * @param offsetList
   * @param param
   */
  async translate(sourceTree, offsetList) {
    const sourceFile = _ParserUtil.default.createSourceFile('', _path.default.basename(this.options.filePath));
    // 1
    const {
      template,
      style,
      script,
      import: importList
    } = sourceTree;
    const importTarget = await this.importToTypescript(importList);
    const templateTarget = await new _TemplateToTypescript.default(this.options, this.compilerOption, importList).translate(template, offsetList);
    const scriptTarget = await new _ScriptToTypescript.default(this.options, this.compilerOption, this.contxt).translate(script, []);
    // 切割script中的转换结果和source map
    const spliteResult = _SourceMapUtil.default.splitSourceMap(scriptTarget.targetTree);
    const styleTarget = new _StyleToTypescript.default(this.options, this.compilerOption).translate(style, []);
    const codeList = this.wrapUx({
      imports: importTarget,
      template: templateTarget.targetTree,
      script: spliteResult.newScriptTree,
      style: styleTarget.targetTree
    });
    sourceFile.addStatements(codeList.filter(Boolean).join('\n'));
    const scriptOriginOffsetLine = this.options.scriptOffsetLine ? this.options.scriptOffsetLine - 1 : 0;

    // 目标代码的行偏移量
    let offsetLines = _SourceMapUtil.default.countStatementLines(codeList[0]) + _SourceMapUtil.default.countStatementLines(codeList[1]);
    return {
      targetTree: sourceFile,
      mapList: [],
      sourceMap: await _SourceMapUtil.default.updateSourceMap(spliteResult.scriptSourceMap, {
        origin: {
          line: scriptOriginOffsetLine,
          col: 0
        },
        target: {
          line: offsetLines,
          col: 0
        }
      }, this.options)
    };
  }
  importToTypescript(importList) {
    return importList.map(item => {
      return `$app_define$("@app-component/${item.name.toLowerCase()}", [], require("${item.src}"))`;
    });
  }
  wrapUx = data => {
    const {
      fileType
    } = this.options;
    switch (fileType) {
      case 'page':
        return this.wrapPageUx(data);
      case 'app':
        return this.wrapAppUx(data);
      default:
        return this.wrapCommonUx(data);
    }
  };
  wrapPageUx = data => {
    const {
      filePath
    } = this.options;
    const fileParse = _path.default.parse(filePath);
    const common = this.wrapCommonUx(data);
    return [...common, `
      $app_define$(
      "@app-component/${fileParse.name}",
      [], module.exports)
      $app_bootstrap$("@app-component/${fileParse.name}", { packagerVersion: '${this.getToolVersion()}'});
      `];
  };
  getToolVersion() {
    return require('../../../../package.json').version;
  }
  wrapCommonUx = data => {
    const {
      imports,
      template,
      script,
      style
    } = data;
    const scriptValue = this.wrapScript(script);
    return [imports.join('\n'), [`const $style$ = ${JSON.stringify(style)}`, `const $template$ = ${template?.getFullText() || `''`}`].join('\n'), `const $script$=${scriptValue || 'null'}`, `module.exports = function ($app_require$, $app_exports$, $app_module$) {
          ${this.ruleScript(Boolean(scriptValue))}
          $app_module$.exports.template = $template$
          $app_module$.exports.style = $style$;
       }
      `];
  };
  wrapAppUx = data => {
    const {
      imports,
      script,
      style
    } = data;
    const scriptValue = this.wrapScript(script);
    return [`${imports.join()}`, `const $style$ = ${JSON.stringify(style)}`, `const $script$=${scriptValue || 'null'}`, `$app_define$('@app-application/app', [], function ($app_require$, $app_exports$, $app_module$) {
          ${this.ruleScript(Boolean(scriptValue))}
          $app_module$.exports.manifest = require("./manifest.json")
          $app_module$.exports.style = { list: [$style$] }
      })
      $app_bootstrap$('@app-application/app', { packagerVersion: "${this.getToolVersion()}" })`];
  };
  ruleScript(hasScript) {
    if (hasScript) {
      return `$script$($app_module$, $app_exports$, $app_require$)
              if ($app_exports$.__esModule && $app_exports$.default) {
                  $app_module$.exports = $app_exports$.default
              }`;
    }
    return '';
  }
  wrapScript(script) {
    const code = script?.getFullText();
    const {
      fileType
    } = this.options;
    return code ? `function __scriptModule__(module, exports, $app_require$) {${code}
      ${fileType === 'page' ? this.contenAccess : ''}}` : '';
  }
}
var _default = exports.default = UxToTypescript;
