// @aiot-toolkit/parser/lib/ux/config/vela/ElementConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const graphCommonAttr = {
  opacity: {},
  'stroke-opacity': {},
  'fill-opacity': {},
  fill: {},
  stroke: {},
  'stroke-width': {},
  'stroke-linecap': {},
  'stroke-dasharray': {},
  'stroke-dashoffset': {},
  transform: {}
};
const elementConfig = {
  commonEvents: ['click', 'focus', 'blur', 'key', 'longpress', 'appear', 'disappear', 'swipe', 'touchstart', 'touchmove', 'touchend', 'touchcancel', 'resize', 'animationstart', 'animationiteration', 'animationend'],
  commonAttributes: {
    id: {},
    key: {},
    style: {},
    class: {},
    disabled: {
      enums: ['true', 'false', '']
    },
    if: {},
    elif: {},
    else: {},
    for: {},
    tid: {},
    show: {},
    'aria-label': {},
    'aria-unfocusable': {
      enums: ['false', 'true']
    },
    forcedark: {
      enums: ['true', 'false']
    },
    focusable: {
      enums: ['false', 'true']
    },
    vtags: {},
    vindex: {},
    autofocus: {
      enums: ['false', 'true']
    },
    descendantfocusability: {
      enums: ['before', 'after', 'block']
    },
    static: {}
  },
  elementConfig: {
    img: {
      aliasName: 'image',
      events: ['complete', 'error'],
      attributes: {
        src: {},
        alt: {},
        enablenightmode: {
          enums: ['true', 'false', '']
        },
        autoplay: {
          enums: ['true', 'false', '']
        }
      }
    },
    div: {
      attributes: {
        enablevideofullscreencontainer: {
          enums: ['false', 'true']
        }
      }
    },
    a: {
      allowTextChildren: true,
      attributes: {
        visited: {
          enums: ['false', 'true']
        },
        href: {},
        value: {}
      }
    },
    text: {
      allowTextChildren: true,
      attributes: {
        value: {},
        type: {
          enums: ['text', 'html']
        }
      }
    },
    'arc-text': {
      allowTextChildren: true,
      attributes: {
        value: {},
        'start-angle': {},
        direction: {
          enums: ['clockwise', 'counterclockwise']
        }
      }
    },
    span: {
      allowTextChildren: true,
      attributes: {
        value: {}
      }
    },
    label: {
      attributes: {
        target: {}
      }
    },
    image: {
      events: ['complete', 'error'],
      attributes: {
        src: {},
        alt: {},
        enablenightmode: {
          enums: ['true', 'false', '']
        },
        autoplay: {
          enums: ['true', 'false', '']
        }
      }
    },
    maml: {
      events: ['complete', 'error'],
      attributes: {
        src: {},
        alt: {},
        enablenightmode: {
          enums: ['true', 'false', '']
        },
        autoplay: {
          enums: ['true', 'false', '']
        }
      }
    },
    slider: {
      events: ['change'],
      attributes: {
        enabled: {
          enums: ['true', 'false', '']
        },
        min: {},
        max: {},
        step: {},
        value: {}
      }
    },
    web: {
      events: ['pagestart', 'pagefinish', 'titlereceive', 'error', 'message', 'progress'],
      attributes: {
        src: {},
        trustedurl: {},
        allowthirdpartycookies: {
          enums: ['true', 'false', '']
        },
        enablenightmode: {
          enums: ['true', 'false', '']
        },
        showloadingdialog: {
          enums: ['true', 'false', '']
        },
        supportzoom: {
          enums: ['true', 'false', '']
        }
      }
    },
    list: {
      events: ['scroll', 'scrollbottom', 'scrolltop', 'scrollend', 'scrolltouchup', 'selected'],
      attributes: {
        scrollpage: {
          enums: ['true', 'false', '']
        },
        focusbehavior: {
          enums: ['aligned', 'edged', 'reverseedged', 'leadingedged', 'trailingedged']
        },
        bounces: {}
      }
    },
    'list-item': {
      attributes: {
        type: {
          required: true
        }
      }
    },
    block: {
      attributes: {
        static: {},
        dynamic: {},
        data: {}
      }
    },
    component: {
      attributes: {
        is: {
          forceDynamic: true
        },
        remotewidget: {}
      }
    },
    slot: {
      attributes: {
        content: {},
        name: {}
      }
    },
    input: {
      events: ['change', 'enterkeyclick', 'selectionchange'],
      attributes: {
        type: {
          enums: ['text', 'button', 'checkbox', 'radio', 'email', 'date', 'time', 'number', 'password', 'tel', 'eventbutton']
        },
        autocomplete: {
          enums: ['on', 'off']
        },
        enterkeytype: {
          enums: ['default', 'next', 'go', 'done', 'send', 'search']
        },
        eventtype: {
          enums: ['shortcut']
        },
        model: {},
        'true-value': {},
        'false-value': {},
        maxlength: {},
        checked: {
          enums: ['false', 'true', '']
        },
        name: {},
        value: {},
        placeholder: {}
      }
    },
    refresh: {
      events: ['refresh'],
      attributes: {
        offset: {},
        refreshing: {
          enums: ['false', 'true']
        },
        type: {
          enums: ['auto', 'pulldown']
        },
        'enable-refresh': {
          enums: ['true', 'false']
        }
      }
    },
    refresh2: {
      events: ['pulldownrefresh', 'pulluprefresh', 'refresh'],
      attributes: {
        pulldownrefreshing: {
          enums: ['false', 'true', '']
        },
        pulluprefreshing: {
          enums: ['false', 'true', '']
        },
        animationduration: {},
        enablepulldown: {
          enums: ['false', 'true', '']
        },
        enablepullup: {
          enums: ['false', 'true', '']
        },
        reboundable: {
          enums: ['false', 'true', '']
        },
        gesture: {
          enums: ['false', 'true', '']
        },
        offset: {},
        refreshing: {
          enums: ['false', 'true', '']
        },
        type: {
          enums: ['auto', 'pulldown']
        }
      }
    },
    'refresh-header': {
      events: ['move'],
      attributes: {
        dragrate: {},
        triggerratio: {},
        tiggersize: {},
        maxdragratio: {},
        maxdragsize: {},
        refreshdisplayratio: {},
        refreshdisplaysize: {},
        spinnerstyle: {
          enums: ['translation', 'front', 'behind']
        },
        autorefresh: {
          enums: ['false', 'true', '']
        },
        translationwithcontent: {
          enums: ['false', 'true', '']
        }
      }
    },
    'refresh-footer': {
      events: ['move'],
      attributes: {
        dragrate: {},
        triggerratio: {},
        tiggersize: {},
        maxdragratio: {},
        maxdragdize: {},
        refreshdisplayratio: {},
        refreshdisplaysize: {},
        spinnerstyle: {
          enums: ['translation', 'front', 'behind']
        },
        autorefresh: {
          enums: ['false', 'true', '']
        },
        translationwithcontent: {
          enums: ['false', 'true', '']
        }
      }
    },
    ad: {
      events: ['load', 'error'],
      attributes: {
        unitid: {},
        type: {}
      }
    },
    swiper: {
      events: ['change', 'swipestart', 'swipeend'],
      attributes: {
        index: {},
        autoplay: {
          enums: ['false', 'true', '']
        },
        interval: {},
        indicator: {
          enums: ['false', 'true', '']
        },
        loop: {
          enums: ['false', 'true', '']
        },
        duration: {},
        vertical: {
          enums: ['false', 'true', '']
        },
        previousmargin: {},
        nextmargin: {},
        enableswipe: {
          enums: ['false', 'true', '']
        }
      }
    },
    progress: {
      attributes: {
        percent: {},
        type: {
          enums: ['horizontal', 'circular', 'arc']
        }
      }
    },
    picker: {
      events: ['change', 'columnchange', 'cancel'],
      attributes: {
        type: {
          required: true,
          enums: ['text', 'date', 'time', 'multi-text']
        },
        start: {},
        end: {},
        range: {},
        selected: {},
        value: {}
      }
    },
    switch: {
      events: ['change'],
      attributes: {
        checked: {
          enums: ['false', 'true', '']
        }
      }
    },
    textarea: {
      events: ['change', 'selectionchange', 'linechange'],
      attributes: {
        placeholder: {},
        maxlength: {},
        model: {}
      }
    },
    video: {
      events: ['playing', 'prepared', 'start', 'pause', 'finish', 'error', 'seeking', 'seeked', 'timeupdate', 'fullscreenchange'],
      attributes: {
        src: {},
        muted: {
          enums: ['true', 'false', '']
        },
        autoplay: {
          enums: ['true', 'false', '']
        },
        controls: {
          enums: ['true', 'false', '']
        },
        poster: {},
        orientation: {
          enums: ['landscape', 'portrait']
        },
        titlebar: {
          enums: ['true', 'false', '']
        },
        title: {},
        playcount: {},
        speed: {}
      }
    },
    camera: {
      events: ['error', 'camerainitdone', 'cameraframe'],
      attributes: {
        deviceposition: {
          enums: ['back', 'front']
        },
        flash: {
          enums: ['auto', 'on', 'off', 'torch']
        },
        framesize: {
          enums: ['low', 'normal', 'high']
        },
        autoexposurelock: {
          enums: ['true', 'false', '']
        },
        autowhitebalancelock: {
          enums: ['true', 'false', '']
        }
      }
    },
    map: {
      events: ['loaded', 'regionchange', 'tap', 'markertap', 'callouttap', 'controltap', 'poitap'],
      attributes: {
        latitude: {},
        longitude: {},
        coordtype: {},
        scale: {},
        rotate: {},
        markers: {},
        showmylocation: {
          enums: ['true', 'false', '']
        },
        polylines: {},
        polygons: {},
        circles: {},
        controls: {},
        groundoverlays: {},
        includepoints: {},
        heatmaplayer: {},
        showcompass: {
          enums: ['true', 'false', '']
        },
        enableoverlooking: {
          enums: ['true', 'false', '']
        },
        enablezoom: {
          enums: ['true', 'false', '']
        },
        enablescroll: {
          enums: ['true', 'false', '']
        },
        enablerotate: {
          enums: ['true', 'false', '']
        },
        showscale: {
          enums: ['true', 'false', '']
        },
        showzoom: {
          enums: ['true', 'false', '']
        }
      }
    },
    custommarker: {
      attributes: {
        custommarkerattr: {}
      }
    },
    canvas: {},
    stack: {
      events: ['fullscreenchange']
    },
    richtext: {
      allowTextChildren: true,
      events: ['start', 'complete'],
      attributes: {
        value: {},
        type: {
          required: true,
          enums: ['html', 'ux', 'mix']
        }
      }
    },
    tabs: {
      events: ['change'],
      attributes: {
        index: {}
      }
    },
    'tab-content': {
      attributes: {
        scrollable: {
          enums: ['true', 'false', '']
        }
      }
    },
    'tab-bar': {
      attributes: {
        mode: {
          enums: ['fixed', 'scrollable']
        }
      }
    },
    popup: {
      events: ['visibilitychange'],
      attributes: {
        target: {
          required: true
        },
        placement: {
          enums: ['left', 'top', 'right', 'bottom', 'topLeft', 'topRight', 'bottomLeft', 'bottomRight']
        }
      }
    },
    rating: {
      events: ['change'],
      attributes: {
        numstars: {},
        rating: {},
        stepsize: {},
        indicator: {
          enums: ['false', 'true']
        }
      }
    },
    marquee: {
      allowTextChildren: true,
      events: ['bounce', 'finish', 'start'],
      attributes: {
        scrollamount: {},
        loop: {},
        direction: {
          enums: ['left', 'right']
        },
        value: {},
        'text-offset': {}
      }
    },
    scrollview: {
      events: ['scroll', 'scrollbegindrag', 'scrollenddrag', 'scrollstart', 'scrollstop'],
      attributes: {
        'scroll-direction': {
          enums: ['vertical', 'horizontal', 'vertical_horizontal']
        },
        'show-scrollbar': {
          enums: ['true', 'false', '']
        }
      }
    },
    drawer: {
      events: ['change', 'scroll'],
      attributes: {
        enableswipe: {
          enums: ['true', 'false', '']
        }
      }
    },
    lottie: {
      events: ['complete', 'error', 'change'],
      attributes: {
        source: {
          required: true
        },
        progress: {},
        speed: {},
        loop: {
          enums: ['false', 'true', '']
        },
        autoplay: {
          enums: ['false', 'true', '']
        },
        rendermode: {
          enums: [`AUTOMATIC`, `HARDWARE`, `SOFTWARE`]
        }
      }
    },
    'drawer-navigation': {
      attributes: {
        direction: {
          enums: ['start', 'end']
        }
      }
    },
    'slide-view': {
      events: ['open', 'close', 'slide', 'buttonclick'],
      attributes: {
        edge: {
          enums: ['right', 'left']
        },
        enableslide: {
          enums: ['true', 'false', '']
        },
        isopen: {
          enums: ['false', 'true', '']
        },
        layer: {
          enums: ['above', 'same']
        },
        buttons: {}
      }
    },
    'section-list': {
      events: ['scroll', 'scrollend', 'scrolltouchup', 'scrolltop', 'scrollbottom']
    },
    'section-group': {
      events: ['change'],
      attributes: {
        expand: {
          enums: ['false', 'true', '']
        }
      }
    },
    'section-header': {},
    'section-item': {},
    'share-button': {
      events: ['success', 'fail', 'cancel'],
      attributes: {
        value: {},
        title: {},
        description: {},
        icon: {},
        url: {},
        path: {},
        params: {},
        platforms: {},
        usepageparams: {
          enums: ['true', 'false', '']
        }
      }
    },
    'shortcut-button': {
      attributes: {
        value: {}
      }
    },
    qrcode: {
      attributes: {
        value: {}
      }
    },
    scroll: {
      events: ['scrolltop', 'scrollbottom', 'scroll'],
      attributes: {
        'scroll-x': {
          enums: ['true', 'false', '']
        },
        'scroll-y': {
          enums: ['true', 'false', '']
        },
        'scroll-top': {},
        'scroll-bottom': {},
        'scroll-left': {},
        'scroll-right': {},
        bounces: {
          enums: ['true', 'false', '']
        }
      }
    },
    'frame-image': {
      events: ['error'],
      attributes: {
        src: {},
        index: {},
        cache: {
          enums: ['false', 'true', '']
        }
      }
    },
    'number-image': {
      events: ['error'],
      attributes: {
        src: {},
        number: {},
        space: {}
      }
    },
    'chars-image': {
      events: ['error'],
      attributes: {
        string: {},
        'map-list': {},
        src: {},
        space: {}
      }
    },
    time: {
      attributes: {
        format: {},
        src: {},
        space: {},
        update: {}
      }
    },
    svg: {
      attributes: {
        width: {},
        height: {},
        viewBox: {}
      }
    },
    rect: {
      attributes: {
        ...graphCommonAttr,
        width: {},
        height: {},
        x: {},
        y: {},
        rx: {},
        ry: {}
      }
    },
    circle: {
      attributes: {
        ...graphCommonAttr,
        cx: {},
        cy: {},
        r: {}
      }
    },
    ellipse: {
      attributes: {
        ...graphCommonAttr,
        cx: {},
        cy: {},
        rx: {},
        ry: {}
      }
    },
    line: {
      attributes: {
        ...graphCommonAttr,
        x1: {},
        y1: {},
        x2: {},
        y2: {}
      }
    },
    polyline: {
      attributes: {
        ...graphCommonAttr,
        points: {}
      }
    },
    polygon: {
      attributes: {
        ...graphCommonAttr,
        points: {}
      }
    },
    path: {
      attributes: {
        ...graphCommonAttr,
        d: {}
      }
    },
    barcode: {
      attributes: {
        value: {}
      }
    },
    chart: {
      attributes: {
        type: {},
        options: {},
        datasets: {}
      }
    },
    'image-animator': {
      attributes: {
        images: {},
        iteration: {},
        reverse: {},
        fixedsize: {},
        duration: {},
        fillmode: {
          enums: ['none', 'forwards']
        }
      }
    }
  }
};
var _default = exports.default = elementConfig;
