// vela-quickapp-test/vela-quickapp-test.js
// EventEmitter start
class EventEmitter {
  constructor() {
    this.fnList = new Map()
  }
  on(name, fn) {
    let callBackList = this.fnList.get(name)
    if (callBackList) {
      this.fnList.set(name, [fn, ...callBackList])
    } else {
      this.fnList.set(name, [fn])
    }
  }

  emit(name, ...args) {
    this.fnList && this.fnList.get(name) && this.fnList.get(name).forEach(item => {
      item.apply(this, args)
    })
  }

  off(name) {
    this.fnList.delete(name)
  }

  once(name, fn) {
    function onceFn() {
      fn()
      this.off(name)
    }
    this.on(name, onceFn)
  }
}
// EventEmitter end
// Runner start
// 监听事件的标识
const constants = {
  EVENT_RUN_BEGIN: 'EVENT_RUN_BEGIN',      // 执行流程开始
  EVENT_RUN_END: 'EVENT_RUN_END',          // 执行流程结束
  EVENT_SUITE_BEGIN: 'EVENT_SUITE_BEGIN',  // 执行suite开始
  EVENT_SUITE_END: 'EVENT_SUITE_END',      // 执行suite开始
  EVENT_TEST_BEGIN: 'EVENT_TEST_BEGIN',      // 执行用例开始
  EVENT_FAIL: 'EVENT_FAIL',                // 执行用例失败
  EVENT_PEDING: 'EVENT_PEDING',
  EVENT_PASS: 'EVENT_PASS'                 // 执行用例成功
}

class Runner extends EventEmitter {
  constructor() {
    super();
    // 记录 suite 根节点到当前节点的路径
    this.suites = [];
  }

  /*
   * 主入口
   */
  async run(root) {
    if (root.hasOnly()) root.filterOnly()
    this.emit(constants.EVENT_RUN_BEGIN);
    await this.runSuite(root);
    this.emit(constants.EVENT_RUN_END);
  }

  /*
   * 执行suite
   */
  async runSuite(suite) {
    // suite执行开始
    this.emit(constants.EVENT_SUITE_BEGIN, suite);

    // 1）执行before钩子函数
    if (suite._beforeAll.length) {
      for (const fn of suite._beforeAll) {
        const result = await fn();
        if (result instanceof Error) {
          this.emit(constants.EVENT_FAIL, `"before all" hook in ${suite.title}: ${result.message}`);
          // suite执行结束
          this.emit(constants.EVENT_SUITE_END);
          return;
        }
      }
    }
  
    // 路径栈推入当前节点
    this.suites.unshift(suite);
  
    // 2）执行test
    if (suite.tests.length) {
      for (const test of suite.tests) {
        // 处理 pending 情况
        if (suite.isPending) test.isPending = true
        await this.runTest(test, suite);
      }
    }
  
    // 3）执行子级suite
    if (suite.suites.length) {
      for (const child of suite.suites) {
        await this.runSuite(child);
      }
    }
  
    // 路径栈推出当前节点
    this.suites.shift(suite);
  
    // 4）执行after钩子函数
    if (suite._afterAll.length) {
      for (const fn of suite._afterAll) {
        const result = await fn();
        if (result instanceof Error) {
          this.emit(constants.EVENT_FAIL, `"after all" hook in ${suite.title}: ${result.message}`);
          // suite执行结束
          this.emit(constants.EVENT_SUITE_END);
          return;
        }
      }
    }

    // suite结束
    this.emit(constants.EVENT_SUITE_END);
  }
  
  /*
   * 执行suite
   */
  async runTest(test, suite) {
    // 1）由suite根节点向当前suite节点，依次执行beforeEach钩子函数
    const _beforeEach = [].concat(this.suites).reverse().reduce((list, suite) => list.concat(suite._beforeEach), []);
    if (_beforeEach.length) {
      for (const fn of _beforeEach) {
        const result = await fn();
        if (result instanceof Error) {
          return this.emit(constants.EVENT_FAIL, `"before each" hook for ${test.title}: ${result.message}`)
        }
      }
    }
  
    // 2）执行测试用例
    this.emit(constants.EVENT_TEST_BEGIN, test, new Date().getTime(), `${test.title}`, suite.fullTitle());
    // skip
    if (test.isPending) return this.emit(constants.EVENT_PEDING, `${test.title}`);
    const result = await test.fn();
    if (result instanceof Error) {
      return this.emit(constants.EVENT_FAIL, `${test.title}`, test, result);
    } else {
      this.emit(constants.EVENT_PASS, `${test.title}`, test);
    }
  
    // 3）由当前suite节点向suite根节点，依次执行afterEach钩子函数
    const _afterEach = [].concat(this.suites).reduce((list, suite) => list.concat(suite._afterEach), []);
    if (_afterEach.length) {
      for (const fn of _afterEach) {
        const result = await fn();
        if (result instanceof Error) {
          return this.emit(constants.EVENT_FAIL, `"after each" hook for ${test.title}: ${result.message}`)
        }
      }
    }
  }
}
// Runner end

// utils start
function adaptPromise(fn) {
  return () => new Promise(resolve => {
    if (fn.length == 0) { // 不使用参数 done
      try {
        const ret = fn();
        // 判断是否返回promise
        if (ret instanceof Promise) {
          return ret.then(resolve, resolve);
        } else {
          resolve();
        }
      } catch (error) {
        resolve(error);
      }
    } else { // 使用参数 done
      function done(error) {
        resolve(error);
      }
      try {
        fn(done);
      } catch (error) {
        resolve(error);
      }
    }
  })
}
// utils end

// suite start
class Suite {
  constructor (props) {
      this.title = props.title; 
      this.suites = [];
      this.tests = [];
      this.isPending = false;
      this._beforeEach = [];
      this._beforeAll = [];
      this._afterEach = [];
      this._afterAll = [];
      this._onlyTests = [];
      this._onlySuites = [];
      this.parent = props.parent || null;
      if (props.parent instanceof Suite) {
        props.parent.suites.push(this);
      }
  }

  fullTitle() {
    if (this.parent) {
      const full = this.parent.fullTitle();
      if (full) {
        return full + ' ' + this.title;
      }
    }
    return this.title;
  };

  hasOnly() {
    return (
      this._onlyTests.length > 0 ||
      this._onlySuites.length > 0 ||
      this.suites.some(function (suite) {
        return suite.hasOnly();
      })
    );
  };

  filterOnly() {
    if (this._onlyTests.length) {
      this.tests = this._onlyTests;
      this.suites = [];
    } else {
      this.tests = [];
      this._onlySuites.forEach(function (onlySuite) {
        if (onlySuite.hasOnly()) {
          onlySuite.filterOnly();
        }
      });
      var onlySuites = this._onlySuites;
      this.suites = this.suites.filter(function (childSuite) {
        return onlySuites.indexOf(childSuite) !== -1 || childSuite.filterOnly();
      });
    }
    return this.tests.length > 0 || this.suites.length > 0;
  };
}
// suite end

// test start
class Test {
	constructor(props) {
		this.title = props.title;
		this.fn = props.fn;
    this.isPending = false;
	}
}
// test start

// reporters start
const colors = {
  pass: 90,
  fail: 31,
  green: 32,
  pending: 36
}

function color(type, str) {
  return '\u001b[' + colors[type] + 'm' + str + '\u001b[0m';
}

function reporters(runner) {

  let indents = 0;
  let passes = 0;
  let failures = 0;
  let pending = 0;
  // 测试用例信息，包括: startTime、endTime、title、fullTitle
  const casesInfo = new Map()
  const result = {
    stats: {
      suites: 0,
      tests: 0,
      passes: 0,
      failures: 0,
      pending: 0,
      start: '',
      end: '',
      duration: 0,
      title: ''
    },
    tests: [],
    failures: [], // 失败的
    passes: [], // 成功的
    pending: []
  }
  // 格式化结果
  function formatResult() {
    casesInfo.forEach((item, key) => {
      if (key.isPending) {
        delete item.startTime

        result.pending.push(item)
      } else {
        item.duration = item.endTime - item.startTime

        delete item.endTime
        delete item.startTime
  
        if (item.code) {
          result.failures.push(item)
        } else {
          result.passes.push(item)
        }
  
        delete item.code
        result.tests.push(item)
      }
    });

    result.stats.tests = result.tests.length
    result.stats.passes = result.passes.length
    result.stats.pending = result.pending.length
    result.stats.failures = result.failures.length
    result.stats.end = new Date();
    result.stats.duration = result.stats.end.getTime() - result.stats.start.getTime()

    return result
  }

  function indent(i = 0) {
    return Array(indents + i).join('  ');
  }

  // 执行开始
  runner.on(constants.EVENT_RUN_BEGIN, function() {
    // 记录整个测试流程的开始
    result.stats.start = new Date();
  });

  // suite执行开始
  runner.on(constants.EVENT_SUITE_BEGIN, function(suite) {
    ++indents;
    // 记录suites
    if (suite.title) result.stats.suites++
    // 取第一个suite的title作为stats的title
    if (!result.stats.title) result.stats.title = suite.title
    console.log(indent(), suite.title);
  });

  // suite执行结束
  runner.on(constants.EVENT_SUITE_END, function() {
    --indents;
    if (indents == 1) console.log();
  });

  // 用例测试开始
  runner.on(constants.EVENT_TEST_BEGIN, function(test, startTime, title, parentTitle) {
    casesInfo.set(test, {
      startTime,
      title,
      fullTitle: parentTitle + ' ' + title
    })
    if (!global.isConcise) {
      const fmt = indent(1) + '  [run]';
      console.log(fmt, title);
    }
  });

  // 用例通过
  runner.on(constants.EVENT_PASS, function(title, test) {
    const caseInfo = casesInfo.get(test)
    if (caseInfo) {
      caseInfo.code = 0
      caseInfo.endTime = new Date().getTime()
    }

    passes++;

    // const fmt = indent(1) + color('green', '  [passed]') + color('pass', ' %s');
    if (!global.isConcise) {
      const fmt = indent(1) + '  [passed]';
      console.log(fmt, title);
    }
  });

  // 用例失败
  runner.on(constants.EVENT_FAIL, function(title, test, err) {
    if (arguments.length === 1) {
      failures++;
      const fmt = indent(1) + color('fail', `  × ${title}`);
      console.log(fmt);
    } else {
      const caseInfo = casesInfo.get(test)
      if (caseInfo) {
        const { actual, expected } = err
  
        caseInfo.code = 1
        caseInfo.endTime = new Date().getTime()
        caseInfo.err = {
          message: err.message || '',
          stack: err.stack || ''
        }
        if (actual !== undefined) caseInfo.err.actual = actual
        if (expected !== undefined) caseInfo.err.expected = expected
      }
      failures++;
  
      if (!global.isConcise) {
        const fmt = indent(1) + '  [failed]';
        console.log(indent(1), ' err:');
        console.log(indent(1), ' ' + JSON.stringify(caseInfo.err));
        console.log(fmt, title);
      }
    }
  });

  // 用例pending
  runner.on(constants.EVENT_PEDING, function(title) {
    pending++;

    // const fmt = indent(1) + color('pending', '  - %s');
    const fmt = indent(1) + '  - ';
    console.log(fmt, title);
  });

  // 执行结束
  runner.once(constants.EVENT_RUN_END, function() {
    // console.log(color('green', '  %d passing'), passes);
    // console.log(color('fail', '  %d failing'), failures);
    // console.log(color('pending', '  %d pending'), pending);

    console.log(indent(2), passes + ' passing');
    console.log(indent(2), failures + ' failing');
    console.log(indent(2), pending + ' pending');

    const result = formatResult()
    // 输出报告
    console.log(JSON.stringify(result, null, "\t"))
    if (global.velaTestResults) global.velaTestResults.push(result)
  });
}
// reporters end

// bdd start
function bdd(context, root) {

  // 记录suite路径
  const suites = [root];

  context.describe = context.context = function (title, fn) {
    const parent = suites[0];
    const suite = new Suite({
      title,
      parent
    });
    suites.unshift(suite);
    fn.call(suite);
    suites.shift(suite);
  }

  context.describe.skip = function(title, fn) {
    const parent = suites[0];
    const suite = new Suite({
      title,
      parent
    });
    suite.isPending = true
    suites.unshift(suite);
    fn.call(suite);
    suites.shift(suite);
  }

  context.describe.only = function(title, fn) {
    const parent = suites[0];
    const suite = new Suite({
      title,
      parent
    });
    suites.unshift(suite);
    parent._onlySuites.push(suite);
    fn.call(suite);
    suites.shift(suite);
  }

  context.it = context.specify = function (title, fn) {
    const parent = suites[0];
    const test = new Test({
      title,
      fn: adaptPromise(fn)
    });
    parent.tests.push(test);
    return { parent, test };
  }

  context.it.skip = function(title, fn) {
    const parent = suites[0];
    const test = new Test({
      title,
      fn: adaptPromise(fn)
    });
    test.isPending = true
    parent.tests.push(test);
  }

  context.it.only = function(title, fn) {
    const { parent, test } = context.it(title, fn);
    parent._onlyTests.push(test)
  };

  context.before = function (fn) {
    const cur = suites[0];
    cur._beforeAll.push(adaptPromise(fn));
  }

  context.after = function (fn) {
    const cur = suites[0];
    cur._afterAll.push(adaptPromise(fn));
  }

  context.beforeEach = function (fn) {
    const cur = suites[0];
    cur._beforeEach.push(adaptPromise(fn));
  }

  context.afterEach = function (fn) {
    const cur = suites[0];
    cur._afterEach.push(adaptPromise(fn));
  }
}
// bdd end

// VelaQuickappTest start
class VelaQuickappTest {
  constructor() {
    // 创建一个根suite
    this.rootSuite = new Suite({
      title: '',
      parent: null
    });

    // 使用bdd测试风格，将API挂载到global对象上
    bdd(global, this.rootSuite);
  }

  async run(callback) {
    const runner = new Runner();
    reporters(runner);
    await runner.run(this.rootSuite);
    // 用于执行页面回调: 例如：history.back()
    if (callback) callback()
  }
}
// VelaQuickappTest end
// setTimeoutDone start
const setTimeoutDone = async function(fn, timeout, done) {

  const mySetTimeout = function () {
    return new Promise((resolve, reject) => {
      setTimeout(_ => {
        try {
          fn()
        } catch (error) {
          resolve(error)
        }
      }, timeout)
    })
  }
  const res = await mySetTimeout()
  if (!!res) {
    if (res instanceof Error) done(res)
    else done(new Error(res))
  }
}
// setTimeoutDone end
// waitForOK start
const waitForOK = function (ms = 10) {
  return new Promise(res => {
    setTimeout(() => {
      res();
    }, ms);
  });
};
// waitForOK end
// NextTime start
class NextTime {
  constructor(start = 0, step = 50) {
    this.time = start
    this.step = step
  }

  getNextTime() {
    this.time = this.time + this.step
    return this.time
  }

  setStep(step) {
    this.step = step
  }
}
const next = new NextTime()
// 默认50ms
const nextTime = next.getNextTime.bind(next)
// NextTime end
export {
  VelaQuickappTest,
  NextTime,
  nextTime,
  setTimeoutDone,
  waitForOK
}
