// @aiot-toolkit/aiotpack/lib/compiler/javascript/vela/utils/UxCompileUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _EntryType = _interopRequireDefault(require("../enum/EntryType"));
var _IManifest = require("../interface/IManifest");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class UxCompileUtil {
  // 摘要文件夹
  static DIGEST_ZIP_DIR = 'META-INF';
  static clean(dirList) {
    if (!dirList) {
      return;
    }
    dirList.forEach(item => {
      if (_fsExtra.default.existsSync(item)) {
        if (_fsExtra.default.statSync(item).isFile()) {
          _fsExtra.default.unlinkSync(item);
        } else {
          _fsExtra.default.rmSync(item, {
            recursive: true,
            force: true
          });
        }
      }
    });
  }

  /**
   * 从项目配置文件中解析出入口文件
   * @param config 项目配置文件的内容，应为json对象
   * @param codeDir 源码目录
   * @param projectPath 项目目录
   * @returns {[入口名]:源文件路径}
   */
  static resolveEntries(config, codeDir, projectPath) {
    const {
      router
    } = config;
    if (!router) {
      throw Error('No routing configured in manifest.json！');
    }
    const appFile = this.resolveFile(_path.default.join(codeDir, 'app'));
    if (!_fsExtra.default.existsSync(appFile)) {
      _sharedUtils.ColorConsole.throw(`App file does not exist`);
    }
    const {
      pages = {},
      widgets = {},
      floatingWindows = {}
    } = router;
    const confsList = [{
      confs: pages,
      type: _EntryType.default.PAGE
    }, {
      confs: floatingWindows,
      type: _EntryType.default.FLOAT
    }, {
      confs: widgets,
      type: _EntryType.default.CARD
    }];

    // 1. 添加app
    const result = {
      app: `./${_path.default.relative(projectPath, appFile)}?uxType=${_EntryType.default.APP}`
    };

    // 2. 添加 route 配置的文件
    confsList.forEach(item => {
      const {
        confs,
        type
      } = item;
      Object.keys(confs).forEach(routePath => {
        if (/^\//.test(routePath)) {
          throw new Error(`Compilation failed: please confirm that '${routePath}' configured by router.pages in manifest.json is the directory name`);
        }
        const conf = confs[routePath];
        //轻卡无需打包
        if (conf.type === _IManifest.LITE_CARD_TYPE) {
          return;
        }
        const entryKey = _path.default.join(routePath, conf.component);
        const filePath = this.resolveFile(_path.default.join(codeDir, entryKey));
        if (!filePath) {
          throw new Error(`Compilation failed: please confirm that the file path ${entryKey} configured in manifest.json exists`);
        }
        let sourceFile = `./${_path.default.relative(projectPath, filePath)}?uxType=${type}`;
        sourceFile = sourceFile.replace(/\\/g, '/');
        result[entryKey] = sourceFile;
      });
    });

    // 3. 添加 workers
    const {
      workers
    } = config;
    if (workers && workers.entries && Array.isArray(workers.entries)) {
      workers.entries.filter(worker => worker.file).forEach(worker => {
        result[worker.file.replace(/\.js$/, '')] = './src/' + worker.file;
      });
    }

    // 4. 添加 services
    const {
      services
    } = config;
    // 数组格式
    if (Array.isArray(services)) {
      services.forEach(item => {
        const {
          name,
          path
        } = item;
        if (name && path) {
          result['services/' + name] = './src/' + path + `?uxType=${_EntryType.default.APP}`;
        }
      });
    }
    // 字典格式
    else {
      for (const key in services) {
        if (Object.hasOwnProperty.call(services, key)) {
          result['services/' + key] = './src/' + services[key].path + `?uxType=${_EntryType.default.APP}`;
        }
      }
    }
    return result;
  }

  /**
   * 通过无后缀的文件名路径获取存在的文件路径
   *
   * @example
   * const path = resolveFile('/project/a')
   * // path = '/project/a.ux'
   * @param filePath
   */
  static resolveFile(filePath) {
    const extensionList = this.getExtensionList();
    for (const item of extensionList) {
      const currentPath = `${filePath}${item}`;
      if (_fsExtra.default.existsSync(currentPath)) {
        return currentPath;
      }
    }
    return '';
  }

  /**
   * 获取ux文件支持的后缀列表
   * @param withDot
   * @returns
   */
  static getExtensionList() {
    let withDot = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : true;
    const result = ['ux', 'hml'];
    if (withDot) {
      return result.map(item => `.${item}`);
    }
    return result;
  }
}
var _default = exports.default = UxCompileUtil;
