// aiot-toolkit/lib/builder/UxBuilderBase.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "IFileLaneEvents", {
  enumerable: true,
  get: function () {
    return _IFileLaneEvents.default;
  }
});
Object.defineProperty(exports, "IFileLaneSuccessData", {
  enumerable: true,
  get: function () {
    return _IFileLaneEvents.IFileLaneSuccessData;
  }
});
exports.default = void 0;
var _aiotpack = require("@aiot-toolkit/aiotpack");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _path = _interopRequireDefault(require("path"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _lodash = _interopRequireDefault(require("lodash"));
var _fileLane = _interopRequireDefault(require("file-lane"));
var _IFileLaneEvents = _interopRequireWildcard(require("file-lane/lib/interface/IFileLaneEvents"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * UxBuilderBase
 */
class UxBuilderBase {
  /**
   * 配置文件
   */
  QUICKAPP_CONFIG = 'quickapp.config.js';
  constructor() {
    let events = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    this.events = events;
  }

  /**
   * ux项目的build函数
   * @param projectPath 项目路径
   * @param options 命令参数
   * @param watch 是否开启监听
   */
  async build(projectPath, options) {
    if (this.fileLane) {
      // 已经在运行中，请先停止
      throw new Error('already running, please stop first: builder.stop()');
    }
    const watch = options.watch || false;
    // 项目配置
    const uxProjectConfig = new _aiotpack.UxConfig(projectPath);
    // 编译配置
    const compilerOption = this.getCompilerOption(projectPath, options);
    const log = {
      level: _sharedUtils.Loglevel.INFO,
      message: ['start build: ', {
        word: JSON.stringify({
          projectPath,
          options: {
            ...compilerOption,
            // 去除敏感信息
            clientRecordPath: undefined,
            dataCoverage: undefined,
            projectPath: undefined,
            webpack: undefined,
            server: undefined
          },
          watch,
          node: process.version,
          platform: process.platform,
          arch: process.arch,
          toolkit: require('../../package.json').version
        }, null, 2)
      }]
    };
    this.handleLogs([log]);
    // 开始编译项目
    this.fileLane = new _fileLane.default(uxProjectConfig, projectPath, compilerOption, this.events);
    await this.fileLane.start({
      watch
    });
  }
  handleLogs(logs) {
    const {
      onLog
    } = this.events;
    if (onLog) {
      onLog(logs);
    } else {
      logs.forEach(log => {
        _sharedUtils.ColorConsole.logger(log);
      });
    }
  }
  async dispose() {
    await this.fileLane?.dispose();
    this.fileLane = undefined;
  }

  /**
   * 获取build的编译配置
   * @param projectPath 项目路径
   * @param options 命令参数
   */
  getCompilerOption(projectPath, options) {
    // 读取项目中文件的配置
    const quickappConfig = this.readQuickAppConfig(projectPath);
    const {
      cli,
      ...otherConfig
    } = quickappConfig || {};
    options = _lodash.default.merge({}, options, cli);
    // 项目配置
    const uxProjectConfig = new _aiotpack.UxConfig(projectPath);
    // 编译配置
    const mode = options.mode || _aiotpack.CompileMode.DEVELOPMENT;
    const compilerOption = _lodash.default.merge({}, _aiotpack.JavascriptDefaultCompileOption, {
      ...options,
      projectPath: _path.default.join(projectPath, uxProjectConfig.output),
      mode
    }, otherConfig);
    return compilerOption;
  }
  readQuickAppConfig(projectPath) {
    const path = _path.default.join(projectPath, this.QUICKAPP_CONFIG);
    if (_fsExtra.default.existsSync(path)) {
      try {
        const data = require(path);
        return data;
      } catch (error) {
        const log = {
          level: _sharedUtils.Loglevel.ERROR,
          message: ['read quickapp.config.js error: ', {
            word: error?.toString() || ''
          }]
        };
        this.handleLogs([log]);
      }
    }
    return;
  }
}
var _default = exports.default = UxBuilderBase;
