// @aiot-toolkit/commander/lib/Command.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _prompts = require("@inquirer/prompts");
var _commander = require("commander");
class Command {
  /**
   * 注册命令行项目
   * @param programConfig 配置
   */
  static registeProgram(programConfig) {
    const {
      name,
      description,
      version,
      commandList,
      defaultCommand
    } = programConfig;
    _commander.program.name(name).description(description).version(version, '-v, --version');
    if (defaultCommand) {
      this.registeCommand(defaultCommand);
    }
    if (commandList) {
      commandList.forEach(item => {
        Command.registeCommand(item);
      });
    }
    _commander.program.parse();
  }
  static registeCommand(command) {
    var _this = this;
    const {
      name,
      description,
      argumentList,
      action,
      paramList,
      waiter
    } = command;
    const instance = name ? _commander.program.command(name).description(description) : _commander.program;

    // 设置argument
    argumentList?.forEach(item => {
      const {
        name,
        description,
        defaultValue
      } = item;
      instance.argument(name, description, defaultValue);
    });

    // 设置option
    paramList?.forEach(item => {
      const {
        type,
        name,
        description,
        defaultValue
      } = item;
      const effectName = type === 'confirm' ? `--${name}` : `--${name} <value>`;
      instance.option(effectName, description, defaultValue);
    });

    // 设置交互式问题
    instance.action(async function () {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }
      let option = args[args.length - 2];
      if (paramList) {
        // 检查询问式参数
        for (let item of paramList) {
          const {
            name,
            enableInquirer
          } = item;
          const cameName = _sharedUtils.StringUtil.hyphenedToCamel(name);
          const valueSource = instance.getOptionValueSource(cameName);
          const haveSet = valueSource && valueSource !== 'default';
          // 如果启用询问，且未已显式设置值，则通过询问收集参数
          if (enableInquirer && !haveSet) {
            const result = await _this.collectParam(item);
            if (result) {
              for (let key in result) {
                option[key] = result[key];
              }
            }
          }
        }

        // 检查弃用的参数
        const deprecatedList = paramList.filter(item => item.deprecated);
        deprecatedList.forEach(item => {
          const {
            name
          } = item;
          const cameName = _sharedUtils.StringUtil.hyphenedToCamel(name);
          if (option.hasOwnProperty(cameName)) {
            _sharedUtils.ColorConsole.warn({
              word: item.name
            }, 'is deprecated', ...(item.deprecatedVersion ? [', will be removed in version:', {
              word: item.deprecatedVersion
            }] : []), '.', {
              word: item.deprecated || ''
            });
          }
        });
      }
      await action.apply(null, args);
      if (waiter) {
        waiter.describe();
        waiter.start();
      }
    });
  }

  /**
   * 采集参数。  根据不同的参数类型使用不同的询问方式
   * @param param
   * @returns
   */
  static async collectParam(param) {
    const {
      name,
      description,
      type,
      defaultValue
    } = param;
    const cameName = _sharedUtils.StringUtil.hyphenedToCamel(name);
    let result;
    switch (type) {
      case 'string':
        const {
          validate
        } = param;
        result = await (0, _prompts.input)({
          message: description,
          default: defaultValue?.toString(),
          ...(validate ? {
            validate
          } : {})
        });
        break;
      case 'confirm':
        result = await (0, _prompts.confirm)({
          message: description,
          default: Boolean(defaultValue)
        });
        break;
      case 'select':
        result = await (0, _prompts.select)({
          message: description,
          choices: param.choices
        });
        break;
      case 'checkbox':
        result = await (0, _prompts.checkbox)({
          message: description,
          choices: param.choices
        });
        break;
      default:
        break;
    }
    return {
      [cameName]: result
    };
  }
}
var _default = exports.default = Command;
