// @aiot-toolkit/commander/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Command", {
  enumerable: true,
  get: function () {
    return _Command.default;
  }
});
Object.defineProperty(exports, "ICommand", {
  enumerable: true,
  get: function () {
    return _ICommand.default;
  }
});
Object.defineProperty(exports, "IParam", {
  enumerable: true,
  get: function () {
    return _IParam.default;
  }
});
Object.defineProperty(exports, "IPersistentCommandData", {
  enumerable: true,
  get: function () {
    return _IPersistentCommandData.default;
  }
});
Object.defineProperty(exports, "IProgram", {
  enumerable: true,
  get: function () {
    return _IProgram.default;
  }
});
Object.defineProperty(exports, "PersistentCommand", {
  enumerable: true,
  get: function () {
    return _PersistentCommand.default;
  }
});
exports.default = void 0;
var _Command = _interopRequireDefault(require("./Command"));
var _ICommand = _interopRequireDefault(require("./interface/ICommand"));
var _IParam = _interopRequireDefault(require("./interface/IParam"));
var _IProgram = _interopRequireDefault(require("./interface/IProgram"));
var _IPersistentCommandData = _interopRequireDefault(require("./persistentCommand/IPersistentCommandData"));
var _PersistentCommand = _interopRequireDefault(require("./persistentCommand/PersistentCommand"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _Command.default;
