// @aiot-toolkit/aiotpack/lib/compiler/javascript/vela/VelaWebpackConfigurator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _UxFileUtils = _interopRequireDefault(require("../../../utils/ux/UxFileUtils"));
var _WrapPlugin = _interopRequireDefault(require("./plugin/WrapPlugin"));
var _UxCompileUtil = _interopRequireDefault(require("./utils/UxCompileUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class VelaWebpackConfigurator {
  createPlugins() {
    // 包裹代码的插件
    const result = [this.createWrapPlugin()];

    // 如果开启 stats 参数，则添加 webpack-bundle-analyzer 插件
    if (this.param.enableStats) {
      result.push(this.createBundleAnalyzerPlugin());
    }
    return result;
  }
  createWrapPlugin() {
    return new _WrapPlugin.default(this.param);
  }
  createBundleAnalyzerPlugin() {
    const BundleAnalyzerPlugin = require('webpack-bundle-analyzer').BundleAnalyzerPlugin;
    return new BundleAnalyzerPlugin({
      analyzerMode: 'static',
      openAnalyzer: false,
      excludeAssets: /^@(system|service)\./
    });
  }

  /**
   * 通过读取 manifest.json 生成 entry
   * @returns
   */
  createEntry() {
    const {
      projectPath,
      sourceRoot
    } = this.param;
    const configFilePath = _UxFileUtils.default.getManifestFilePath(this.param.projectPath, this.param.sourceRoot);
    if (_fsExtra.default.existsSync(configFilePath)) {
      const config = _fsExtra.default.readJSONSync(configFilePath);
      return _UxCompileUtil.default.resolveEntries(config, _path.default.resolve(projectPath, sourceRoot), projectPath);
    } else {
      throw new Error(`Configuration file does not exist: ${configFilePath}`);
    }
  }
  createRules() {
    const {
      projectPath,
      sourceRoot
    } = this.param;
    const srcPath = _path.default.resolve(projectPath, sourceRoot);
    return [{
      test: /\.js$/,
      use: [{
        loader: require.resolve('babel-loader'),
        options: {
          cwd: this.param.projectPath,
          cacheDirectory: true
        }
      }]
    }, {
      test: /\.json$/,
      include: [_path.default.join(srcPath, 'manifest.json')]
    }, {
      test: /\.(png|jpe?g|gif|svg|bmp|webp|mp4|wmv|avi|mpg|rmvb|mov|flv|otf|ttf|ttc|woff|eot)$/i,
      use: {
        loader: require.resolve('url-loader'),
        options: {
          limit: 0,
          name: `dynamicAssets/[name].[hash:8].[ext]`,
          publicPath: '/',
          esModule: false
        }
      }
    }];
  }
}
var _default = exports.default = VelaWebpackConfigurator;
