// @aiot-toolkit/aiotpack/lib/loader/ux/JsLoader.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parser = require("@aiot-toolkit/parser");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
/**
 * JsLoader
 */
class JsLoader {
  async parser(files) {
    const onLog = () => {};
    const result = [];
    for (const item of files) {
      if (!item.content) {
        result.push({
          path: item.path,
          content: item.content
        });
      } else {
        const options = {
          filePath: item.path,
          projectPath: this.context.projectPath,
          content: item.content.toString(),
          projectType: _sharedUtils.ProjectType.getProjectType(this.context.projectPath),
          onLog
        };
        result.push({
          path: item.path,
          content: (await new _parser.ScriptToTypescript(options, this.compilerOption, this.context).translate({
            content: new _parser.ScriptParser(options).parser(item.content.toString()).ast.content
          }, [])).targetTree.getFullText()
        });
      }
    }
    return result;
  }
}
var _default = exports.default = JsLoader;
