// @aiot-toolkit/parser/lib/ux/utils/BabelUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var parser = _interopRequireWildcard(require("@babel/parser"));
var t = _interopRequireWildcard(require("@babel/types"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * @see：node_modules/@vue/compiler-core/dist/compiler-core.cjs.prod.js的 walkIdentifiers函数
 * BabelUtil
 */
class BabelUtil {
  /**
   * 是否是对象的标识符
   *
   * 例如：  a = {}
   * @param node
   * @returns
   */
  static isObjectIdentifier(node) {
    return node && (node.type === 'ObjectProperty' || node.type === 'ObjectMethod') && !node.computed;
  }

  /**
   * 是否是对象键的标识符
   *
   * 例如： {x: **}  中的 x
   * @param node
   * @param parent
   * @returns
   */
  static isObjectKeyIdentifier(node, parent) {
    return this.isObjectIdentifier(parent) && parent.key === node;
  }

  /**
   * 生成 babel的函数体
   * @param code
   * @returns
   */
  static generateFunctionBody(code) {
    const func = `function() {${code}}`;
    return parser.parseExpression(func).body;
  }

  /**
   * 插入或更新函数到对象
   *
   * @param functionName 函数名. 如果存在此函数，则把body插入到最后；否则，添加函数
   * @param body 新增的函数体
   * @param path 对象路径(babel的路径)
   */
  static insertOrUpdateFunctionToObject(functionName, body, path) {
    if (path.node.declaration.type === 'ObjectExpression') {
      const properties = path.node.declaration.properties;
      const fun = properties.find(p => p.type === 'ObjectMethod' && p.key.name === functionName);
      if (fun) {
        fun.body.body.push(...body.body);
      } else {
        properties.push(t.objectMethod('method', {
          name: functionName,
          type: 'Identifier'
        }, [], body));
      }
    }
  }

  /**
   * 插入import, 如果 `importName` 存在，则不执行操作
   *
   * @example
   * import router from '@system/router'
   *
   * @param importName import的名称, 例如：router
   * @param importPath import的路径, 例如：@system/router
   * @param param babel节点， 只有 `Program` 类型时才执行
   */
  static insertImportToProgram(importName, importPath, param) {
    if (param.type === 'Program') {
      // 检查是否存在

      const hasImport = param.body.some(node => node.type === 'ImportDeclaration' && node.specifiers.some(specifier => specifier.local.name === importName));
      if (hasImport) {
        return;
      }
      const importDeclaration = t.importDeclaration([t.importDefaultSpecifier(t.identifier(importName))], t.stringLiteral(importPath));
      param.body.unshift(importDeclaration);
    }
  }
}
var _default = exports.default = BabelUtil;
