// @aiot-toolkit/parser/lib/ux/utils/ElementConfigUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _TemplateUtil = _interopRequireDefault(require("../translate/vela/utils/TemplateUtil"));
var _QuickAppDocHelp = require("@aiot-toolkit/shared-utils/lib/utils/QuickAppDocHelp");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * ElementConfigUtil
 */
class ElementConfigUtil {
  constructor(config) {
    this.config = config;
  }
  getElementConfig = name => {
    const element = this.config.elementConfig[name];
    return element;
  };
  allowTextChildren = nodeName => {
    const config = this.config.elementConfig[nodeName];
    if (config && config.allowTextChildren) {
      return config.allowTextChildren;
    }
    return false;
  };

  /**
   *
   * 对非自定义组件检查元素合法性
   * 1. 是否有配置
   * 2. 检查属性值是否合法
   * 3. 检查是否缺少必填属性
   * @param tagName 元素名称  <com id="a"> 中的 com
   * @param importList
   * @returns
   */
  validateElement = (options, tag, attributes, importList) => {
    const tagName = tag.name;
    const tagLocation = tag.location;
    const {
      onLog,
      filePath,
      docHelp
    } = options;
    // 1
    if (!importList?.includes(tagName)) {
      const config = this.getElementConfig(tagName);
      if (!config) {
        onLog({
          level: _sharedUtils.Loglevel.WARN,
          message: [{
            word: `[${tagName}]`
          }, `Unknown element`],
          position: tagLocation,
          filePath
        });
      }

      // 2
      if (attributes.length) {
        attributes.forEach(item => {
          this.validateAttribute(options, tagName, item);
        });
      }

      // 3
      if (config && config.attributes) {
        const requireList = Object.keys(config.attributes).filter(attributeName => {
          return config.attributes && config.attributes[attributeName]?.required;
        });
        const missingAttribute = requireList.filter(attributeName => attributes.findIndex(item => item.name === attributeName) < 0);
        if (missingAttribute.length) {
          onLog({
            filePath,
            level: _sharedUtils.Loglevel.WARN,
            position: tagLocation,
            message: [{
              word: `[${tagName}]`
            }, `missing attributes`, {
              word: missingAttribute.join('、')
            }, docHelp?.createWidgetReferenceMessage({
              componentName: tagName,
              chapter: _QuickAppDocHelp.QuickAppDocWidgetType.ATTRIBUTE
            })]
          });
        }
      }
    }
    return true;
  };

  /**
   * 检查属性
   * 1. 是否忽略： data-* 忽略
   * 2. 如果是事件
   *    a. 检查是否在 元素 events 或 commonEvents 中
   *    b. 检查是否有值
   * 3. 不是事件
   *    a. 检查是否在attributes配置中
   *    b. 是否有 enum，如果有，则检查值是否在配置中(不检查动态值)
   *    c. 是否有 validate，如果有，则执行并返回结果
   * @param tagName 元素标签名
   * @param attributeName 属性名
   * @param attributeValue 属性值
   */
  validateAttribute = (options, tagName, attribute) => {
    const elementConfig = this.getElementConfig(tagName);
    const {
      onLog,
      filePath,
      docHelp
    } = options;
    const {
      name,
      value,
      nameLocation
    } = attribute;
    // 1
    if (_TemplateUtil.default.isDataSetAttribute(name)) {
      return true;
    }
    // 2
    const isEvent = _TemplateUtil.default.isEventAttribute(name);
    if (isEvent) {
      const eventName = _TemplateUtil.default.translateEventName(name);
      const eventList = [...(elementConfig?.events || []), ...this.config.commonEvents];
      if (!eventList.includes(eventName)) {
        onLog({
          filePath,
          position: nameLocation,
          level: _sharedUtils.Loglevel.WARN,
          message: [{
            word: `[${name}]`
          }, `unsupport event`, docHelp?.createWidgetReferenceMessage({
            componentName: tagName,
            chapter: _QuickAppDocHelp.QuickAppDocWidgetType.EVENT
          })]
        });
        return false;
      }
      if (!value) {
        onLog({
          filePath,
          position: nameLocation,
          level: _sharedUtils.Loglevel.WARN,
          message: [{
            word: `[${name}]`
          }, `Event Value Cannot Be Empty`]
        });
        return false;
      }
    } else {
      if (!elementConfig) {
        return true;
      }
      const attributes = {
        ...elementConfig.attributes,
        ...this.config.commonAttributes
      };
      if (!attributes) {
        return true;
      }

      // 3.a
      const attributeConfig = attributes[name];
      if (!attributeConfig) {
        onLog({
          filePath,
          position: nameLocation,
          level: _sharedUtils.Loglevel.WARN,
          message: [`unsupport attribute`, {
            word: `[${tagName}.${name}]`
          }, docHelp?.createWidgetReferenceMessage({
            componentName: tagName
          })]
        });
        return true;
      }

      // 3.b
      const {
        enums,
        validate
      } = attributeConfig;
      if (enums && !_TemplateUtil.default.isDynamicValue(value) && !enums.includes(value)) {
        onLog({
          filePath,
          position: nameLocation,
          level: _sharedUtils.Loglevel.WARN,
          message: [{
            word: `[${tagName}-->${name}]`
          }, `value is`, {
            word: value
          }, `, but the valid value is`, {
            word: enums.join('/')
          }, docHelp?.createWidgetReferenceMessage({
            componentName: tagName,
            chapter: _QuickAppDocHelp.QuickAppDocWidgetType.ATTRIBUTE
          })]
        });
        return false;
      }

      // 3.c
      if (validate) {
        const errors = validate(value, onLog);
        if (errors.length) {
          return false;
        }
      }
    }
    return true;
  };
}
var _default = exports.default = ElementConfigUtil;
