// @aiot-toolkit/parser/lib/ux/utils/StyleMapUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _url = require("url");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class StyleMapUtil {
  constructor(sourceMapConsumer, lineOffset, options) {
    this.sourceMapConsumer = sourceMapConsumer;
    this.lineOffset = lineOffset;
    this.options = options;
  }
  /**
   * 1. 根据生成结果位置找转换前的行列值
   * 2. 根据样式代码总体偏移量，计算出当前节点行在ux中的行
   *  2.1 查看map信息，原始行总是从第三行开始计算，-2是去掉生成map信息时多余的偏移量
   *  2.2 再加上样式代码总体偏移量，就可以计算出实际所在行数
   * 3. 返回节点原始位置信息
   *  3.1 原始位置的startLine即为上一步计算了偏移量的行信息
   *  3.2 原始位置的startColumn为查找到的节点列信息（目前看结果不存在列偏移）
   *  3.3 原始位置的endLine为startLine+生成结果中所占据行数的值
   *  3.4 原始位置的endColumn根据map结果，发现存在1个位置的差异，所以-1
   *  3.5 source存储来源的文件
   * @param location 生成结果中节点的位置信息
   * @param consumer 经过解析器获取的map信息生成的consumer对象
   * @param offset 当前样式代码在ux文件中的偏移量
   * @returns
   */
  transfromLocToPosition(location) {
    const {
      projectPath
    } = this.options;
    const nodeStartPosition = this.sourceMapConsumer.originalPositionFor({
      line: location.start.line,
      column: location.start.column
    });
    const nodeEndPosition = this.sourceMapConsumer.originalPositionFor({
      line: location.end.line,
      column: location.end.column
    });
    let nodeRelativePath = '';
    if (nodeStartPosition.source?.endsWith('.ux') && nodeStartPosition.line) {
      nodeStartPosition.line = nodeStartPosition.line - 1 + this.lineOffset;
    }
    if (nodeEndPosition.source?.endsWith('.ux') && nodeEndPosition.line) {
      nodeEndPosition.line = nodeEndPosition.line - 1 + this.lineOffset;
    }
    // nodeStartPosition.source在scss转css时会添加file://前缀，在这里需要去掉，以免影响路径的计算
    if (nodeStartPosition.source) {
      let decodeStr = decodeURI(nodeStartPosition.source);
      if (decodeStr.startsWith('file://')) {
        decodeStr = (0, _url.fileURLToPath)(decodeStr);
      }
      nodeRelativePath = _path.default.relative(projectPath, decodeStr).replaceAll(_path.default.sep, _path.default.posix.sep);
    }
    return {
      pos: 0,
      end: 0,
      startLine: nodeStartPosition.line || 0,
      startColumn: nodeStartPosition.column || 0,
      endLine: nodeEndPosition.line || 0,
      endColumn: nodeEndPosition.column || 0,
      source: nodeRelativePath
    };
  }
}
var _default = exports.default = StyleMapUtil;
