// @aiot-toolkit/parser/lib/ux/utils/UxUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * UxParserUtil
 */
class UxUtil {
  /**
   * 重组文件名
   * @param fileName
   * @param extValue
   * @returns
   */
  static spliceFileName(fileName, extValue) {
    const {
      name
    } = _path.default.parse(fileName);
    return name && extValue ? `${name}.${extValue}` : '';
  }
  /**
   * 获取css的lang类型
   * @param node
   * @returns
   */
  static getLangType(node) {
    let langType = 'css';
    for (let attr of node?.attrs || []) {
      if (attr.name === 'lang') {
        langType = attr.value;
        break;
      }
    }
    return langType;
  }
  /**
   * 更新文件路径中的后缀
   * @param path
   * @param extValue
   * @returns
   */
  static updateFileName(path, extValue) {
    const dirName = _path.default.dirname(path);
    let newFileName = UxUtil.spliceFileName(path, extValue);
    return newFileName ? _path.default.posix.join(dirName, newFileName) : path;
  }
  /**
   * 将全局变量根据样式语言类型转为对应的变量声明
   * {
   *  "a": "#fff"
   * }
   * 转为
   * $a: #fff 或者 @a: #fff
   * @param globalVar
   * @param langType
   * @returns
   */
  static convertGlobalJson(globalVar) {
    let langType = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'css';
    if (langType === 'css' && globalVar) {
      return '';
    }
    const langSymbol = langType === 'less' ? '@' : '$';
    const varArray = [];
    for (const key in globalVar) {
      const value = globalVar[key];
      varArray.push(`${langSymbol}${key}: ${value};`);
    }
    return varArray.join('\n');
  }

  /**
   * 是否是绝对资源路径
   *
   * 1. 以/开头：/a/b/c.png
   * 2. 以 ***://开头的视为绝对路径: package://a.png
   * @param path
   */
  static isAbsoluteResource(path) {
    if (!path) {
      return true;
    }
    if (path.startsWith('/')) {
      return true;
    }
    const reg = /^\w+:\/\//;
    return reg.test(path);
  }
  static importToTypescript(importList, compilerOption) {
    let appImport = [];
    for (let item of importList) {
      const {
        name,
        src
      } = item;
      const pushList = [`$app_exports$['${name.toLowerCase()}'] = require('${src}')`];

      // 如果开启统计，添加引用组件耗时日志
      if (compilerOption.enableStats) {
        pushList.unshift(`console.time('PERF: require @app-component/${name}')`);
        pushList.push(`console.timeEnd('PERF: require @app-component/${name}')`);
      }
      appImport.push(...pushList);
    }
    return appImport;
  }
}
var _default = exports.default = UxUtil;
