// aiot-toolkit/lib/utils/RequestUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _axios = _interopRequireDefault(require("axios"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * RequestUtils
 */
class RequestUtils {
  static async sendReq(client, api, params) {
    return new Promise(resolve => {
      const requrl = `http://${client.ip}:${client.port}${api}`;
      let options = {
        host: client.ip,
        port: client.port,
        path: api,
        timeout: 3000
      };
      if (params) {
        options = Object.assign({}, options, {
          headers: params
        });
      }
      _axios.default.post(requrl, options).then(data => {
        _sharedUtils.ColorConsole.log(`### App Server ### Request ${requrl} succeeded`);
        resolve(data.toString());
      }).catch(error => {
        if (error.respose && error.response.status === 408) {
          // 超时处理
          _sharedUtils.ColorConsole.error(`### App Server ### Request ${requrl} timed out, please try again`);
        } else {
          _sharedUtils.ColorConsole.error(`### App Server ### Request ${requrl} error message: ${error?.toString() || 'unknown error'}`);
        }
      });
    });
  }
  static downloadFile(url, fileName) {
    return new Promise(async (resolve, reject) => {
      // 开始下载
      _sharedUtils.ColorConsole.log(`Start downloading file：${fileName}, address：${url}`);
      const response = await _axios.default.get(url, {
        responseType: 'arraybuffer'
      });
      if (response.status === 200) {
        const targetDir = _path.default.join(__dirname, './apk');
        _fsExtra.default.ensureDirSync(targetDir);
        _fsExtra.default.writeFileSync(_path.default.join(targetDir, fileName), response.data);
        resolve(`Download file ${fileName} successfully`);
      } else {
        reject(`Download failed, status code：${response.status}`);
      }
    });
  }
}
var _default = exports.default = RequestUtils;
