// aiot-toolkit/lib/utils/VelaAvdUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _emulator = require("@aiot-toolkit/emulator");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _prompts = require("@inquirer/prompts");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class VelaVvdUtils {
  static sdkHome = (() => _emulator.defaultSDKHome)();
  static vvdHome = (() => _emulator.defaultVvdHome)();
  static vvdManager = (() => new _emulator.VvdManager({
    sdkHome: VelaVvdUtils.sdkHome,
    vvdHome: VelaVvdUtils.vvdHome
  }))();

  /** 校验VVD名称 */
  static validateVvdName(vvdName) {
    if (!vvdName) {
      return `Please enter vvd name`;
    }
    if (!vvdName.startsWith('Vela')) {
      return 'The vvd name must starts with Vela';
    }
    const vvdDir = this.vvdManager.getVvdDir(vvdName);
    if (_fsExtra.default.existsSync(vvdDir)) {
      return `The vvd already exists. Please change vvd name`;
    }
    return true;
  }

  /**
   * 命令行访问方式创建模拟器
   */
  static async createVelaVvdByInquire() {
    try {
      // vvdName
      const name = await (0, _prompts.input)({
        message: 'vvd name starting with Vela. (eg. Vela_Virtual_Device)',
        default: 'Vela_Virtual_Device',
        validate(value) {
          return VelaVvdUtils.validateVvdName(value);
        }
      });
      // 镜像
      const velaImage = await (0, _prompts.select)({
        message: 'vela image.',
        choices: _emulator.VelaImageVersionList.map(item => {
          return {
            value: item.value,
            name: item.label
          };
        })
      });
      // skin或者size
      let skin;
      let width = '466';
      let height = '466';
      // vela4.0不允许自定义分辨率
      if ((0, _emulator.isMiwearImageType)(velaImage)) {
        const needSkin = await (0, _prompts.confirm)({
          message: 'need vvd skin?',
          default: true
        });
        if (needSkin) {
          const skinList = await VelaVvdUtils.vvdManager.getVelaSkinList();
          skin = await (0, _prompts.select)({
            message: 'vvd skin.',
            choices: skinList.map(item => {
              return {
                name: item.name,
                value: item,
                description: item.name
              };
            })
          });
        } else {
          width = await (0, _prompts.input)({
            message: 'vvd width.',
            default: '466'
          });
          height = await (0, _prompts.input)({
            message: 'vvd height.',
            default: '466'
          });
        }
      }
      const vvdManager = new _emulator.VvdManager({
        sdkHome: VelaVvdUtils.sdkHome
      });
      if ((await vvdManager.hasSDKPartUpdate()).length > 0) {
        const downloder = await vvdManager.downloadSDK({
          imageTypeArr: [velaImage],
          force: false,
          cliProgress: true,
          parallelDownloads: 6
        });
        await downloder.downlodPromise;
      }
      const imageDir = _path.default.resolve(VelaVvdUtils.sdkHome, _emulator.SDKParts.SYSTEM_IMAGES, velaImage);

      // 创建vvd文本文件
      const vvdParams = {
        name,
        skin: skin?.name,
        width,
        height,
        arch: _emulator.IVvdArchType.arm,
        imageDir,
        imageType: velaImage,
        'skin.path': skin?.path,
        customLcdRadius: ''
      };
      VelaVvdUtils.vvdManager.createVvd(vvdParams);
      _sharedUtils.ColorConsole.success(`Create vvd succeed.`);
    } catch (e) {
      _sharedUtils.ColorConsole.throw(`Create vvd failed. Error: ${e.message}`);
    }
  }
  static async checkEmulatorEnv() {
    return (await this.vvdManager.hasSDKPartUpdate()).length === 0;
  }

  /** 初始化/重置模拟器环境 */
  static async initEmulatorEnv() {
    const downloder = await VelaVvdUtils.vvdManager.downloadSDK({
      force: true,
      cliProgress: true,
      parallelDownloads: 6
    });
    await downloder.downlodPromise;
  }
}
var _default = exports.default = VelaVvdUtils;
