// @aiot-toolkit/parser/lib/utils/ParserUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _tsMorph = require("ts-morph");
/**
 * ParserUtil
 */
class ParserUtil {
  /**
   * 根据偏移列表，查找 pos 对应的偏移量
   * @param pos
   * @param offsetList
   * @returns
   */
  static getOffsetPos(pos, offsetList) {
    if (!offsetList.length) {
      return 0;
    }
    // 找到"最后一个 < pos"的项，返回此项的值
    const item = offsetList.concat().reverse().find(item => {
      return item.pos < pos;
    });
    return item ? item.offset : 0;
  }

  /**
   * 获取指定位置的列偏移量
   *
   * 偏移量= 从offsetList 中获取 (在同一行  且 位置小于 position ) 的项，偏移量进行累加。
   *
   * 确保同一行有多个偏移项也可获取正确的偏移量
   * @param position
   * @param offsetList
   */
  static getOffsetColumn(position, offsetList) {
    if (!offsetList?.length) {
      return 0;
    }
    const {
      line,
      pos
    } = position;
    // 从offsetList 中获取 (在同一行  且 位置小于 position ) 的项，偏移量进行累加
    const effectOffsetList = offsetList.filter(item => item.line === line && item.pos < pos);
    const result = effectOffsetList.length ? effectOffsetList.reduce((preValue, currentValue) => {
      return currentValue.currentOffset + preValue;
    }, 0) : 0;
    return result;
  }
  static createSourceFile(source, fileName) {
    if (!this.project) {
      this.project = this.createProject();
    }
    const name = fileName || 'temp.ts';
    const oldFile = this.project.getSourceFile(name);
    if (oldFile) {
      this.project.removeSourceFile(oldFile);
    }
    return this.project.createSourceFile(name, source);
  }
  static createProject() {
    return new _tsMorph.Project({
      useInMemoryFileSystem: true,
      manipulationSettings: {
        indentationText: _tsMorph.IndentationText.TwoSpaces
      }
    });
  }
}
var _default = exports.default = ParserUtil;
