// @aiot-toolkit/emulator/lib/shared/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RpkStartFailedReason = exports.RpkInstallFailedReason = void 0;
exports.getDefaultImage = getDefaultImage;
exports.isMiwearImageType = isMiwearImageType;
exports.isVelaImageType = isVelaImageType;
var _Vvd = require("../typing/Vvd");
/**
 * 本文件用于导出一些同构内容，请勿引用非同构内容，例如 nodejs 特性内容，或者 browser 特性内容
 * @TODO 项目转为 ESM 模块后，可以删除
 * */

function isVelaImageType(value) {
  return Object.values(_Vvd.VelaImageType).includes(value);
}
function isMiwearImageType(val) {
  return [_Vvd.VelaImageType.REL, _Vvd.VelaImageType.VELA_MIWEAR_WATCH_5, _Vvd.VelaImageType.VELA_MIWEAR_MINISOUND_5].includes(val);
}
function getDefaultImage() {
  return _Vvd.VelaImageType.VELA_MIWEAR_WATCH_5;
}
let RpkStartFailedReason = exports.RpkStartFailedReason = /*#__PURE__*/function (RpkStartFailedReason) {
  RpkStartFailedReason["getManifestFailed"] = "Get app manifest.json failed";
  return RpkStartFailedReason;
}({});
let RpkInstallFailedReason = exports.RpkInstallFailedReason = /*#__PURE__*/function (RpkInstallFailedReason) {
  RpkInstallFailedReason["Busy"] = "Installation process is busy";
  RpkInstallFailedReason["Timeout"] = "Install timeout";
  RpkInstallFailedReason["Poweroff"] = "Device poweroff";
  RpkInstallFailedReason["LowStorage"] = "Application exceeds maximum quantity or remaining space is low";
  return RpkInstallFailedReason;
}({});
