// @aiot-toolkit/aiotpack/lib/compiler/tools/icu/ImageIcu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _path = _interopRequireDefault(require("path"));
var _os = _interopRequireDefault(require("os"));
var _fs = _interopRequireDefault(require("fs"));
var _child_process = require("child_process");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class ImageIcu {
  static EXT = '.bin';
  static INCLUDE_LIST = ['.png', '.jpg', '.jpeg', '.webp', '.bmp'];
  /**
   * @private
   * @returns
   */
  getBinPath() {
    const platform = _os.default.platform();
    const arch = _os.default.arch();
    let result = _path.default.join(__dirname, `icu_${platform}_${arch}`);
    if (platform === 'win32') {
      result += '.exe';
    }
    return result;
  }

  /**
   * 文件转换为 bin 格式
   * @param {*} resourcePath 源文件
   * @param {*} folder 产物目录
   * @param {'lvgl'|'png'} format
   * @returns
   */
  translate(resourcePath, folder) {
    let format = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 'lvgl';
    if (!_fs.default.existsSync(resourcePath) || !ImageIcu.canTranslate(resourcePath)) {
      return;
    }
    const binPath = this.getBinPath();
    if (_fs.default.existsSync(binPath)) {
      const pathParse = _path.default.parse(resourcePath);

      // 后缀，lvgl格式的后缀为"bin", 其它保持原后缀
      const extDic = {
        lvgl: ImageIcu.EXT,
        png: pathParse.ext
      };
      const ext = extDic[format];
      if (!_fs.default.existsSync(folder)) {
        _fs.default.mkdirSync(folder, {
          recursive: true
        });
      }
      const command = `"${binPath}" convert "${resourcePath}" -O "${folder}" -F ${format} -C i8 -r`;
      const message = (0, _child_process.execSync)(command).toString('utf-8');
      if (message) {
        console.log(`convert error: path=${resourcePath} message=${message}`);
        return '';
      }
      const result = _path.default.join(pathParse.dir, `${pathParse.name}${ext}`);
      return result;
    } else {
      return resourcePath;
    }
  }

  /**
   * 资源转换为 bin 文件路径
   *
   * 如果是 png, jpg，jpeg, webp, bmp，则转换为 .bin 后缀；其它类型不处理
   *
   * **此方法仅返回转换后的路径， 不会实际转换文件**
   * @param {string} path
   * @example assetsToBinPath('images/test.png') // images/test.bin
   *
   * @return {string} 可转则返回对应的 bin 路径；否则返回''
   */
  static assetsToBinPath(assetsPath) {
    if (!assetsPath) {
      return '';
    }
    const pathParse = _path.default.parse(assetsPath);
    if (ImageIcu.canTranslate(assetsPath)) {
      return _path.default.join(pathParse.dir, `${pathParse.name}${ImageIcu.EXT}`);
    }
    return '';
  }

  /**
   * 是否可转换
   * @param {string} assetsPath
   * @param {'lvgl'|'png'} format
   * @returns
   */
  static canTranslate(assetsPath) {
    let format = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 'lvgl';
    if (!assetsPath) {
      return false;
    }
    const pathParse = _path.default.parse(assetsPath);
    const ext = pathParse.ext.toLowerCase();
    if (format === 'png') {
      return ext === '.png';
    }
    return ImageIcu.INCLUDE_LIST.includes(ext);
  }
}
var _default = exports.default = ImageIcu;
