// @aiot-toolkit/parser/lib/ux/config/android/ElementConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const elementConfig = {
  commonEvents: ['click', 'focus', 'blur', 'key', 'longpress', 'appear', 'disappear', 'swipe', 'touchstart', 'touchmove', 'touchend', 'touchcancel', 'resize', 'animationstart', 'animationiteration', 'animationend'],
  elementConfig: {
    img: {
      aliasName: 'image',
      events: ['complete', 'error']
    },
    div: {},
    a: {
      allowTextChildren: true
    },
    text: {
      allowTextChildren: true
    },
    span: {
      allowTextChildren: true
    },
    image: {
      events: ['complete', 'error']
    },
    maml: {
      events: ['complete', 'error']
    },
    slider: {
      events: ['change']
    },
    web: {
      events: ['pagestart', 'pagefinish', 'titlereceive', 'error', 'message', 'progress']
    },
    list: {
      events: ['scroll', 'scrollbottom', 'scrolltop', 'scrollend', 'scrolltouchup', 'selected']
    },
    'list-item': {},
    block: {},
    component: {
      attributes: {
        is: {
          forceDynamic: true
        }
      }
    },
    slot: {},
    input: {
      events: ['change', 'enterkeyclick', 'selectionchange'],
      attributes: {
        type: {
          enums: ['text', 'button', 'checkbox', 'radio', 'email', 'date', 'time', 'number', 'password', 'tel', 'eventbutton']
        },
        autocomplete: {
          enums: ['on', 'off']
        },
        enterkeytype: {
          enums: ['default', 'next', 'go', 'done', 'send', 'search']
        },
        eventtype: {
          enums: ['shortcut']
        }
      }
    },
    refresh: {
      events: ['refresh']
    },
    refresh2: {
      events: ['pulldownrefresh', 'pulluprefresh', 'refresh']
    },
    'refresh-header': {
      events: ['move']
    },
    'refresh-footer': {
      events: ['move']
    },
    ad: {
      events: ['load', 'error']
    },
    swiper: {
      events: ['change', 'swipestart', 'swipeend']
    },
    progress: {},
    picker: {
      events: ['change', 'columnchange', 'cancel'],
      attributes: {
        type: {
          required: true,
          enums: ['text', 'date', 'time', 'multi-text']
        }
      }
    },
    switch: {
      events: ['change']
    },
    textarea: {
      events: ['change', 'selectionchange', 'linechange']
    },
    video: {
      events: ['playing', 'prepared', 'start', 'pause', 'finish', 'error', 'seeking', 'seeked', 'timeupdate', 'fullscreenchange']
    },
    camera: {
      events: ['error', 'camerainitdone', 'cameraframe']
    },
    map: {
      events: ['loaded', 'regionchange', 'tap', 'markertap', 'callouttap', 'controltap', 'poitap']
    },
    custommarker: {},
    canvas: {},
    stack: {
      events: ['fullscreenchange']
    },
    richtext: {
      allowTextChildren: true,
      events: ['start', 'complete']
    },
    tabs: {
      events: ['change']
    },
    'tab-content': {},
    'tab-bar': {},
    popup: {
      events: ['visibilitychange']
    },
    rating: {
      events: ['change']
    },
    select: {
      events: ['change']
    },
    option: {
      allowTextChildren: 'content'
    },
    marquee: {
      allowTextChildren: true,
      events: ['bounce', 'finish', 'start']
    },
    scrollview: {
      events: ['scroll', 'scrollbegindrag', 'scrollenddrag', 'scrollstart', 'scrollstop']
    },
    drawer: {
      events: ['change', 'scroll']
    },
    lottie: {
      events: ['complete', 'error', 'change']
    },
    'drawer-navigation': {},
    'slide-view': {
      events: ['open', 'close', 'slide', 'buttonclick']
    },
    'section-list': {
      events: ['scroll', 'scrollend', 'scrolltouchup', 'scrolltop', 'scrollbottom']
    },
    'section-group': {
      events: ['change']
    },
    'section-header': {},
    'section-item': {},
    'share-button': {
      events: ['success', 'fail', 'cancel']
    },
    'shortcut-button': {},
    qrcode: {},
    scroll: {
      events: ['scrolltop', 'scrollbottom', 'scroll']
    },
    'frame-image': {
      events: ['error']
    },
    'number-image': {
      events: ['error']
    },
    'chars-image': {
      events: ['error']
    },
    time: {},
    svg: {},
    rect: {},
    circle: {},
    ellipse: {},
    line: {},
    polyline: {},
    polygon: {},
    path: {},
    barcode: {},
    chart: {}
  }
};
var _default = exports.default = elementConfig;
