// file-lane/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FileLane", {
  enumerable: true,
  get: function () {
    return _FileLane.default;
  }
});
Object.defineProperty(exports, "FileLaneUtil", {
  enumerable: true,
  get: function () {
    return _FileLaneUtil.default;
  }
});
Object.defineProperty(exports, "FollowWork", {
  enumerable: true,
  get: function () {
    return _IFileLaneConfig.FollowWork;
  }
});
Object.defineProperty(exports, "IFileLaneConfig", {
  enumerable: true,
  get: function () {
    return _IFileLaneConfig.default;
  }
});
Object.defineProperty(exports, "IFileLaneContext", {
  enumerable: true,
  get: function () {
    return _IFileLaneContext.default;
  }
});
Object.defineProperty(exports, "IFileParam", {
  enumerable: true,
  get: function () {
    return _IFileParam.default;
  }
});
Object.defineProperty(exports, "ILoader", {
  enumerable: true,
  get: function () {
    return _ILoader.default;
  }
});
Object.defineProperty(exports, "IPlugin", {
  enumerable: true,
  get: function () {
    return _IPlugin.default;
  }
});
Object.defineProperty(exports, "PreWork", {
  enumerable: true,
  get: function () {
    return _IFileLaneConfig.PreWork;
  }
});
exports.default = void 0;
var _FileLane = _interopRequireDefault(require("./FileLane"));
var _IFileLaneConfig = _interopRequireWildcard(require("./interface/IFileLaneConfig"));
var _IFileLaneContext = _interopRequireDefault(require("./interface/IFileLaneContext"));
var _IFileParam = _interopRequireDefault(require("./interface/IFileParam"));
var _ILoader = _interopRequireDefault(require("./interface/ILoader"));
var _IPlugin = _interopRequireDefault(require("./interface/IPlugin"));
var _FileLaneUtil = _interopRequireDefault(require("./utils/FileLaneUtil"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
var _default = exports.default = _FileLane.default;
