// @aiot-toolkit/shared-utils/lib/utils/CommonUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _crypto = _interopRequireDefault(require("crypto"));
var _ajv = _interopRequireDefault(require("ajv"));
var _fs = _interopRequireDefault(require("fs"));
var _path = _interopRequireDefault(require("path"));
var _qrcode = _interopRequireDefault(require("qrcode"));
var _ColorConsole = _interopRequireDefault(require("../ColorConsole"));
var _ILog = require("../interface/ILog");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * 不好分类的工具函数
 */
class CommonUtil {
  /**
   * 根据 buffer 生成 hash 值
   *
   * 用于唯一标识流内容，通常用于文件校验
   * @param buffer
   * @returns hash 值
   */
  static calcDataDigest(buffer) {
    const hash = _crypto.default.createHash('SHA256');
    hash.update(buffer);
    return hash.digest();
  }
  static calcImageDigest(filePath) {
    const data = _fs.default.readFileSync(filePath);
    const hash = _crypto.default.createHash('md5').update(data);
    return hash.digest('hex').substring(0, 8);
  }

  /**
   * 从开发者项目的 node_module 中加载指定依赖；
   * 如果不存在，提示下载；存在则返回加载的依赖
   * @param packageName
   */
  static requireNodeModule(projectPath, packageName) {
    const modulePath = _path.default.resolve(projectPath, `./node_modules/${packageName}`);
    if (!_fs.default.existsSync(modulePath)) {
      return Promise.reject(new Error(_ColorConsole.default.createMessage([{
        word: packageName,
        style: _ColorConsole.default.getStyle(_ILog.Loglevel.ERROR)
      }, 'is missing, run', {
        word: `npm i ${packageName}`,
        style: _ColorConsole.default.getStyle(_ILog.Loglevel.ERROR)
      }, 'to install it'])));
    }
    return Promise.resolve(require(modulePath));
  }
  static requireModule(path) {
    let throwError = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
    if (!_fs.default.existsSync(path)) {
      if (throwError) {
        throw new Error(`${path} is missing`);
      }
      return;
    }
    try {
      return require(path);
    } catch (error) {
      if (throwError) {
        throw error;
      }
      return;
    }
  }

  /**
   * 验证 json 内容
   * @param json json 内容
   * @param schema 验证配置, 详见：https://github.com/ajv-validator/ajv
   *
   *
   * @example
   * ```
   * CommonUtil.validateJson({a:123},
   * {
   *  a:{
   *     type:'number'
   *   }
   * }
   * )
   * ```
   * @returns
   */
  static validateJson(json, schema) {
    if (!json || !schema) {
      return;
    }
    const ajv = new _ajv.default({
      allErrors: true,
      allowUnionTypes: true
    });
    ajv.validate(schema, json);
    if (ajv.errors) {
      const errors = ajv.errors.map(item => {
        const {
          message,
          instancePath,
          keyword
        } = item;
        const errorMessage = `${instancePath ? `${instancePath}: ` : ''}${message}`;
        if (keyword === 'required') {
          return new TypeError(errorMessage);
        }
        return new Error(errorMessage);
      });
      return errors;
    }
  }
  static outputQRCodeOnTerminal(text) {
    return _qrcode.default.toString(text, {
      type: 'utf8',
      margin: 0
    }).then(qrcode => {
      console.log(qrcode);
    });
  }
}
var _default = exports.default = CommonUtil;
