// @aiot-toolkit/shared-utils/lib/utils/FileUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _rimraf = require("rimraf");
var _fswin = _interopRequireDefault(require("fswin"));
var _minimatch = require("minimatch");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * FileUtil
 */
class FileUtil {
  /**
   * 判断字符串是否符合包含、不包含的条件
   *
   * 规则为：
   * 1. 如果设置了不包含且匹配，则不包含
   * 2. 否则，如果设置了包含且不匹配，则不包含
   * 3. 否则，包含
   * @param value
   * @param include
   * @param exclude
   */
  static include(path, include, exclude) {
    if (exclude && this.match(path, exclude)) {
      return false;
    }
    if (include && !this.match(path, include)) {
      return false;
    }
    return true;
  }
  /**
   * 判断字符串是否匹配指定规则
   * @param value
   * @param matchType
   * @returns
   */
  static match(value, matchType) {
    switch (typeof matchType) {
      case 'function':
        return matchType(value);
      case 'string':
        return (0, _minimatch.minimatch)(value, matchType, {
          dot: true
        });
      case 'object':
        if (matchType instanceof RegExp) {
          return matchType.test(value);
        } else if (matchType instanceof Array) {
          // 递归检查每一项
          let result = matchType.find(item => this.match(value, item));
          if (result) {
            return true;
          }
        }
        return false;
      default:
        return false;
    }
  }
  /**
   * 获取指定目录所有层级的文件
   * @param dir
   * @returns
   */

  static readAlldirSync(dir, includes, excludes) {
    const files = _fsExtra.default.readdirSync(dir, {
      withFileTypes: true
    });
    const result = [];
    files.forEach(item => {
      const itemPath = _path.default.join(dir, item.name);
      if (item.isDirectory()) {
        // 当目录不被排除时，递归获取子目录
        if (this.include(itemPath, undefined, excludes)) {
          const children = this.readAlldirSync(itemPath, includes, excludes);
          result.push(...children);
        }
      } else {
        if (this.include(itemPath, includes, excludes)) {
          result.push(itemPath);
        }
      }
    });
    return result;
  }

  /**
   * 替换文件内容
   *
   * @see replaceKeywords
   * @param filePath
   * @param keywordDic
   */
  static replaceFileContent(filePath, keywordDic) {
    let content = _fsExtra.default.readFileSync(filePath, 'utf-8');
    content = this.replaceKeywords(content, keywordDic);
    _fsExtra.default.writeFileSync(filePath, content);
  }

  /**
   * 替换字符串内容
   *
   * @example
   * // 替换 abc{{value}} 为 abc123
   * replaceKeywords(`abc{{value}}`, {value: '123'})
   *
   * @param content
   * @param keywordDic
   * @returns
   */
  static replaceKeywords(content, keywordDic) {
    return content.replace(/{{(.*?)}}/gm, (origin, key) => {
      key = key.trim();
      return keywordDic[key] !== undefined ? keywordDic[key] : origin;
    });
  }
  /**
   * 更新文件路径
   * @param filePath
   * @param newFileName
   * @returns
   * @example updateFileName('./a/b/c.ux', 'd.js') 转换结果为'./a/b/d.js'
   */
  static updateFileName(filePath, newFileName) {
    return filePath && newFileName ? _path.default.join(_path.default.dirname(filePath), newFileName) : '';
  }

  /**
   * 复制文件
   * @param sourceDir 源目录
   * @param targetDir 目标目录
   * @param excludes  不包含的文件
   * @param includes 包含的文件
   */
  static copyFiles(sourceDir, targetDir, excludes, includes) {
    const files = FileUtil.readAlldirSync(sourceDir, includes, excludes);
    for (let file of files) {
      const targetPath = _path.default.join(targetDir, _path.default.relative(sourceDir, file));
      _fsExtra.default.copySync(file, targetPath, {
        overwrite: true
      });
    }
  }
  /**
   * 检查指定的文件列表是否全部存在
   * @param fileList 文件列表
   */
  static checkFilePath(fileList) {
    return fileList?.every(filePath => _fsExtra.default.existsSync(filePath));
  }
  static async del(patterns, options) {
    return (0, _rimraf.rimrafSync)(patterns, {
      glob: true,
      ...options
    });
  }

  /**
   * 创建目录
   *
   * *支持 windows 系统创建隐藏目录
   * @param path
   * @param options
   * @returns
   */
  static mkdirSync(path, options) {
    const result = _fsExtra.default.mkdirSync(path, options);
    if (_fswin.default && options?.hidden) {
      _fswin.default.setAttributesSync(path.toString(), {
        IS_HIDDEN: true
      });
    }
    return result;
  }
}
var _default = exports.default = FileUtil;
