// @aiot-toolkit/shared-utils/lib/utils/NetworkUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ansiColors = _interopRequireDefault(require("ansi-colors"));
var _cliProgress = _interopRequireDefault(require("cli-progress"));
var _koa = _interopRequireDefault(require("koa"));
var _koaRouter = _interopRequireDefault(require("koa-router"));
var _koaStatic = _interopRequireDefault(require("koa-static"));
var _os = _interopRequireDefault(require("os"));
var _portfinder = _interopRequireDefault(require("portfinder"));
var _simpleGit = _interopRequireDefault(require("simple-git"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * NetworkUtil
 */
class NetworkUtil {
  /**
   * clone git 项目
   * @param url
   * @param localPath
   * @param showProgress 是否显示进度条，默认为 true，会在命令行工具中显示绿色进度条
   * @returns
   */
  static gitClone(url, localPath) {
    let showProgress = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
    return new Promise((resolve, reject) => {
      let bar;
      if (showProgress) {
        bar = new _cliProgress.default.SingleBar({
          format: `progress: {percentage}% ${_ansiColors.default.green(`{bar}`)}  {value}/{total}`
        }, _cliProgress.default.Presets.shades_classic);
        bar.start(100, 0);
      }
      (0, _simpleGit.default)({
        progress: data => {
          if (bar) {
            bar.setTotal(data.total);
            bar.update(data.processed);
          }
        }
      }).clone(url, localPath).then(() => {
        resolve();
      }).catch(error => {
        reject(error);
      }).finally(() => {
        if (bar) {
          bar.stop();
        }
      });
    });
  }

  /**
   * 创建 http 服务器
   * @param data
   * @returns
   */
  static async createHttpServer(data) {
    const {
      wantPort,
      routeConfigList,
      events,
      staticFolder
    } = data;
    const app = new _koa.default();
    const port = await _portfinder.default.getPortPromise({
      port: wantPort
    });
    if (staticFolder) {
      app.use((0, _koaStatic.default)(staticFolder));
    }
    const router = new _koaRouter.default();
    for (const config of routeConfigList) {
      const {
        routeList
      } = config;
      for (const route of routeList) {
        const {
          path,
          method = 'get',
          handler
        } = route;
        router[method](path, handler);
      }
    }
    app.use(router.routes());
    const server = app.listen(port, () => {
      events?.onSuccess?.(port);
    });
    return {
      server,
      port,
      app
    };
  }
  static getIPv4IPAddress() {
    const ifaces = _os.default.networkInterfaces();
    if (!ifaces) {
      return;
    }
    for (const prop in ifaces) {
      const iface = ifaces[prop];
      if (iface) {
        for (let item of iface) {
          if (item.family === 'IPv4' && !item.internal && item.address !== '127.0.0.1') {
            return item.address;
          }
        }
      }
    }
  }

  /**
   * 从请求中获取客户端 ip
   * @param ctx koa的请求上下文
   * @returns
   */
  static getClientIp(ctx) {
    const request = ctx.request;
    const ip = request.headers['x-forwarded-for'] || request.header['x-real-ip'] || request.socket.remoteAddress;
    if (ip) {
      return NetworkUtil.stripPrefixForIPV4MappedIPV6Address(ip.toString());
    }
  }

  /**
   * 从网络地址中获取真实 ip
   *
   * @example
   * @param ip
   * @returns
   */
  static stripPrefixForIPV4MappedIPV6Address(ip) {
    if (/^::1$/.test(ip)) {
      ip = '127.0.0.1';
    }
    if (/^::.{0,4}:(\d{1,3}\.){3}\d{1,3}/.test(ip)) {
      ip = ip.replace(/^.*:/, '');
    }
    return ip;
  }
}
var _default = exports.default = NetworkUtil;
