// @aiot-toolkit/shared-utils/lib/utils/QuickAppDocHelp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VelaAppDocHelp = exports.QuickAppDocWidgetType = exports.QuickAppDocHelper = void 0;
var _ColorConsole = _interopRequireDefault(require("../ColorConsole"));
var _ILog = require("../interface/ILog");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * 文档帮助
 */

class QuickAppDocHelper {
  HOST = 'https://doc.quickapp.cn';
  getWidgetsUrl(option) {
    const {
      componentName,
      chapter
    } = option;
    return `${this.HOST}/widgets/${componentName}.html${chapter ? `#${chapter}` : ''}`;
  }
  createWidgetReferenceMessage(option) {
    return {
      word: `(Reference: ${this.getWidgetsUrl(option)})`,
      style: {
        ..._ColorConsole.default.getStyle(_ILog.Loglevel.INFO),
        italic: true,
        bold: true
      }
    };
  }
}
exports.QuickAppDocHelper = QuickAppDocHelper;
class VelaAppDocHelp {
  HOST = 'https://iot.mi.com/vela/quickapp';
  categoryDic = {
    container: ['div', 'list', 'list-item', 'scroll', 'stack', 'swiper'],
    basic: ['text', 'span', 'a', 'image', 'image-animator', 'progress', 'marquee', 'chart', 'qrcode', 'barcode'],
    form: ['input', 'picker', 'switch', 'slider']
  };
  findCategoryName(componentName) {
    for (let key in this.categoryDic) {
      if (this.categoryDic[key].includes(componentName)) {
        return key;
      }
    }
    return '';
  }
  getWidgetsUrl(option) {
    const {
      componentName,
      chapter
    } = option;
    const componentCategory = this.findCategoryName(componentName);
    return `${this.HOST}/zh/components/${componentCategory}/${componentName}.html${chapter ? `#${chapter}` : ''}`;
  }
  createWidgetReferenceMessage(option) {
    return {
      word: `(Reference: ${this.getWidgetsUrl(option)})`,
      style: {
        ..._ColorConsole.default.getStyle(_ILog.Loglevel.INFO),
        italic: true,
        bold: true
      }
    };
  }
}
exports.VelaAppDocHelp = VelaAppDocHelp;
let QuickAppDocWidgetType = exports.QuickAppDocWidgetType = /*#__PURE__*/function (QuickAppDocWidgetType) {
  QuickAppDocWidgetType["ATTRIBUTE"] = "\u5C5E\u6027";
  QuickAppDocWidgetType["STYLE"] = "\u6837\u5F0F";
  QuickAppDocWidgetType["EVENT"] = "\u4E8B\u4EF6";
  return QuickAppDocWidgetType;
}({});
