// @aiot-toolkit/shared-utils/lib/utils/StringUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/**
 * StringUtil
 */
class StringUtil {
  /**  使用正则表达式去除开头和结尾的 ' " ` */
  static removeQuotes(str) {
    return str.replace(/^['"`]+|['"`]+$/g, '');
  }

  /**
   * 中划线文本转为驼峰
   *
   * aaa-bbb 转换为 aaaBbb
   * @param value
   * @returns
   */
  static hyphenedToCamel(value) {
    return value?.replace(/-([a-z])/g, (_, match) => match.toUpperCase());
  }

  /** 首字母大写 */
  static capitalizeFirstLetter(str) {
    if (!str) return '';
    return str.charAt(0).toUpperCase() + str.slice(1);
  }

  /** 逗号分割的字符串转数组 */
  static string2arrayByComma(value) {
    if (!value) return [];
    return value.replace(' ', '').split(',');
  }
  /**
   * 数组转字符串
   * @param arr
   * @param styleTransform 是否为style解析树的数组转换
   * @returns
   */
  static arrayToString(arr) {
    if (Array.isArray(arr)) {
      if (arr.length <= 0) {
        return '[]';
      }
    } else {
      return '';
    }
    const tempArr = arr.map(item => {
      let valueStr = '';
      switch (typeof item) {
        case 'object':
          if (Array.isArray(item)) {
            valueStr = StringUtil.arrayToString(item);
          } else {
            valueStr = StringUtil.objectToString(item);
          }
          break;
        case 'string':
          valueStr = JSON.stringify(item);
          break;
        default:
          valueStr = item;
          break;
      }
      return valueStr;
    });
    return `[${tempArr.join(',')}]`;
  }
  /**
   * 对象转字符串
   * @param obj
   * @param styleTransform 是否为style解析树的数组转换
   * @returns
   */
  static objectToString(obj) {
    return `{${Object.keys(obj).map(key => {
      const value = obj[key];
      if (value === undefined) {
        return '';
      }
      let valueStr = '';
      switch (typeof value) {
        case 'object':
          if (Array.isArray(value)) {
            valueStr = StringUtil.arrayToString(value);
          } else {
            valueStr = StringUtil.objectToString(value);
          }
          break;
        case 'function':
          valueStr = value.toString();
          break;
        default:
          valueStr = value;
          break;
      }
      return `"${key}":${valueStr}`;
    }).filter(item => Boolean(item)).join(',\n')}}`;
  }
}
var _default = exports.default = StringUtil;
