// @aiot-toolkit/aiotpack/lib/compiler/javascript/JavascriptCompiler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _core = require("@rspack/core");
var _lodash = _interopRequireDefault(require("lodash"));
var _path = _interopRequireDefault(require("path"));
var _CompileMode = _interopRequireDefault(require("../enum/CompileMode"));
var _AndroidWebpackConfigurator = _interopRequireDefault(require("./android/AndroidWebpackConfigurator"));
var _VelaWebpackConfigurator = _interopRequireDefault(require("./vela/VelaWebpackConfigurator"));
var _UxCompileUtil = _interopRequireDefault(require("./vela/utils/UxCompileUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class JavascriptCompiler {
  QUICKAPP_CONFIG = 'quickapp.config.js';
  constructor(context, onLog) {
    this.context = context;
    this.onLog = onLog;
  }
  async compile(param) {
    return new Promise(async (resolve, reject) => {
      try {
        await this.clean(param);
        const config = this.createWebpackConfig(param);
        (0, _core.rspack)(config, (error, stats) => {
          if (error) {
            reject({
              errors: [error]
            });
          } else {
            const statsObj = stats?.toJson({});
            if (statsObj) {
              const {
                errors,
                warnings
              } = statsObj;
              if (errors?.length || warnings?.length) {
                reject({
                  errors,
                  warnings
                });
              } else {
                resolve();
              }
            }
          }
        });
      } catch (error) {
        reject({
          errors: [error]
        });
      }
    });
  }

  /**
   * 压缩插件
   *
   * dev模式，只处理console，其它禁用
   * prod模式，其它使用默认值
   * @param mode
   * @returns
   */
  createMinimizerRspackPlugin(compileOption) {
    const {
      mode,
      dropConsole
    } = compileOption;
    const translateDropConsole = () => {
      let dropConsoleValue;
      if (dropConsole === undefined) {
        dropConsoleValue = false;
      } else if (dropConsole === 'true') {
        dropConsoleValue = true;
      } else if (dropConsole === 'false') {
        dropConsoleValue = false;
      } else {
        dropConsoleValue = dropConsole;
      }
      return dropConsoleValue;
    };
    const createCompressValue = dropConsoleValue => {
      const result = {};
      if (typeof dropConsoleValue === 'boolean') {
        result.drop_console = dropConsoleValue;
      } else if (typeof dropConsoleValue === 'string') {
        result.pure_funcs = dropConsoleValue.split(',').map(item => item.trim()).filter(Boolean).map(item => `console.${item}`);
      }
      return result;
    };
    if (mode === _CompileMode.default.DEVELOPMENT) {
      return new _core.rspack.SwcJsMinimizerRspackPlugin({
        minimizerOptions: {
          module: true,
          minify: false,
          mangle: false,
          compress: {
            defaults: false,
            ...createCompressValue(translateDropConsole())
          }
        }
      });
    } else {
      return new _core.rspack.SwcJsMinimizerRspackPlugin({
        minimizerOptions: {
          module: true,
          compress: {
            ...createCompressValue(translateDropConsole())
          }
        }
      });
    }
  }
  createWebpackConfig(param) {
    const configurator = this.getConfigurator(param);
    if (!configurator) {
      throw new Error(`This project is not supported`);
    }
    configurator.param = param;
    // 如果configurator有 create 函数，直接返回
    // 否则，结合configurator生成默认配置，并使用
    if (configurator.create) {
      return configurator.create();
    }
    const {
      projectPath,
      mode,
      devtool,
      outputPath
    } = param;
    const buildPath = _path.default.resolve(projectPath, outputPath);
    const quickAppConfig = _sharedUtils.CommonUtil.requireModule(_path.default.join(param.projectPath, this.QUICKAPP_CONFIG));
    let result = {
      context: projectPath,
      mode,
      devtool: devtool,
      output: {
        globalObject: 'window',
        filename: '[name].js',
        publicPath: './',
        path: buildPath
      },
      module: {
        rules: [{
          test: /\.ux$/,
          // 匹配以 .ux 结尾的文件
          exclude: /node_modules/,
          use: [{
            loader: _path.default.join(__dirname, '../javascript/vela/utils/webpackLoader/extractMapData.js')
          }]
        },
        // node_modules下的ux文件
        {
          test: /\.ux$/,
          include: /node_modules/,
          use: [{
            loader: _path.default.join(__dirname, '../javascript/vela/utils/webpackLoader/WebpackUxLoader'),
            options: {
              compileParam: param,
              context: this.context,
              onLog: this.onLog
            }
          }]
        }, {
          test: /\.js$/,
          include: /node_modules/,
          use: [{
            loader: _path.default.join(__dirname, '../javascript/vela/utils/webpackLoader/WebpackJsLoader'),
            options: {
              compileParam: param,
              context: this.context,
              onLog: this.onLog
            }
          }]
        }]
      },
      optimization: {
        minimize: true,
        minimizer: [this.createMinimizerRspackPlugin(param)]
      },
      resolve: {
        extensions: ['.js', '.ts', '.ux']
      },
      stats: {
        builtAt: false,
        entrypoints: false,
        children: false,
        chunks: false,
        chunkModules: false,
        chunkOrigins: false,
        modules: false,
        version: false,
        assets: false
      }
    };
    if (configurator.createEntry) {
      result.entry = configurator.createEntry();
    }
    if (configurator.createRules && result.module) {
      const readyRules = result.module.rules;
      const configuratorRules = configurator.createRules();
      result.module.rules = Array.isArray(readyRules) && readyRules.length > 0 ? [...configuratorRules, ...readyRules] : configuratorRules;
      // 判断devtool类型 inline-source-map时，添加第0列关系映射
      if (devtool === 'inline-source-map') {
        result.module.rules.unshift({
          test: /\.ux$/,
          // 匹配以 .ux 结尾的文件
          use: [{
            loader: _path.default.join(__dirname, '../javascript/vela/utils/webpackLoader/addColSourceMap.js')
          }]
        });
      }
    }
    if (configurator.createPlugins) {
      result.plugins = configurator.createPlugins();
    }
    if (configurator.hook) {
      configurator.hook(result);
    }

    // 合并用户自定义的webpack配置
    // 数组，则合并数组(默认合并数组中相同序号的项)
    result = _lodash.default.mergeWith({}, result, quickAppConfig?.webpack, (source, target) => {
      if (Array.isArray(target) && Array.isArray(source)) {
        return [...target, ...source];
      }
    });

    // 触发用户配置的posthook
    if (quickAppConfig?.postHook) {
      quickAppConfig.postHook(result);
    }
    return result;
  }
  getConfigurator(param) {
    const {
      projectPath,
      sourceRoot
    } = param;
    const projectType = _sharedUtils.ProjectType.getProjectType(projectPath, sourceRoot);
    switch (projectType) {
      case _sharedUtils.ProjectType.VELA_UX:
        return new _VelaWebpackConfigurator.default();
      case _sharedUtils.ProjectType.ANDDROID_UX:
        return new _AndroidWebpackConfigurator.default();
      default:
        break;
    }
  }
  async clean(param) {
    const {
      outputPath,
      releasePath,
      projectPath
    } = param;
    _UxCompileUtil.default.clean([outputPath, releasePath].map(item => _path.default.resolve(projectPath, item)));
  }
}
var _default = exports.default = JavascriptCompiler;
