// @aiot-toolkit/parser/lib/ux/translate/vela/utils/AttributeConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.eventAttributeTranslate = exports.defaultAttributeTranslate = exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _tsMorph = require("ts-morph");
var _ParserUtil = _interopRequireDefault(require("../../../../utils/ParserUtil"));
var _StyleParser = _interopRequireDefault(require("../../../parser/StyleParser"));
var _StyleToTypescript = _interopRequireDefault(require("../StyleToTypescript"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const ATTRIBUTE_CONFIG = {
  class: {
    aliasName: 'classList',
    translate: function (value) {
      const {
        result,
        dynamic
      } = value;
      // 如果是动态值，运行时转成数组
      // 如果是静态值，编译时直接转成数组
      if (dynamic) {
        return `function() { 
          const $classValue$ = (${result})
          if(typeof $classValue$ === 'string'){
            return $classValue$.split(' ').map(item=>item.trim()).filter(Boolean)
          }
          return $classValue$;
        }`;
      }
      return `[${JSON.parse(result).split(' ').map(item => `"${item.trim()}"`).filter(Boolean).join(',')}]`;
    }
  },
  style: {
    translate: async function (value, option, compilerOption, attribute) {
      const {
        result,
        dynamic
      } = value;
      // 如果是动态值，运行时转成对象
      // 如果是静态值，用 styleParse-->styletoTypescript 进行标准的 css 转换
      if (dynamic) {
        return `function() {
        return global.$translateStyle$(${result})
      }`;
      } else {
        const logs = [];
        const onLog = log => {
          logs.push(log);
        };
        // style{ 刚好匹配 style=, 这样不需要额外算位置
        const inlineStyle = `style{${JSON.parse(result)}}`;
        const styleTree = await new _StyleParser.default({
          ...option,
          onLog
        }, compilerOption, path => path, 0).parser(inlineStyle, `inline.css`);
        const styleResult = new _StyleToTypescript.default({
          ...option,
          onLog
        }, compilerOption).translate(styleTree.ast, []);
        const nameLocation = attribute.nameLocation;
        if (nameLocation) {
          const {
            startLine,
            startColumn
          } = nameLocation;
          logs.forEach(log => {
            const position = log.position;
            if (position) {
              position.startLine += startLine - 1;
              position.startColumn += startColumn;
            }
            option.onLog(log);
          });
        }
        const value = styleResult.targetTree?.[0]?.[1];
        return JSON.stringify(value);
      }
    }
  },
  model: {
    translate: function (value) {
      const {
        result
      } = value;
      return `{
        value: function() { return ${result} },
        callback: function(evt) { ${result} = evt.detail}
      }`;
    }
  },
  remotewidget: {
    translate: function (value) {
      const {
        result
      } = value;
      return `function(){ return ${result};}`;
    }
  }
};

/**
 * 转换动态值的默认方法
 * @param value
 * @param disabledWrap 是否禁用函数包裹。因为普通属性不用函数无法动态响应，而 for 的 list 属性设置函数又报错；所以这里临时加一个参数，在生成__list__的地方手动设置为 true
 * @returns
 */
const defaultAttributeTranslate = function (value) {
  let disabledWrap = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : false;
  if (disabledWrap) {
    return value;
  }
  return `function() { return ${value} }`;
};

/**
 * 转换事件属性的值
 *
 * 1. 普通函数不转换: function(){}
 * 2. 函数名：_vm_.a  转换为 _vm_.a(event)
 * 3. 函数调用：_vm_.a(x, y)  转换为 _vm_.a(x, y, event)
 * 4. 箭头函数不转换: ()=>{}
 *
 * @param value
 * @returns
 */
exports.defaultAttributeTranslate = defaultAttributeTranslate;
const eventAttributeTranslate = (options, value, attribute) => {
  const sourceFile = _ParserUtil.default.createSourceFile(value);
  const statementList = sourceFile.getStatements();
  const {
    onLog,
    filePath
  } = options;
  const {
    nameLocation
  } = attribute;
  // 值不是一个表达式，报错
  if (statementList.length !== 1) {
    onLog({
      filePath,
      position: nameLocation,
      level: _sharedUtils.Loglevel.ERROR,
      message: [`expected 1 expression actual ${statementList.length} expressions`, {
        word: value
      }]
    });
    return '';
  }
  // 获取表达式的类型

  // 1
  const statement = statementList[0];
  if (statement.getKind() === _tsMorph.SyntaxKind.FunctionDeclaration) {
    return value;
  }
  const express = statement.getExpression();
  switch (true) {
    // 2
    case express instanceof _tsMorph.PropertyAccessExpression:
      return `function(evt) { return ${value}(evt) }`;
    // 3
    case express instanceof _tsMorph.CallExpression:
      const argList = express.getArguments().map(item => item.getText());
      const name = express.getExpression().getText();
      return `function(evt) { return ${name}(${[...argList, 'evt'].join(', ')}) }`;
    // 4
    case express instanceof _tsMorph.ArrowFunction:
      return value;
    default:
      onLog({
        filePath,
        position: attribute.nameLocation,
        level: _sharedUtils.Loglevel.THROW,
        message: [`Unsupported Event value:`, {
          word: value
        }, ', Expected Values:', {
          style: _sharedUtils.ColorConsole.getStyle(_sharedUtils.Loglevel.SUCCESS),
          word: ['Arrow Function: onClick="()=>{}"', 'Function: onClick="function(){}"', 'FunctionName: onClick="funName"', 'CallExpression: onClick="funName(x, y)"'].map((item, index) => {
            return `\r\n${index + 1}. ${item}`;
          }).join('')
        }]
      });
      return '';
  }
};
exports.eventAttributeTranslate = eventAttributeTranslate;
var _default = exports.default = ATTRIBUTE_CONFIG;
