// @aiot-toolkit/parser/lib/ux/translate/vela/utils/ExtendedBoxStyle.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _StyleUtil = _interopRequireDefault(require("../../../utils/StyleUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const positionList = ['Top', 'Right', 'Bottom', 'Left'];
const unitList = ['px', 'dp', 'Percentage'];
const borderWidthList = ['thin', 'medium', 'thick'];
const borderStyleList = ['none', 'hidden', 'dotted', 'dashed', 'solid', 'double', 'groove', 'ridge', 'inset', 'outset'];
const timingFunctions = ['linear', 'ease', 'ease-in', 'ease-out', 'ease-in-out', 'cubic-bezier', 'step-start', 'step-end', 'jump-start', 'jump-end', 'jump-none', 'jump-both', 'start', 'end'];
const REGEXP_PROPERTY_NAME = /^[a-zA-Z-]+[a-zA-Z0-9-]*/;
/**
 * ExtendedBoxStyle
 */

/**
 * ExtendedBoxStyle
 */
class ExtendedBoxStyle {
  static extendBoxStyle(sourceNode) {
    const targetList = {};
    const {
      name
    } = sourceNode;
    switch (name) {
      case 'borderWidth':
      case 'borderColor':
      case 'margin':
      case 'padding':
        ExtendedBoxStyle.extendPosition(sourceNode, targetList);
        break;
      case 'border':
      case 'borderLeft':
      case 'borderRight':
      case 'borderTop':
      case 'borderBottom':
        ExtendedBoxStyle.extendBorderStyle(sourceNode, targetList);
        break;
      case 'transition':
        ExtendedBoxStyle.extendTransition(sourceNode, targetList);
        break;
      default:
        break;
    }
    return targetList;
  }
  /**
   * 扩展border样式
   * 扩展border的前提是border的格式符合CSS标准
   * 1. 展开border节点为borderWidth、borderStyle、borderColor
   * 2. 扩展borderWidth、borderColor属性
   * 3. 将border[Left、Right、Top、Bottom]由第1步生成的结果赋值给targetList
   * @param sourceNode
   * @param targetList
   */
  static extendBorderStyle(sourceNode, targetList) {
    const {
      value: nodeValue,
      name
    } = sourceNode;
    const extendedBorder = {};
    // 1.
    if (Array.isArray(nodeValue)) {
      nodeValue.forEach(item => {
        ExtendedBoxStyle.extendValueTypeNode(item, extendedBorder, name);
      });
    } else {
      ExtendedBoxStyle.extendValueTypeNode(nodeValue, extendedBorder, name);
    }
    // 2.
    if (extendedBorder.borderColor) {
      ExtendedBoxStyle.extendPosition({
        name: 'borderColor',
        value: [{
          value: extendedBorder.borderColor,
          unit: ''
        }]
      }, targetList);
    }
    if (extendedBorder.borderStyle) {
      targetList['borderStyle'] = extendedBorder.borderStyle;
    }
    if (extendedBorder.borderWidth) {
      ExtendedBoxStyle.extendPosition({
        name: 'borderWidth',
        value: [{
          value: extendedBorder.borderWidth,
          unit: ''
        }]
      }, targetList);
    }
    // 3.
    switch (name) {
      case 'borderLeft':
      case 'borderRight':
      case 'borderTop':
      case 'borderBottom':
        Object.keys(extendedBorder).map(key => {
          targetList[key] = extendedBorder[key] || '';
        });
        break;
      default:
        break;
    }
  }
  /**
   * 扩展元素与位置
   * margin->0：'marginTop', 1：'marginRight', 2：'marginBottom', 3：'marginLeft'
   * 解析时，只有valueList.length > 1才会返回列表
   * 1. 节点值为数组且长度大于1时，待扩展值顺序为数组值的顺序
   * 2. 若待扩展元素无对应数组值，则取对称位置的数组值index-2
   * 3. 节点值为非数组，则取对应节点值即可
   * @param sourceNode
   * @param targetList
   */
  static extendPosition(sourceNode, targetList) {
    const {
      name,
      value: nodeValue
    } = sourceNode;
    // 4. 处理borderWidth,borderColor
    const nameList = ExtendedBoxStyle.splitByUppercase(name);
    positionList.map((pos, index) => {
      const namePos = nameList.length > 1 ? `${nameList[0]}${pos}${nameList[1]}` : `${nameList[0]}${pos}`;
      let value = '';
      // 1.
      if (Array.isArray(nodeValue) && nodeValue.length > 1) {
        // 2.
        const a = nodeValue[index];
        const b = nodeValue[index - 2];
        value = index + 1 <= nodeValue.length ? a.value + a.unit : b.value + b.unit;
      } else {
        // 3.
        const c = nodeValue[0];
        value = c.value + c.unit;
      }
      targetList[namePos] = value;
    });
  }
  /**
   * 按照大写字符拆分字符串
   * bordeWidth -> border,width
   * @param str
   * @returns
   */
  static splitByUppercase(str) {
    return str.split(/(?=[A-Z])/);
  }
  /**
   * 展开border的每个值
   * @param sourceNode
   * @param extendedBorder
   */
  static extendValueTypeNode(sourceNode, extendedBorder, name) {
    // border borderLeft ---> borderWidth borderLeftWidth
    const value = _StyleUtil.default.translateNodeValue(sourceNode);
    if (ExtendedBoxStyle.isBorderWidth(sourceNode)) {
      const width = name + 'Width';
      extendedBorder[width] = value;
    } else if (borderStyleList.indexOf(sourceNode.value) > -1) {
      const style = name + 'Style';
      extendedBorder[style] = value;
    } else {
      const color = name + 'Color';
      extendedBorder[color] = value;
    }
  }
  /**
   * 判断border节点值是否有borderWidth
   * @param nodeValue
   * @returns
   */
  static isBorderWidth(nodeValue) {
    return unitList.indexOf(nodeValue.unit) > -1 || borderWidthList.indexOf(nodeValue.value) > -1;
  }
  /**
   * 第一个时间为 transitionDuration
   * @param sourceNode
   * @param targetList
   */
  static extendTransition(sourceNode, targetList) {
    // 默认值
    targetList.transitionProperty = 'all';
    targetList.transitionTimingFunction = 'ease';
    targetList.transitionDuration = '0s';
    targetList.transitionDelay = '0s';
    let isDuration = true;
    let isFirstProperty = true;
    const {
      value: propertyValues
    } = sourceNode;
    if ('value' in propertyValues[0]) {
      propertyValues.forEach(pValue => {
        pValue = pValue;
        // 是否为timing function
        if (timingFunctions.filter(func => pValue.value.startsWith(func)).length) {
          return targetList.transitionTimingFunction = pValue.value;
        }
        // 是否为duration time
        if (pValue.unit && isDuration) {
          isDuration = false;
          return targetList.transitionDuration = pValue.value + pValue.unit;
        }
        // 是否为delay time
        if (pValue.unit && !isDuration) {
          return targetList.transitionDelay = pValue.value + pValue.unit;
        }
        // 是否为property
        if (REGEXP_PROPERTY_NAME.test(pValue.value)) {
          if (isFirstProperty) {
            isFirstProperty = false;
            targetList.transitionProperty = pValue.value;
          } else {
            targetList.transitionProperty += `, ${pValue.value}`;
          }
          return;
        }
      });
    }
  }
}
var _default = exports.default = ExtendedBoxStyle;
