// @aiot-toolkit/parser/lib/ux/translate/vela/utils/SourceMapUtil.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sourceMap = require("source-map");
/**
 * SourceMapUtil
 */
class SourceMapUtil {
  /**
   * 更新source map
   * @param sourceMapStr 更新前的source map字符串
   * @param offsetInfo 目标代码合并时用到的偏移量
   * @param options
   * @returns 返回更新了map的generator对象
   */
  static async updateSourceMap(sourceMapStr, offsetInfo, options) {
    if (!sourceMapStr) {
      return new _sourceMap.SourceMapGenerator();
    }
    let {
      consumer,
      tempGenerator
    } = await SourceMapUtil.createConsumerAndGenerator(sourceMapStr);
    // 更新map
    SourceMapUtil.updateMappings(consumer, tempGenerator, offsetInfo);
    // 更新sourcesContent
    tempGenerator.setSourceContent(consumer.sources[0], options.content);
    return tempGenerator;
  }
  /**
   * 基于sourceMap字符串创建一个consumer对象、一个mappings为空的tempGenerator对象
   * @param sourceMap
   * @returns
   */
  static async createConsumerAndGenerator(sourceMap) {
    const consumer = await new _sourceMap.SourceMapConsumer(sourceMap);
    // 创建临时source map对象，并置空
    const tempGenerator = await SourceMapUtil.createTempGenerator(sourceMap);
    return {
      consumer,
      tempGenerator
    };
  }
  /**
   * 创建一个mappings为空的tempGenerator对象
   * @param sourceMap
   * @returns
   */
  static async createTempGenerator(sourceMap) {
    const tempSourceMap = JSON.parse(sourceMap);
    tempSourceMap.mappings = [];
    const tempConsumer = await new _sourceMap.SourceMapConsumer(tempSourceMap);
    return _sourceMap.SourceMapGenerator.fromSourceMap(tempConsumer);
  }
  /**
   * 更新tempGenerator的mappings
   * @param consumer 更新前mappings存储在该对象中
   * @param tempGenerator 待更新mappings的对象
   * @param updateInfo 传递更新过程中使用的偏移量
   */
  static updateMappings(consumer, tempGenerator, updateInfo) {
    consumer.eachMapping(mapping => {
      mapping.originalLine += updateInfo.origin.line;
      mapping.originalColumn += updateInfo.origin.col;
      mapping.generatedLine += updateInfo.target.line;
      mapping.generatedColumn += updateInfo.target.col;
      tempGenerator.addMapping({
        generated: {
          line: mapping.generatedLine,
          column: mapping.generatedColumn
        },
        original: {
          line: mapping.originalLine,
          column: mapping.originalColumn
        },
        source: mapping.source,
        name: mapping.name
      });
    });
  }
  /**
   * 从babel后的代码中切割出源码和source map
   * @param appScriptTree
   * @param project
   * @returns
   */
  static splitSourceMap(appScriptTree) {
    // 取出script中的source map
    let sourceMapReg = new RegExp('//# sourceMappingURL=data:application/json;base64,');
    let result = appScriptTree.getFullText().split(sourceMapReg);
    if (!result[1]) {
      sourceMapReg = new RegExp('//# sourceMappingURL=data:application/json;charset=utf-8;base64,');
      result = appScriptTree.getFullText().split(sourceMapReg);
    }
    appScriptTree.replaceWithText(result[0]);
    // 内联的sourcemap编码为base64
    return {
      newScriptTree: appScriptTree,
      scriptSourceMap: result[1] ? Buffer.from(result[1], 'base64').toString() : ''
    };
  }
  /**
   * 计算addStatments中字符串所占行数
   * @param str
   * @returns
   */
  static countStatementLines(originStr) {
    if (!originStr) {
      return 0;
    }
    let lineCount = (originStr.match(/\n/g) || []).length + 1;
    return lineCount;
  }
}
var _default = exports.default = SourceMapUtil;
