// @aiot-toolkit/shared-utils/lib/ColorConsole.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _chalk = _interopRequireDefault(require("chalk"));
var _dayjs = _interopRequireDefault(require("dayjs"));
var _ILog = require("./interface/ILog");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * Log
 */
class ColorConsole {
  static defaultStyle = {
    color: '#cccccc'
  };
  static styleDic = (() => ({
    [_ILog.Loglevel.SUCCESS]: {
      color: '#008000',
      icon: '✅'
    },
    [_ILog.Loglevel.INFO]: {
      color: '#03A9F4',
      icon: '🔷'
    },
    [_ILog.Loglevel.WARN]: {
      color: '#FF9800',
      icon: '🟠'
    },
    [_ILog.Loglevel.ERROR]: {
      color: '#ff0000',
      icon: '❌'
    },
    [_ILog.Loglevel.THROW]: {
      color: '#ff0000',
      bold: true,
      icon: '❌'
    },
    [_ILog.Loglevel.LOG]: ColorConsole.defaultStyle
  }))();
  static getStyle(logLevel) {
    return ColorConsole.styleDic[logLevel] || ColorConsole.defaultStyle;
  }
  static log() {
    for (var _len = arguments.length, message = new Array(_len), _key = 0; _key < _len; _key++) {
      message[_key] = arguments[_key];
    }
    return ColorConsole._log(_ILog.Loglevel.LOG, ...message);
  }
  static error() {
    for (var _len2 = arguments.length, message = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
      message[_key2] = arguments[_key2];
    }
    return ColorConsole._log(_ILog.Loglevel.ERROR, ...message);
  }
  static info() {
    for (var _len3 = arguments.length, message = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
      message[_key3] = arguments[_key3];
    }
    return ColorConsole._log(_ILog.Loglevel.INFO, ...message);
  }
  static success() {
    for (var _len4 = arguments.length, message = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
      message[_key4] = arguments[_key4];
    }
    return ColorConsole._log(_ILog.Loglevel.SUCCESS, ...message);
  }
  static throw() {
    for (var _len5 = arguments.length, message = new Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
      message[_key5] = arguments[_key5];
    }
    ColorConsole._log(_ILog.Loglevel.THROW, ...message);
  }
  static warn() {
    for (var _len6 = arguments.length, message = new Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
      message[_key6] = arguments[_key6];
    }
    return ColorConsole._log(_ILog.Loglevel.WARN, ...message);
  }

  /**
   * 创建复杂格式的消息，可单独设置词语样式
   *
   * @example
   * ```js
   * createMessage("my name is", {message:'****', style:{ color:'red' }})
   * ```
   * @param words
   * @returns
   */
  static createMessage(words) {
    const list = Array.isArray(words) ? words : [words];
    const result = list.map(item => {
      if (typeof item === 'object') {
        return ColorConsole.formatWord(item.word, item.style);
      }
      return ColorConsole.formatWord(item, ColorConsole.defaultStyle);
    }).join(' ');
    return result;
  }
  static _log(logLevel) {
    const style = ColorConsole.styleDic[logLevel] || ColorConsole.defaultStyle;
    for (var _len7 = arguments.length, message = new Array(_len7 > 1 ? _len7 - 1 : 0), _key7 = 1; _key7 < _len7; _key7++) {
      message[_key7 - 1] = arguments[_key7];
    }
    console.log(ColorConsole.createDefaultPrefix({
      level: logLevel,
      message
    }).join(' ') + ColorConsole.createMessage(message.map(item => {
      if (typeof item === 'object') {
        return {
          ...item,
          style: item.style || style
        };
      }
      return item;
    })));
  }
  static formatWord(word, style) {
    if (word === undefined || word === null) {
      return '';
    }
    const {
      color,
      bgColor,
      bold,
      italic,
      underline
    } = style || ColorConsole.defaultStyle;
    let result = word;
    if (color) {
      result = _chalk.default.hex(color)(result);
    }
    if (bgColor) {
      result = _chalk.default.bgHex(bgColor)(result);
    }
    if (bold) {
      result = _chalk.default.bold(result);
    }
    if (italic) {
      result = _chalk.default.italic(result);
    }
    if (underline) {
      result = _chalk.default.underline(result);
    }
    return result;
  }

  /**
   * 按照日志类型打印不同颜色的日志
   * @param log 默认日志类型为info
   * @returns 返回打印的日志信息
   */
  static logger(log) {
    const str = ColorConsole.logToString(log);
    console.log(str);
  }

  /**
   * 日志转换为带颜色的字符串
   *
   * @param log 日志对象
   * @param prefix 前缀列表
   * false-不使用前缀
   * string[]-自定义前缀
   * 默认-[图标、时间、toolkit字样]
   *
   * @example logToString(log, ['myPrefix'])
   */
  static logToString(log, prefix) {
    const {
      message,
      position,
      filePath,
      level = _ILog.Loglevel.INFO
    } = log;

    // 位置消息
    let positionMessage = filePath ? {
      style: {
        ...ColorConsole.getStyle(_ILog.Loglevel.LOG),
        underline: true
      },
      word: [filePath, position?.startLine, position?.startColumn].filter(item => Boolean(item)).join(':')
    } : null;
    const logs = Array.isArray(message) ? message : [message];
    if (positionMessage) {
      logs.unshift(positionMessage);
    }

    // 前缀
    let effectPrefix;
    if (prefix === false) {
      prefix = [];
    }
    if (prefix === undefined) {
      effectPrefix = ColorConsole.createDefaultPrefix(log);
    } else {
      effectPrefix = prefix;
    }
    const style = ColorConsole.styleDic[level] || ColorConsole.defaultStyle;
    return [...effectPrefix].join(' ') + ColorConsole.createMessage(logs.map(item => {
      if (typeof item === 'object') {
        return {
          ...item,
          // 对象格式消息的样式依次使用：[消息自身样式, log样式, 默认样式]
          style: item.style || style
        };
      }
      return item;
    }));
  }

  /**
   * 创建默认消息前缀列表
   *
   * 1. 图标
   * 2. 时间
   * 3. toolkit 字样
   *
   * @param log
   */
  static createDefaultPrefix(log) {
    const {
      level = _ILog.Loglevel.INFO
    } = log;
    const style = ColorConsole.styleDic[level] || ColorConsole.defaultStyle;
    const time = ColorConsole.formatWord(`[${(0, _dayjs.default)().format('MM-DD HH:mm:ss')}] [toolkit]: `, style);
    const icon = style.icon ? style.icon : '';
    return [icon, time];
  }
}
var _default = exports.default = ColorConsole;
