// @aiot-toolkit/aiotpack/lib/compiler/javascript/android/plugin/WrapPlugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _core = require("@rspack/core");
var _webpackSources = require("webpack-sources");
var _UxFileUtils = _interopRequireDefault(require("../../../../utils/ux/UxFileUtils"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * WrapPlugin
 */
class WrapPlugin {
  constructor(param) {
    this.param = param;
  }
  apply(compiler) {
    compiler.hooks.compilation.tap('WrapPlugin', compilation => {
      compilation.hooks.processAssets.tap({
        name: 'WrapPlugin',
        stage: _core.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE
      }, () => {
        this.wrap(compilation);
      });
    });
  }

  /**
   * 添加包含代码
   *
   * 1. 仅 page 文件添加
   * @param compilation
   */
  wrap(compilation) {
    // 获取入口文件
    const entrys = Object.keys(compilation.options.entry).map(item => `${item}.js`);
    entrys.forEach(entry => {
      if (compilation.assets[entry]) {
        const source = compilation.assets[entry];
        const isApp = entry === 'app.js';
        compilation.assets[entry] = isApp ? this.wrapApp(source) : this.wrapPage(source);
      }
    });
  }
  wrapPage(source) {
    return new _webpackSources.ConcatSource(`
(function () {
  var createPageHandler = function () {
  // 转换动态 style 的函数
  var $translateStyle$ = function (value) {
    if (typeof value === 'string') {
      return Object.fromEntries(value.split(';').filter(item => Boolean(item && item.trim())).map(
        item => {
          const matchs = item.match(/([^:]+):(.*)/)
          if (matchs && matchs.length> 2) {
            return [matchs[1].trim().replace(/-([a-z])/g, (_, match) => match.toUpperCase()), matchs[2].trim()]
          }
          return []
        }))}
    return value
  }
  return`, source, `;
  };
  if (typeof window === "undefined") {
    return createPageHandler();
  } else {
    window.createPageHandler = createPageHandler;
  }
})();`);
  }
  wrapApp(source) {
    const {
      projectPath,
      sourceRoot
    } = this.param;
    return new _webpackSources.ConcatSource(`
(function () {
  var $app_define_wrap$ = $app_define_wrap$ || function () {};
  var manifestJson =${JSON.stringify(_UxFileUtils.default.getManifestInfo(projectPath, sourceRoot))}
  var createAppHandler = function () {
      `, source, `
  }
    if (typeof window === "undefined") {
    return createAppHandler();
  } else {
    window.createAppHandler = createAppHandler;
    // H5注入manifest以获取features
    global.manifest = manifestJson;
  }
})();
`);
  }
}
var _default = exports.default = WrapPlugin;
