// @aiot-toolkit/parser/lib/ux/translate/vela/wrap/CbTranslate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ElementConfig = _interopRequireDefault(require("../../../config/vela/ElementConfig"));
var _ElementConfigUtil = _interopRequireDefault(require("../../../utils/ElementConfigUtil"));
var _VelaContext = _interopRequireDefault(require("../VelaContext"));
var _TemplateUtil = _interopRequireDefault(require("../utils/TemplateUtil"));
var _TemplateNodeType = _interopRequireDefault(require("../../../enum/TemplateNodeType"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * cb 函数生成
 *
 * 当满条件时生成
 * 1. 是 block 节点
 *    1. 如果是根元素，生成普通cb 函数
 *    2. 如果"有static 属性" 或 "同时有dynamic/data"，生成 opts 包含 dynamic， data 的 cb 函数
 *
 */
class CbTranslate {
  constructor(options, compilerOption, extractionInformation) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.extractionInformation = extractionInformation;
    this.templateUtil = new _TemplateUtil.default(_VelaContext.default, options, compilerOption, new _ElementConfigUtil.default(_ElementConfig.default));
  }
  match(node) {
    const {
      tagName,
      parentNode
    } = node;
    if (tagName === 'block') {
      const staticAttribute = _TemplateUtil.default.getAttribute(node, 'static');
      const dynamicAttribute = _TemplateUtil.default.getAttribute(node, 'dynamic');
      const dataAttribute = _TemplateUtil.default.getAttribute(node, 'data');
      const isRoot = !parentNode || parentNode.nodeName === _TemplateNodeType.default.DOCUMENT_FRAGMENT;
      if (isRoot || staticAttribute || dynamicAttribute && dataAttribute) {
        return true;
      }
    }
    return false;
  }
  async translate(node, childrenExpression, identifiers, extraParams) {
    const name = '__cb__';
    const opts = {};
    const {
      hasStaticAttribute,
      hasEvent
    } = this.extractionInformation;
    const {
      parentNode,
      attrs
    } = node;
    const isRoot = !parentNode || parentNode.nodeName === _TemplateNodeType.default.DOCUMENT_FRAGMENT;
    if (isRoot) {
      return await this.templateUtil.createWrapCallExpression({
        name,
        opts: await this.templateUtil.attributesToObject(attrs, identifiers, {
          nameToCamel: true
        }),
        elementExpression: `function(){
          return [${childrenExpression}]
        }`
      });
    } else {
      // static: 有static 属性
      if (hasStaticAttribute) {
        opts.static = true;
      }
      // interactive: 有 static 属性 + 有事件属性
      if (hasStaticAttribute && hasEvent) {
        opts.interactive = true;
      }
      // dynamic + data
      const dynamicAttribute = _TemplateUtil.default.getAttribute(node, 'dynamic');
      const dataAttribute = _TemplateUtil.default.getAttribute(node, 'data');
      if (dynamicAttribute && dataAttribute) {
        opts.dynamic = true;
        opts.data = await this.templateUtil.translateAttributeValue(dataAttribute, identifiers);
      }
      return this.templateUtil.createWrapCallExpression({
        name,
        opts,
        elementExpression: `function(${opts.dynamic ? '$data, ' : '_0,'}${extraParams || ''}){
          return [${childrenExpression}]
        }`
      });
    }
  }
}
var _default = exports.default = CbTranslate;
