// @aiot-toolkit/parser/lib/ux/translate/vela/wrap/CiTranslate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _ElementConfig = _interopRequireDefault(require("../../../config/vela/ElementConfig"));
var _ElementConfigUtil = _interopRequireDefault(require("../../../utils/ElementConfigUtil"));
var _VelaContext = _interopRequireDefault(require("../VelaContext"));
var _TemplateUtil = _interopRequireDefault(require("../utils/TemplateUtil"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * CiTranslate
 */
class CiTranslate {
  constructor(options, compilerOption) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.templateUtil = new _TemplateUtil.default(_VelaContext.default, options, compilerOption, new _ElementConfigUtil.default(_ElementConfig.default));
  }
  match(node) {
    return Boolean(_TemplateUtil.default.getAttribute(node, 'if')) || Boolean(_TemplateUtil.default.getAttribute(node, 'elif')) || Boolean(_TemplateUtil.default.getAttribute(node, 'else'));
  }

  /**
   * 转有 if 属性的元素
   * 1. 转换为 ci
   * 2. if 属性的修饰符放到 ci 的 opts 中
   * @param node
   * @param childrenExpression
   * @param identifiers
   * @param extraParams
   * @returns
   */
  async translate(node, childrenExpression, identifiers, extraParams) {
    const {
      onLog,
      filePath
    } = this.options;
    if (this.match(node)) {
      const name = '__ci__';
      const condition = this.templateUtil.getIfCoditionList(node);
      if (!condition) {
        onLog({
          filePath,
          position: _TemplateUtil.default.parse5LocationToPosition(node.sourceCodeLocation?.startTag),
          level: _sharedUtils.Loglevel.THROW,
          message: [{
            word: 'if'
          }, `attribute missing`]
        });
        return '';
      }
      const shownValue = this.templateUtil.genIfConditionList(condition, identifiers);
      const modifiers = this.getModifiers(node);
      const opts = {
        shown: `function(){ return ${shownValue}; }`,
        modifiers
      };
      return this.templateUtil.createWrapCallExpression({
        name,
        opts,
        elementExpression: `function(${extraParams || ''}){
          return [${childrenExpression}]
        }`
      });
    }
    return childrenExpression;
  }
  getModifiers(node) {
    return _TemplateUtil.default.createDecoratorData(node, 'if', 'shown');
  }
}
var _default = exports.default = CiTranslate;
