// @aiot-toolkit/parser/lib/ux/translate/vela/plugins/e2e.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = e2e;
var t = _interopRequireWildcard(require("@babel/types"));
var parser = _interopRequireWildcard(require("@babel/parser"));
var _path = _interopRequireDefault(require("path"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _sharedUtils = require("@aiot-toolkit/shared-utils");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/**
 * 用于添加应用自动化测试代码
 * 分为两种情况
 * 1. app文件
 * 2. page 文件
 */
function e2e() {
  return {
    visitor: {
      ExportDefaultDeclaration(path, _ref) {
        let {
          opts
        } = _ref;
        const opts2 = opts;
        const options = opts2.options;
        switch (options.fileType) {
          case 'page':
            addPageTestCode(path, opts2);
            break;
          case 'app':
            addAppTestCode(path);
            break;
        }
      }
    }
  };
}

/**
 * 添加 app 的测试代码
 *
 * 引入测试框架文件
 * @param path
 * @param opts
 */
function addAppTestCode(path) {
  const runtimePath = _path.default.join(__dirname, '../runtime/velaTestLibrary.js');
  const appImport = t.importDeclaration([], t.stringLiteral(runtimePath));
  if (path.parent.type === 'Program') {
    path.parent.body.unshift(appImport);
  }
}

/**
 * 添加页面的测试代码
 * 1. 如果测试文件存在，则
 *    a. import 测试文件
 *    b. 设置测试函数: 如果 onTestStart 方法已存在，则在其内部添加测试代码；否则，添加 onTestStart方法
 *
 *
 * 2. 如果启用自动路由，且当前页面是首页
 *    a. 从routerList 中过滤出有测试文件的路由
 *    b. 在末尾添加 $generateAutoRunTest
 *    c. 在onShow 方法中调用 $generateAutoRunTest
 *
 *
 * @description 示例
 *
 * 源码 `src/home/myHome.ux`
 * ```
 *
 * export default {}
 * ```
 *
 * 产物
 * ```
 * import fnTestCase from 'test/home/myHome.js'
 *
 * export default {
 *  onTestStart: (){
 *      // code
 *   }
 * }
 * ```
 *
 * @param path
 * @param opts
 */
function addPageTestCode(path, opts) {
  const e2eConfig = getE2eConfig(opts);
  const testFilePath = getTestFilePath(opts.options.filePath, opts, e2eConfig);
  // 1
  if (_fsExtra.default.existsSync(testFilePath)) {
    // 1.a
    const relativePath = _path.default.relative(_path.default.dirname(opts.options.filePath), testFilePath);
    const vmImport = t.importDeclaration([t.importDefaultSpecifier(t.identifier('fnTestCase'))], t.stringLiteral(relativePath));
    if (path.parent.type === 'Program') {
      path.parent.body.unshift(vmImport);
    }

    // 1.b
    if (path.node.declaration.type === 'ObjectExpression') {
      const properties = path.node.declaration.properties;
      insertFun('onTestStart', generateTestFn(), properties);
    }
  }

  // 2
  if (isIndexPage(opts) && e2eConfig.autoRoute) {
    let {
      routerList,
      pageSleep
    } = e2eConfig.autoRoute;
    routerList = routerList?.filter(item => hasTestFile(item, opts, e2eConfig));
    if (!routerList?.length) {
      console.log('missing autoRoute.routerList');
      return;
    }
    // 2.a

    if (path.node.declaration.type === 'ObjectExpression') {
      const properties = path.node.declaration.properties;

      // 2.b
      insertFun('$generateAutoRunTest', generateAutoRouterFn(pageSleep, routerList), properties);
      // 2.c
      insertFun('onShow', generateAutoRouterInOnShow(), properties);
    }
  }
}

/**
 * 插入函数
 *
 * 如果已存在，把函数体插入到指定函数末尾；不存在，则添加
 *
 * @param name 函数名
 * @param body 函数体
 */
function insertFun(name, body, properties) {
  const fun = properties.find(p => p.type === 'ObjectMethod' && p.key.name === name);
  if (fun) {
    fun.body.body.push(...body.body);
  } else {
    properties.push(t.objectMethod('method', {
      name: name,
      type: 'Identifier'
    }, [], body));
  }
}
function hasTestFile(router, opts, e2eConfig) {
  const manifest = getManifest(opts);
  if (manifest) {
    const {
      pages
    } = manifest.router;
    const pageRouter = pages[router];
    if (!pageRouter) {
      opts.options.onLog({
        level: _sharedUtils.Loglevel.WARN,
        message: [{
          word: router
        }, `missing router in manifest.json`]
      });
      return false;
    }
    const pagePath = `${opts.compilerOption.sourceRoot}/${router}/${pageRouter.component}`;
    const testFilePath = getTestFilePath(pagePath, opts, e2eConfig);
    return _fsExtra.default.existsSync(testFilePath);
  }
  return false;
}
function isIndexPage(opts) {
  const manifest = getManifest(opts);
  if (manifest) {
    const {
      entry,
      pages
    } = manifest.router;
    const entryItem = pages[entry];
    const entryPath = `${entry}/${entryItem.component}`;
    const sourcePath = _path.default.join(opts.options.projectPath, opts.compilerOption.sourceRoot);
    const filePath = _path.default.relative(sourcePath, opts.options.filePath);
    const filePathParsed = _path.default.parse(filePath);
    return entryPath === `${filePathParsed.dir}/${filePathParsed.name}`;
  }
  return false;
}
function getManifest(opts) {
  const manifestPath = _path.default.resolve(opts.options.projectPath, opts.compilerOption.sourceRoot, 'manifest.json');
  if (manifestPath) {
    return _fsExtra.default.readJSONSync(manifestPath);
  }
}
function getE2eConfig(opts) {
  const defaultTestDir = 'test';
  let result = {
    testDir: _path.default.join(opts.options.projectPath, defaultTestDir)
  };
  const {
    projectPath
  } = opts.options;
  const {
    e2eConfigPath
  } = opts.compilerOption;
  if (e2eConfigPath) {
    const configFilePath = _path.default.resolve(projectPath, e2eConfigPath);
    const configFileDir = _path.default.dirname(configFilePath);
    if (_fsExtra.default.existsSync(configFilePath)) {
      result = _fsExtra.default.readJSONSync(configFilePath, 'utf-8');

      // 目录转换为绝对路径
      if (result) {
        result.testDir = _path.default.resolve(configFileDir, result.testDir || defaultTestDir);
      }
    } else {
      throw new Error(`e2e Configuration file does not exist: ${configFilePath}`);
    }
  }
  return result;
}

/**
 * 获取源文件对应的测试文件地址
 * @param sourceFile 源文件
 * @param opts
 * @param e2eConfig
 * @returns
 */
function getTestFilePath(sourceFile, opts, e2eConfig) {
  const {
    projectPath
  } = opts.options;
  const {
    sourceRoot
  } = opts.compilerOption;
  const relativeSrc = _path.default.relative(_path.default.join(projectPath, sourceRoot || ''), sourceFile);
  const parsed = _path.default.parse(relativeSrc);
  return _path.default.join(e2eConfig.testDir, parsed.dir, `${parsed.name}.js`);
}

/**
 * 获取测试函数的函数体
 * @returns
 */
function generateTestFn() {
  const fn = `function(){
  this.$options = this._options || this.$options;
  if (this.$options._descriptor) {
    this.$options._descriptor["back"] = {
      access: "public",
    };
  }
  global.CASE_TEST_START = global.CASE_TEST_START || 1000;
  global.qat = new VelaQuickappTest();
  setTimeout(() => {
    typeof fnTestCase === 'function' && fnTestCase(this)
    qat.run(() => {
      if (this.back !== "false") {
        console.info("拥有关联测试用例，测试完毕，返回到之前的页面");
        function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
        var _system = _interopRequireDefault($app_require$('@app-module/system.router'));
        _system.default.back();
      }
    });
  }, global.CASE_TEST_START);
  }`;
  return parser.parseExpression(fn).body;
}
function generateAutoRouterFn() {
  let sleep = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 3000;
  let routerList = arguments.length > 1 ? arguments[1] : undefined;
  const fn = `function(){
  const routerList = ${JSON.stringify(routerList)}
  if (!routerList) {
      console.log('missing routerList');
      return;
    }
    if (this.autoTestPageIndex === undefined) {
      this.autoTestPageIndex = -1;
    }

    setTimeout(() => {
      if (++this.autoTestPageIndex < routerList.length) {
        const uri = routerList[this.autoTestPageIndex];
        console.log(\`auto test page=\${uri}, index=\${this.autoTestPageIndex}\`);
        router.push({ uri });
      } else {
        console.log('auto test finish');
      }
    }, ${sleep});
  }`;
  return parser.parseExpression(fn).body;
}
function generateAutoRouterInOnShow() {
  const fn = `function(){
  this.$generateAutoRunTest();
  }`;
  return parser.parseExpression(fn).body;
}
