// @aiot-toolkit/aiotpack/lib/compiler/javascript/vela/model/Package.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
/**
 * 文件资源
 */

/**
 * 定义文件包
 *
 * 相当于虚拟的 rpk 包
 */
class Package {
  /**
   * 包前缀
   */

  /**
   * 子包名称；整包没有
   */

  /**
   * 是否为独立包
   */
  standalone = false;

  /**
   * 后缀名
   */

  /**
   * 资源路径匹配
   */

  _resourceList = [];

  /**
   * 资源文件列表
   */
  get resourceList() {
    return this._resourceList;
  }
  _resourceDic = {};

  /**
   * 资源文件字典，以高效检查资源是否存在
   */
  get resourceDic() {
    return this._resourceDic;
  }

  /**
   * 是否已包含指定文件
   * @param filePath 文件路径
   * @returns
   */
  include(filePath) {
    return this.resourceDic[filePath];
  }
  getResource(filePath) {
    return this.resourceList.find(item => item.fileBuildPath === filePath);
  }

  /**
   * 添加资源
   * @param data
   * @returns
   */
  addResource(data) {
    const {
      fileBuildPath
    } = data;
    if (this.include(fileBuildPath)) {
      _sharedUtils.ColorConsole.throw(`File ${fileBuildPath} is added repeatedly`);
      return;
    }
    this._resourceDic[fileBuildPath] = true;
    this._resourceList.push(data);
  }
  constructor(data) {
    Object.assign(this, data);
    this.fileName = [this.filePrefix, this.fileSubname, this.fileSuffix].filter(item => Boolean(item)).join('.');
  }
}
var _default = exports.default = Package;
