// @aiot-toolkit/aiotpack/lib/utils/ux/FileCompare.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.compareFiles = compareFiles;
var _crypto = require("crypto");
var _fs = _interopRequireDefault(require("fs"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
function compareFiles(filePath1, filePath2) {
  const startTime = Date.now();

  // 快速检查：文件大小不同则内容必然不同
  const size1 = _fs.default.statSync(filePath1).size;
  const size2 = _fs.default.statSync(filePath2).size;
  if (size1 !== size2) {
    return {
      isEqual: false,
      details: {
        hash1: '',
        hash2: '',
        fileSize1: size1,
        fileSize2: size2,
        comparisonTime: Date.now() - startTime
      }
    };
  }

  // 计算两个文件的哈希
  const hash1 = computeFileHash(filePath1);
  const hash2 = computeFileHash(filePath2);
  return {
    isEqual: hash1 === hash2,
    details: {
      hash1,
      hash2,
      fileSize1: size1,
      fileSize2: size2,
      comparisonTime: Date.now() - startTime
    }
  };
}
function computeFileHash(filePath) {
  const fileData = _fs.default.readFileSync(filePath);
  const hash = (0, _crypto.createHash)('SHA256').update(fileData);
  return hash.digest('hex');
}
