// @aiot-toolkit/aiotpack/lib/utils/ux/ManifestSchema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const ManifestSchema = {
  type: 'object',
  required: ['package', 'name', 'icon', 'versionCode', 'config', 'router'],
  properties: {
    package: {
      type: 'string',
      minLength: 1
    },
    name: {
      type: 'string',
      minLength: 1
    },
    icon: {
      type: 'string',
      minLength: 1
    },
    banner: {
      type: 'string'
    },
    versionName: {
      type: 'string'
    },
    versionCode: {
      type: ['number']
    },
    minPlatformVersion: {
      type: ['number']
    },
    features: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            type: 'string'
          }
        }
      }
    },
    config: {
      type: 'object',
      properties: {
        logLevel: {
          enum: ['off', 'error', 'warn', 'info', 'log', 'debug']
        },
        designWidth: {
          type: ['number', 'string']
        },
        data: {
          // TODO 全局数据对象，属性名不能以$或_开头，
          type: 'object'
        },
        background: {
          // TODO 更详细的后台运行配置信息，
          type: 'object'
        },
        network: {
          // TODO 更详细的网络配置信息，
          type: 'object'
        }
      }
    },
    router: {
      // TODO 更详细的后台运行配置信息，
      type: 'object',
      required: ['entry', 'pages'],
      properties: {
        entry: {
          type: 'string'
        },
        // TODO 看下 key 不确定能否校验值，
        pages: {
          type: 'object'
        },
        errorPage: {
          type: 'string'
        },
        // TODO 看下 key 不确定能否校验值，
        widgets: {
          type: 'object'
        }
      }
    },
    display: {
      type: 'object',
      properties: {
        backgroundColor: {
          // TODO HEXColor
          type: 'string'
        },
        fullScreen: {
          type: 'boolean'
        },
        titleBar: {
          type: 'boolean'
        },
        titleBarBackgroundColor: {
          // TODO HEXColor
          type: 'string'
        },
        titleBarTextColor: {
          // TODO HEXColor
          type: 'string'
        },
        menu: {
          type: 'boolean'
        },
        windowSoftInputMode: {
          enum: ['adjustPan', 'adjustResize']
        },
        pages: {
          type: 'object'
        },
        orientation: {
          enum: ['portrait', 'landscape']
        },
        statusBarImmersive: {
          type: 'boolean'
        },
        statusBarTextStyle: {
          enum: ['light', 'dark', 'auto']
        },
        statusBarBackgroundColor: {
          type: 'string'
        },
        statusBarBackgroundOpacity: {
          type: 'number'
        },
        fitCutout: {
          enum: ['none', 'portrait', 'landscape']
        },
        textSizeAdjust: {
          enum: ['none', 'auto']
        },
        themeMode: {
          enum: [-1, 0, 1]
        },
        menuBarData: {
          // TODO 详细校验
          type: 'object'
        },
        forceDark: {
          type: 'boolean'
        },
        pageCache: {
          type: 'boolean'
        },
        cacheDuration: {
          type: 'number'
        }
      }
    },
    subpackages: {
      // TODO 与后面的分包校验融合
      type: 'array',
      items: {
        type: 'object',
        properties: {
          name: {
            type: 'string'
          },
          // TODO 详细的规则校验
          resource: {
            type: 'string'
          }
        },
        // 是否必须?
        required: ['name', 'resource']
      }
    },
    menuBarData: {
      type: 'object',
      properties: {
        menuBar: {
          type: 'boolean'
        },
        menuBarStyle: {
          enum: ['dark', 'light']
        },
        shareTitle: {
          type: 'string'
        },
        shareDescription: {
          type: 'string'
        },
        shareIcon: {
          type: 'string'
        },
        shareCurrentPage: {
          type: 'string'
        },
        shareParams: {
          type: 'string'
        },
        shareUrl: {
          type: 'string'
        }
      }
    },
    deviceTypeList: {
      type: 'array',
      items: {
        type: 'string'
      }
    }
  }
};
var _default = exports.default = ManifestSchema;
