// @aiot-toolkit/aiotpack/lib/utils/ux/UxLoaderUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _generator = require("@aiot-toolkit/generator");
var _parser = require("@aiot-toolkit/parser");
var _UxToTypescript = _interopRequireDefault(require("@aiot-toolkit/parser/lib/ux/translate/vela/UxToTypescript"));
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var parse5 = _interopRequireWildcard(require("aiot-parse5"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _path = _interopRequireDefault(require("path"));
var _QuickAppDocHelp = require("@aiot-toolkit/shared-utils/lib/utils/QuickAppDocHelp");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const BinaryPlugin = require('@aiot-toolkit/parser/lib/ux/translate/vela/protobuf/BinaryPlugin');
const {
  extractFunctions
} = require('@aiot-toolkit/parser/lib/ux/translate/vela/protobuf/protobufControl');
/**
 * UxLoaderUtils
 */
class UxLoaderUtils {
  // 页面组件需要添加ViewModel数据校验和处理代码
  static contenAccess = `\n
  const moduleOwn = exports.default || module.exports
  const accessors = ['public', 'protected', 'private']

  if (moduleOwn.data && accessors.some(function (acc) { return moduleOwn[acc] })) {
    throw new Error('页面VM对象中的属性data不可与"' + accessors.join(',') + '"同时存在，请使用private替换data名称')
  }
  else if (!moduleOwn.data) {
    moduleOwn.data = {}
    moduleOwn._descriptor = {}
    accessors.forEach(function (acc) {
      const accType = typeof moduleOwn[acc]
      if (accType === 'object') {
        moduleOwn.data = Object.assign(moduleOwn.data, moduleOwn[acc])
        for (const name in moduleOwn[acc]) {
          moduleOwn._descriptor[name] = { access: acc }
        }
      }
      else if (accType === 'function') {
        console.warn('页面VM对象中的属性' + acc + '的值不能是函数，请使用对象')
      }
    })
  }`;
  /**
   * 转换单个ux文件为js文件
   * @param file ux文件内容
   * @param context 文件所在项目上下文
   * @param project
   * @returns
   */
  static async compileUxToJavascript(file, context, isAppUx, compilerOption) {
    const logs = [];
    const {
      path: filePath,
      content
    } = file;
    const {
      name,
      ext
    } = _path.default.parse(filePath);
    const fullName = `${name}${ext}`;
    // 转换后新的文件名
    const newFileName = `${name}.js`;
    //判断文本，空文本则结束该文件处理
    if (!content) {
      _sharedUtils.ColorConsole.warn(`The file '${fullName}' under the path '${filePath}' has no content `);
      return {
        files: [],
        logs
      };
    }

    // 区分页面组件和子组件
    const isPageUx = UxLoaderUtils.isPageUx(context, filePath);
    // 配置转换参数
    const options = {
      projectPath: context.projectPath,
      projectType: _sharedUtils.ProjectType.VELA_UX,
      filePath,
      fileType: isAppUx ? 'app' : isPageUx ? 'page' : '',
      content: content.toString(),
      onLog: log => {
        logs.push(log);
      },
      docHelp: new _QuickAppDocHelp.VelaAppDocHelp()
    };
    // 收集图片资源
    const ImageResources = [];
    const collectImageResource = originFilePath => {
      const assetDir = 'dynamicAssets';
      const sourcePath = _path.default.resolve(options.projectPath, compilerOption.sourceRoot);
      const assetPath = _path.default.resolve(sourcePath, assetDir);
      const {
        name,
        ext
      } = _path.default.parse(originFilePath);
      const hashName = _sharedUtils.CommonUtil.calcImageDigest(originFilePath);
      const newFilePath = _path.default.join(assetPath, `${name}-${hashName}${ext}`);
      const relativePath = _path.default.posix.sep + _path.default.relative(sourcePath, newFilePath).split(_path.default.sep).join(_path.default.posix.sep);
      ImageResources.push({
        path: newFilePath,
        content: _fsExtra.default.readFileSync(originFilePath)
      });
      return relativePath;
    };
    // 开始转换
    const globalVar = 'globalVar' in context && context.globalVar || {};
    const parserResult = await new _parser.UxParser(options, compilerOption, globalVar, collectImageResource).parser();
    // 区分app.ux和一般ux
    // app.ux解析结果中加上manifest.json的内容
    const manifestJson = `require('./manifest.json')`;
    function translateStyleFunc() {
      return `
  var $translateStyle$ = function (value) {
    if (typeof value === 'string') {
      return Object.fromEntries(
        value
          .split(';')
          .filter((item) => Boolean(item && item.trim()))
          .map((item) => {
            const matchs = item.match(/([^:]+):(.*)/);
            if (matchs && matchs.length > 2) {
              return [
                matchs[1]
                  .trim()
                  .replace(/-([a-z])/g, (_, match) => match.toUpperCase()),
                matchs[2].trim(),
              ];
            }
            return [];
          })
      );
    }
    return value;
  };
  global.$translateStyle$ = $translateStyle$
  `;
    }
    const integrateFunction = (appImport, appStyleTree, appTemplateTree, appScriptTree) => {
      const scriptCode = appScriptTree.getText();
      const hasScript = Boolean(scriptCode) && scriptCode !== '""';
      // script代码放在第三个位置不能变化，影响更新source map
      return isAppUx ? [`${appImport.join('\n')}`, `var $app_style$ = ${UxLoaderUtils.wrapStyle(appStyleTree, file, compilerOption)}`, hasScript ? `var $app_script$ = ${UxLoaderUtils.wrapScript(false, appScriptTree)}` : '', hasScript ? `$app_script$({}, $app_exports$, $app_require$);` : `$app_exports$.default = {}`, `$app_exports$.default.style = $app_style$;`, `$app_exports$.default.manifest = ${manifestJson}`, translateStyleFunc()] : [`${appImport.join('\n')}`, `var $app_style$ = ${UxLoaderUtils.wrapStyle(appStyleTree, file, compilerOption)}`, hasScript ? `var $app_script$ = ${UxLoaderUtils.wrapScript(isPageUx, appScriptTree)}` : '', `var $app_template$ = ${UxLoaderUtils.wrapTempalte(appTemplateTree, file, compilerOption)}`, `${UxLoaderUtils.getReturnType(isPageUx)} function ($app_exports$) {`, hasScript ? `$app_script$({}, $app_exports$, $app_require$);` : `$app_exports$.default = {}`, `$app_exports$.default.template = $app_template$;`, `$app_exports$.default.style = $app_style$;`, `}`];
    };
    const {
      targetTree,
      mapList,
      sourceMap
    } = await new _UxToTypescript.default(options, integrateFunction, compilerOption, context).translate(parserResult.ast, []);
    const {
      code
    } = new _generator.TypescriptGenerator().generate({
      sourceFilePath: fullName,
      targetFilePath: newFileName,
      ast: targetTree,
      mapList
    });
    // 返回转换后的文件内容以及map内容
    return {
      files: [{
        path: filePath,
        content: code
      }, {
        path: _sharedUtils.FileUtil.updateFileName(filePath, `${name}${ext}.map`),
        content: sourceMap.toString()
      }, ...ImageResources],
      logs
    };
  }
  /**
   * 转换app.ux文件为js文件
   * @param file
   * @param resultFiles
   * @param context
   */
  static async compileAppUxToJavascript(file, context, compileOption) {
    const {
      path: filePath,
      content
    } = file;
    let appContent = content;
    // 校验是否为src/app.ux
    let isAppUx = UxLoaderUtils.isAppUx(filePath);
    if (isAppUx) {
      appContent = UxLoaderUtils.getAppUxContent(content);
    }
    // 解析script和style两部分
    const compileResult = await UxLoaderUtils.compileUxToJavascript({
      path: filePath,
      content: appContent
    }, context, isAppUx, compileOption);
    return compileResult;
  }
  /**
   * 判断是否为src/app.ux
   * @param filePath
   * @returns
   */
  static isAppUx(filePath) {
    const parentDirName = _path.default.basename(_path.default.dirname(filePath));
    const fileName = _path.default.basename(filePath);
    return parentDirName === 'src' && fileName === 'app.ux';
  }
  /**
   * 判断是否为页面组件
   * @param context
   * @param filePath
   * @returns
   */
  static isPageUx(context, filePath) {
    return Boolean(context.entries?.includes(filePath));
  }
  /**
   * 获取app.ux的内容，且只返回script和style内容
   * @param fileContent
   * @returns
   */
  static getAppUxContent(fileContent) {
    const appUxAst = parse5.parseFragment(fileContent, {
      scriptingEnabled: false,
      sourceCodeLocationInfo: true
    });
    //   创建一个临时父节点对象
    const parentNode = parse5.defaultTreeAdapter.createElement('div', parse5.html.NS.HTML, []);
    for (const childNode of appUxAst.childNodes) {
      const {
        nodeName
      } = childNode;
      if (nodeName !== 'template') {
        parse5.defaultTreeAdapter.appendChild(parentNode, childNode);
      }
    }
    return parse5.serialize(parentNode);
  }
  static addTemplateProtobuf(file, code) {
    const funToString = fun => {
      const {
        aiotfor,
        aiotcb
      } = fun;
      // 如果aiotfor存在，需要把函数体替换为aiotfor.funcArr
      if (aiotfor) {
        const {
          funcArr = [],
          key,
          value
        } = aiotfor;
        const forFun = `function(${key || '$idx'}, ${value || '$item'}) {
          return [
            ${funcArr.map(item => funToString(item)).join(',\r\n')}
          ]
        }`;
        return forFun;
      }
      if (aiotcb) {
        const {
          funcArr = []
        } = aiotcb;
        const cbFun = `function($data){
          return [
            ${funcArr.map(item => funToString(item)).join(',\r\n')}
          ]
        }`;
        return cbFun;
      }
      return fun.toString();
    };
    const stringifyObjectWithFunctions = obj => {
      const processValue = value => {
        if (typeof value === 'function') {
          return value.toString();
        } else if (typeof value === 'object' && value !== null) {
          return stringifyObjectWithFunctions(value);
        } else {
          return JSON.stringify(value, undefined, 2);
        }
      };
      const entries = Object.entries(obj).map(_ref => {
        let [key, value] = _ref;
        return `"${key}": ${processValue(value)}`;
      });
      return `{
        ${entries.join(',\r\n ')}
      }`;
    };
    const {
      path
    } = file;
    const aiot = extractFunctions(code);
    const {
      funcArr = [],
      optArr
    } = aiot;
    const templateResult = BinaryPlugin.addTemplate(path, path, code);
    BinaryPlugin.addTagName(path, path);
    const result = `function (vm) {
      const _vm_ = vm || this
      return aiot.__cv__(
        "${templateResult.name}",
        ${templateResult.index},
        {
          __vm__:_vm_,
          __func__:[
            ${funcArr.map(item => funToString(item)).join(',\r\n')}
          ],
          __optsArr__:[
            ${optArr.map(item => stringifyObjectWithFunctions(item))},
          ]
        }
      )
    }`;
    return result;
  }

  /**
   * 给template增加外层包裹内容
   * @param templateTree
   * @returns
   */
  static wrapTempalte(templateTree, file, compilerOption) {
    const result = `function (vm) {
      const _vm_ = vm || this
      return ${templateTree.getFullText()}
    }`;
    if (compilerOption.enableProtobuf) {
      return this.addTemplateProtobuf(file, result);
    }
    return result;
  }
  static wrapStyle(styleNodes, file, compilerOption) {
    const code = JSON.stringify(styleNodes);
    if (compilerOption.enableProtobuf) {
      if (styleNodes?.length) {
        const result = BinaryPlugin.addStyle(file.path, file.path, code);
        return `['${result.name}', ${result.index}]`;
      }
    }
    return code;
  }

  /**
   * 给script增加外层包裹内容
   * @param isPageUx
   * @param appScriptTree
   * @returns
   */
  static wrapScript(isPageUx, appScriptTree) {
    if (isPageUx) {
      // 页面组件添加ViewModel处理代码
      appScriptTree.addStatements(UxLoaderUtils.contenAccess);
    }
    return `function __scriptModule__(module, exports, $app_require$) {\t${appScriptTree.getFullText()}\n}`;
  }
  static getReturnType(isPageUx) {
    return isPageUx ? `$app_exports$['entry'] =` : 'module.exports = ';
  }
}
var _default = exports.default = UxLoaderUtils;
