// aiot-toolkit/lib/starter/AndroidUxStart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _aiotpack = require("@aiot-toolkit/aiotpack");
var _commander = require("@aiot-toolkit/commander");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _http = _interopRequireDefault(require("http"));
var _path = _interopRequireDefault(require("path"));
var _AndroidUxBuilder = _interopRequireDefault(require("../builder/AndroidUxBuilder"));
var _IStarter = _interopRequireDefault(require("./IStarter"));
var _PackageRouter = _interopRequireDefault(require("./androidRouter/PackageRouter"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * AndroidUxStart
 *
 * 1. 启动本机开发者http服务器，并提供二维码，以下载 rpk
 * 2. 打包 rpk，并监听文件变化，重新打包
 */
class AndroidUxStart extends _IStarter.default {
  builder = (() => new _AndroidUxBuilder.default())();
  projectPath = '';
  params = (() => [...this.builder.params])();
  get waiter() {
    return new _commander.PersistentCommand({
      description: 'you can press follow keys to do something',
      options: [{
        key: 'q',
        description: 'show qrcode',
        action: () => {
          this.showAddress();
        }
      }, {
        key: '?',
        description: 'show waiter desc',
        action: () => {
          this.waiter.clearLog();
          this.waiter.describe();
        }
      }]
    });
  }

  /**
   * 启动
   * 1. build 项目
   * 2. build 成功，则创建http服务器
   */
  async start(projectPath, options) {
    this.projectPath = projectPath;
    const buildOption = {
      ..._aiotpack.JavascriptDefaultCompileOption,
      ...options
    };
    await this.build(projectPath, buildOption);
    await this.createServer(buildOption);
  }

  /**
   * 创建服务器
   *
   * 1. 使用 koa 创建服务器
   * 2. 显示服务器地址和二维码
   * @returns
   */
  async createServer(options) {
    return new Promise(async resolve => {
      const routeConfigList = [new _PackageRouter.default({
        projectPath: this.projectPath,
        options
      })];
      const data = await _sharedUtils.NetworkUtil.createHttpServer({
        wantPort: options.server.port,
        routeConfigList,
        staticFolder: _path.default.join(__dirname, './androidRouter/h5'),
        events: {
          onSuccess: () => {
            this.server = data.server;
            this.port = data.port;
            this.showAddress();
            resolve(data);
          }
        }
      });
    });
  }
  showAddress() {
    const ip = _sharedUtils.NetworkUtil.getIPv4IPAddress();
    const {
      port
    } = this;
    if (!ip) {
      const localUrl = `http://localhost:${port}`;
      _sharedUtils.ColorConsole.warn(`devServer`, localUrl);
    } else {
      const lanUrl = `http://${ip}:${port}`;
      // 显示二维码
      _sharedUtils.ColorConsole.info(`devServer`, lanUrl);
      _sharedUtils.CommonUtil.outputQRCodeOnTerminal(lanUrl);
    }
  }
  async build(projectPath, options) {
    this.builder.events = {
      onBuildSuccess: data => {
        _sharedUtils.ColorConsole.info(`build time: ${data.costTime}ms`);
        this.noticeDeviceListUpdate(projectPath, options);
      }
    };
    await this.builder.build(projectPath, {
      ...options,
      watch: true
    });
  }
  dispose() {
    this.server?.close();
  }

  /**
   * 通知已知的设备，应用有更新
   * @param projectPath
   * @param options
   */
  noticeDeviceListUpdate(projectPath, options) {
    const {
      clientRecordPath
    } = options;
    const json = _fsExtra.default.readJSONSync(clientRecordPath, {
      throws: false
    });
    const deviceList = json?.records?.[projectPath];
    if (deviceList) {
      deviceList.forEach(item => {
        const {
          ip,
          port
        } = item;
        if (!ip) {
          return;
        }
        const url = `http://${ip}:${port}/update`;
        const requestOption = {
          host: ip,
          port,
          path: '/update',
          timeout: 3000
        };
        _http.default.request(requestOption, () => {
          _sharedUtils.ColorConsole.success(`Notify the phone to update the rpk file success`, {
            word: url
          });
        }).on('error', error => {
          _sharedUtils.ColorConsole.warn(`Notify the phone to update the rpk file`, {
            word: 'error: '
          }, {
            word: error.message
          });
        }).on('timeout', () => {
          _sharedUtils.ColorConsole.warn(`Notify the phone to update the rpk file`, {
            word: 'timeout: '
          }, {
            word: url
          });
        });
      });
    }
  }
}
var _default = exports.default = AndroidUxStart;
