// aiot-toolkit/lib/starter/VelaUxStarter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _emulator = require("@aiot-toolkit/emulator");
var _constants = require("@aiot-toolkit/emulator/lib/emulatorutil/constants");
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _prompts = require("@inquirer/prompts");
var _path = _interopRequireDefault(require("path"));
var _portfinder = _interopRequireDefault(require("portfinder"));
var _VelaUxBuilder = _interopRequireDefault(require("../builder/VelaUxBuilder"));
var _VelaAvdUtils = _interopRequireDefault(require("../utils/VelaAvdUtils"));
var _IStarter = _interopRequireDefault(require("./IStarter"));
var _fsExtra = _interopRequireDefault(require("fs-extra"));
var _aiotpack = require("@aiot-toolkit/aiotpack");
var _events = _interopRequireDefault(require("events"));
var _FileLaneTriggerType = _interopRequireDefault(require("file-lane/lib/enum/FileLaneTriggerType"));
var _IVelaUxBuilderOption = require("../interface/IVelaUxBuilderOption");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * VelaUxStarter
 * ux快应用启动器
 */
class VelaUxStarter extends _IStarter.default {
  constructor(name, description) {
    super(name, description);
    this.name = name;
    this.description = description;
    this.event = new _events.default();
    this.builder = new _VelaUxBuilder.default({
      onBuildSuccess: data => {
        _sharedUtils.ColorConsole.success({
          word: 'build success: '
        }, `${data.costTime}ms`);
        this.lastRpk = data.info?.rpk;
        this.event.emit('buildSuccess', data);
      }
    });
  }
  params = (() => [{
    name: 'disableNSH',
    description: 'disable goldfish NSH terminal',
    defaultValue: false,
    type: 'confirm'
  }, {
    name: 'watch',
    description: 'recompile project while file changes',
    defaultValue: false,
    type: 'confirm'
  }, {
    name: 'openVNC',
    description: 'open vnc',
    defaultValue: false,
    type: 'confirm'
  }, {
    type: 'confirm',
    name: 'open-nuttx',
    description: 'deprecated',
    deprecated: 'please remove it now',
    deprecatedVersion: '2.1.0'
  }, ..._IVelaUxBuilderOption.velaUxBuilderParams])();
  async start(projectPath, options) {
    const check = await _VelaAvdUtils.default.checkEmulatorEnv();
    if (!check) {
      const errorMsg = 'Your Vela emulator is not ready. Please use the `npx aiot initEmulatorEnv` command to initialize the emulator environment.';
      _sharedUtils.ColorConsole.throw(errorMsg);
      throw new Error(errorMsg);
    }
    // 获取已经创建的模拟器列表
    const vvdList = _VelaAvdUtils.default.vvdManager.getVvdList();
    let vvdName;
    if (vvdList.length === 0) {
      _sharedUtils.ColorConsole.error('no vela emulator available, please create it first.');
      _VelaAvdUtils.default.createVelaVvdByInquire();
      return;
    } else if (vvdList.length === 1) {
      const needtoRun = await (0, _prompts.confirm)({
        message: 'there is only one emulator, need to Run?',
        default: true
      });
      if (!needtoRun) {
        _VelaAvdUtils.default.createVelaVvdByInquire();
        return;
      }
      vvdName = vvdList[0].name;
      _sharedUtils.ColorConsole.warn(`Start run by :${vvdName} `);
    } else {
      vvdName = await (0, _prompts.select)({
        message: 'name of the avd to start',
        choices: vvdList.map(item => {
          return {
            value: item.name
          };
        })
      });
    }
    // 检查选择的模拟器是否符合要求
    if (vvdName && !this.isAvailableEmulator(vvdName)) {
      _sharedUtils.ColorConsole.throw(`this emulator is unavailable, please create a new emulator.`);
      return;
    }

    // build
    const compilerOption = this.builder.getCompilerOption(projectPath, options);
    // start build
    await this.builder.build(projectPath, options);

    // 设置 debugPort
    let debugPort;
    if (options.devtool) {
      debugPort = await _portfinder.default.getPortPromise({
        port: _constants.defaultDebugPort
      });
    }

    // start emulator
    const startRes = await _VelaAvdUtils.default.vvdManager.startVvd({
      vvdName,
      origin: _emulator.IStartOrigin.Terminal,
      debugPort
    });
    this.emulatorInstance = startRes.emulatorInstance;
    if (!this.emulatorInstance) {
      throw new Error('emulatorInstance is undefined');
    }
    const projectInfo = _aiotpack.UxFileUtils.getManifestInfo(projectPath, compilerOption.sourceRoot);
    if (this.lastRpk && _fsExtra.default.existsSync(this.lastRpk)) {
      const targetPath = await this.emulatorInstance?.pushRpk(this.lastRpk, projectInfo.package);
      await this.emulatorInstance?.install(targetPath, {
        packageName: projectInfo.package
      });
      await this.emulatorInstance?.startApp(projectInfo.package, options.devtool);
    }
    this.event.on('buildSuccess', async data => {
      if (data.info?.diffFile) {
        const targetPath = `/data/tmp/${projectInfo.package}`;
        await this.emulatorInstance?.push(data.info.diffFile, targetPath);
        await this.emulatorInstance?.unzip(targetPath, `${this.emulatorInstance.appDir}/${projectInfo.package}`);
        await this.emulatorInstance?.reloadApp(projectInfo.package);
      } else if (data.info?.trigger === _FileLaneTriggerType.default.START && data.info.rpk) {
        const targetPath = await this.emulatorInstance?.pushRpk(data.info.rpk, projectInfo.package);
        await this.emulatorInstance?.install(targetPath, {
          packageName: projectInfo.package
        });
        await this.emulatorInstance?.startApp(projectInfo.package, options.devtool);
      }
    });
  }

  /**
   * 检查选择的模拟器在当前环境下是否可以使用
   * 下面的几种情况会导致模拟器不可用：1. 模拟器绑定的Vela镜像不存在；2. Vela4.0缺少coredump.core或者vela_data.bin文件
   * @param avdName 模拟器名称
   * @returns {boolean}
   */
  isAvailableEmulator(avdName) {
    const {
      imageDir
    } = _VelaAvdUtils.default.vvdManager.getVvdInfo(avdName);
    // 没有avdImagePath，即对应的configIni里没有 image.sysdir.1 字段
    // 有avdImagePath，但是这个目录下没有nuttx
    if (!imageDir || !_fsExtra.default.existsSync(_path.default.resolve(imageDir, 'nuttx'))) {
      return false;
    }
    return true;
  }
}
var _default = exports.default = VelaUxStarter;
