// @aiot-toolkit/parser/lib/ux/config/CompressConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cssAttrMap = void 0;
// CSS属性名全写-缩写对照表
const cssAttrMap = exports.cssAttrMap = {
  // boxModel
  width: 'w',
  height: 'h',
  paddingLeft: 'pL',
  paddingRight: 'pR',
  paddingTop: 'pT',
  paddingBottom: 'pB',
  marginLeft: 'mL',
  marginRight: 'mR',
  marginTop: 'mT',
  marginBottom: 'mB',
  borderLeftWidth: 'bLW',
  borderTopWidth: 'bTW',
  borderRightWidth: 'bRW',
  borderBottomWidth: 'bBW',
  borderLeftColor: 'bLC',
  borderTopColor: 'bTC',
  borderRightColor: 'bRC',
  borderBottomColor: 'bBC',
  borderStyle: 'bS',
  borderRadius: 'bR',
  borderBottomLeftRadius: 'bBLR',
  borderBottomRightRadius: 'bBRR',
  borderTopLeftRadius: 'bTLR',
  borderTopRightRadius: 'bTRR',
  indicatorSize: 'iS',
  // flexbox
  flex: 'f',
  flexGrow: 'fG',
  flexShrink: 'fS',
  flexBasis: 'fB',
  flexDirection: 'fD',
  flexWrap: 'fW',
  justifyContent: 'jC',
  alignItems: 'aI',
  alignContent: 'aC',
  alignSelf: 'aS',
  // position
  position: 'p',
  top: 't',
  bottom: 'b',
  left: 'l',
  right: 'r',
  zIndex: 'zI',
  // common
  opacity: 'o',
  background: 'bg',
  backgroundColor: 'bgC',
  backgroundImage: 'bgI',
  backgroundSize: 'bgS',
  backgroundRepeat: 'bgR',
  backgroundPosition: 'bgP',
  display: 'd',
  visibility: 'v',
  // text
  lines: 'ls',
  color: 'c',
  fontSize: 'foS',
  fontStyle: 'fSt',
  fontWeight: 'foW',
  textDecoration: 'tD',
  textAlign: 'tA',
  lineHeight: 'lH',
  textOverflow: 'tO',
  // animation
  transform: 'ts',
  transformOrigin: 'tsO',
  animationName: 'aN',
  animationDuration: 'aD',
  animationTimingFunction: 'aTF',
  animationDelay: 'aDe',
  animationIterationCount: 'aIC',
  animationFillMode: 'aFM',
  // custom
  placeholderColor: 'pC',
  selectedColor: 'sC',
  textColor: 'tC',
  timeColor: 'tiC',
  textHighlightColor: 'tHC',
  strokeWidth: 'sW',
  progressColor: 'prC',
  indicatorColor: 'iC',
  indicatorSelectedColor: 'iSC',
  slideWidth: 'slW',
  slideMargin: 'sM',
  resizeMode: 'rM',
  columns: 'col',
  columnSpan: 'cS',
  maskColor: 'mC',
  // custom style
  starBackground: 'sB',
  starForeground: 'sF',
  starSecondary: 'sS'
};
