// @aiot-toolkit/aiotpack/lib/compiler/javascript/vela/utils/webpackLoader/addColSourceMap.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _parser = require("@aiot-toolkit/parser");
async function addColSourceMap(source, map) {
  const callback = this.async();
  if (!map) {
    callback(null, source, map);
  }
  let {
    consumer,
    tempGenerator
  } = await _parser.SourceMapUtil.createConsumerAndGenerator(JSON.stringify(map));
  let recordLine = -1;
  consumer.eachMapping(mapping => {
    if (mapping.source === consumer.sources[consumer.sources.length - 1]) {
      if (recordLine < mapping.generatedLine && (mapping.generatedColumn !== 0 || mapping.originalColumn !== 0)) {
        tempGenerator.addMapping({
          generated: {
            line: mapping.generatedLine,
            column: 0
          },
          original: {
            line: mapping.originalLine,
            column: 0
          },
          source: mapping.source,
          name: ''
        });
      }
      recordLine = mapping.generatedLine;
    }
    tempGenerator.addMapping({
      generated: {
        line: mapping.generatedLine,
        column: mapping.generatedColumn
      },
      original: {
        line: mapping.originalLine,
        column: mapping.originalColumn
      },
      source: mapping.source,
      name: mapping.name
    });
  });
  callback(null, source, JSON.parse(tempGenerator.toString()));
}
var _default = exports.default = addColSourceMap;
