// @aiot-toolkit/parser/lib/ux/translate/vela/protobuf/BinaryPlugin.js
const Path = require('path')
const { createBin } = require('./protobufControl')

const contentType = {
  style: 'style',
  template: 'template',
  tagName: 'tagName'
}

const ext = '.bin'

class BinaryPlugin {
  /**
   * 所有要创建二进制文件的数据
   *
   * @type {{[binFileName:string]:{
   *    sourceFileName:string,
   *    tagName?:string,
   *    style?:string
   *    template?:string
   * }[]}}
   */
  static binaryData = {}

  static config = {}

  static createBinFiles() {
    const keys = Object.keys(BinaryPlugin.binaryData)
    if (keys.length > 0) {
      keys.forEach((binPath) => {
        // 写入文件
        // 1. 获取style template内容
        // 2. 获取bin文件名= build目录 + 文件名相对src的路径
        // 3. 创建buffer数据
        // 4. 写入文件
        const data = BinaryPlugin.binaryData[binPath]
        const styleList = []
        const templateList = []
        const tagList = []

        data.forEach((item) => {
          tagList.push(item.tagName || '')
          styleList.push(item.style || '[]')
          templateList.push(item.template || '')
        })

        const ext = Path.extname(binPath)

        // 如果不是hml文件，生成style.bin
        if (ext !== '.hml') {
          const cssBinPath = BinaryPlugin.getBinPath(binPath, 'style')
          createBin(
            cssBinPath,
            new Array(tagList.length).fill(''),
            styleList,
            new Array(templateList.length).fill('')
          )
        }

        // 如果不是css类型的文件，生成template.bin
        if (!['.css', '.less', '.scss', '.sass'].includes(ext)) {
          const templateBinPath = BinaryPlugin.getBinPath(binPath, 'template')
          createBin(templateBinPath, tagList, new Array(styleList.length).fill('[]'), templateList)
        }
      })
    }
  }

  static reset() {
    BinaryPlugin.binaryData = {}
  }

  /**
   *
   * @private
   * @param {string} binPath
   * @param {'style'|'template'} type
   */
  static getBinPath(binPath, type) {
    const srcPath = Path.join(this.config.projectPath, this.config.source)
    const relativePath = Path.relative(srcPath, binPath)
    const parse = Path.parse(relativePath)

    const result = Path.join(
      this.config.outputProjectPath,
      this.config.output,
      Path.dirname(relativePath),
      `${parse.name}.${type}${ext}`
    )

    return result
  }

  /**
   *
   * @private
   * @param {string} binFileName 要生成的bin文件路径
   * @param {string} sourceFileName 源码文件路径
   * @param {string} type 添加内容的类型, 参考contentType
   * @param {string} content
   *
   * @returns {{index:number, name:string}}
   */
  static addContent(binFileName, sourceFileName, type, content, pagePath = '') {
    const binaryData = BinaryPlugin.binaryData
    // 1. 如果没有对应的bin数据，则创建

    if (!binaryData[binFileName]) {
      binaryData[binFileName] = []
    }

    const array = binaryData[binFileName]
    let item
    // 2. 通过sourceFileName寻找是否已存在对应的数据，如果存在，记录位置并直接使用；不存在，则新建
    const index = array.findIndex((item) => item.sourceFileName === sourceFileName)
    if (index >= 0) {
      item = array[index]
      item[type] = content
    } else {
      item = {
        sourceFileName,
        [type]: content
      }
      array.push(item)
    }

    const binPath = BinaryPlugin.getBinPath(binFileName, type)
    let relativePath = Path.relative(
      Path.resolve(this.config.outputProjectPath, this.config.output),
      binPath
    )

    relativePath = relativePath.replace(/\\/g, '/')

    return {
      index: index >= 0 ? index : array.length - 1,
      name: relativePath
    }
  }

  static addTagName(binFileName, sourceFileName, content, pagePath = '') {
    return BinaryPlugin.addContent(
      binFileName,
      sourceFileName,
      contentType.tagName,
      content,
      pagePath
    )
  }

  static addTemplate(binFileName, sourceFileName, content, pagePath = '') {
    return BinaryPlugin.addContent(
      binFileName,
      sourceFileName,
      contentType.template,
      content,
      pagePath
    )
  }

  static addStyle(binFileName, sourceFileName, content, pagePath = '') {
    return BinaryPlugin.addContent(
      binFileName,
      sourceFileName,
      contentType.style,
      content,
      pagePath
    )
  }
}

module.exports = BinaryPlugin
