// @aiot-toolkit/parser/lib/ux/translate/vela/protobuf/protobuf.js
const vdom_pb = require('./vdom_pb')
/* eslint-disable */

class AIOT {
  constructor(aiotf) {
    this.funcArr = []
    this.chld = []
    this.idCount = -1
    this.expCell = new vdom_pb.ExportCell()
    this.aiotf = aiotf
    this.optArr = []
  }

  init(expName) {
    this.expCell.setTagname(expName)
  }

  parseObj(opts) {
    let styVaA = new vdom_pb.StyleValueArr()

    for (let key in opts) {
      let sKey = new vdom_pb.StyleKey()
      let onesvl = new vdom_pb.OneStyleValue()

      sKey.setValue(key)
      onesvl.setKey(sKey)

      let typ = Object.prototype.toString.call(opts[key])
      if (typ === '[object Number]') {
        // TODO:double
        onesvl.setDoublevl(opts[key])
      } else if (typ === '[object String]') {
        // console.log(`String ${opts[key]}`)
        onesvl.setStringvl(opts[key])
      } else if (typ === '[object Function]') {
        if (this.aiotf == undefined) {
          this.funcArr.push(opts[key])
          onesvl.setFuncid(this.funcArr.length - 1)
        } else {
          this.aiotf.funcArr.push(opts[key])
          onesvl.setFuncid(this.aiotf.funcArr.length - 1)
        }
      } else if (typ === '[object Object]') {
        // console.log(`object ${opts[key]}`)
        let styVaA_ = new vdom_pb.StyleValueArr()
        styVaA_ = this.parseObj(opts[key])
        onesvl.setStylevlarr(styVaA_)
      } else if (typ === '[object Array]') {
        //TODO:default string arr
        // console.log(`Array ${opts[key]}`)
        let strarr = new vdom_pb.StringArr()
        strarr.setStringvlarrList(opts[key])
        onesvl.setStringarr(strarr)
      } else if (typ === '[object Boolean]') {
        onesvl.setBoolvl(opts[key])
      }
      styVaA.addVlarr(onesvl)
    }
    return styVaA
  }

  __cdc__(tagName, opts, childsArr) {
    return this.__ce__(tagName, opts, childsArr, '__cdc__')
  }

  __ce__(tagName, opts, childsArr, type = '__ce__') {
    let tempOne = new vdom_pb.TemplateOne()
    let ce = new vdom_pb.TemplateUnit()

    ce.setType(type)
    ce.setTag(tagName)
    if (Array.isArray(childsArr) && childsArr.length > 0) {
      let arr = []
      for (let key in childsArr) {
        arr.push(childsArr[key].id_)
      }
      tempOne.setChldsidList(arr)
    }
    tempOne.setId(++this.idCount)

    let styVaA_ = this.parseObj(opts.__opts__)
    ce.setOptsarrList(styVaA_.getVlarrList())

    tempOne.setCv(ce)
    this.expCell.addTemplatearr(tempOne)

    return { id_: this.idCount, aiot_: this }
  }
  __ci__(opts, renderFunc) {
    let tempOne = new vdom_pb.TemplateOne()
    let ci = new vdom_pb.TemplateUnit()

    ci.setType('__ci__')
    tempOne.setId(++this.idCount)

    let styVaA_ = this.parseObj(opts.__opts__)
    ci.setOptsarrList(styVaA_.getVlarrList())

    //   function(aiot) {

    let aiotif = new AIOT(this)
    aiotif.funcArr = this.funcArr
    aiotif.optArr = this.optArr

    let renderArr = renderFunc(aiotif)
    let typ = Object.prototype.toString.call(renderArr)
    if (typ === '[object Object]') {
      ci.setTemplaterenderList(renderArr.aiot_.expCell.getTemplatearrList())
    } else if (typ === '[object Array]') {
      for (let key in renderArr) {
        ci.setTemplaterenderList(renderArr[key].aiot_.expCell.getTemplatearrList())
        break
      }
    }

    this.funcArr = aiotif.aiotf.funcArr

    tempOne.setCv(ci)
    this.expCell.addTemplatearr(tempOne)

    return { id_: this.idCount, aiot_: this }
  }
  __cf__(opts, renderFunc) {
    let tempOne = new vdom_pb.TemplateOne()
    let cf = new vdom_pb.TemplateUnit()

    cf.setType('__cf__')
    tempOne.setId(++this.idCount)

    let styVaA_ = this.parseObj(opts.__opts__)
    cf.setOptsarrList(styVaA_.getVlarrList())

    let aiotfor = new AIOT()
    aiotfor.optArr = this.optArr
    let key_, value_

    let renderArr = renderFunc(key_, value_, aiotfor)
    let typ = Object.prototype.toString.call(renderArr)
    if (typ === '[object Object]') {
      cf.setTemplaterenderList(renderArr.aiot_.expCell.getTemplatearrList())
    } else if (typ === '[object Array]') {
      for (let key in renderArr) {
        cf.setTemplaterenderList(renderArr[key].aiot_.expCell.getTemplatearrList())
        break
      }
    }

    if (aiotfor.funcArr.length > 0) {
      const fun = function($idx, $item) {
        return aiotfor.aiotf.funcArr
      }
      fun.aiotfor = aiotfor
      fun.aiotfor.key = opts.__opts__['key']
      fun.aiotfor.value = opts.__opts__['value']
      this.funcArr.push(fun)
      cf.setFuncrender(this.funcArr.length - 1)
    } else {
      cf.setFuncrender(-1)
    }

    tempOne.setCv(cf)
    this.expCell.addTemplatearr(tempOne)

    return { id_: this.idCount, aiot_: this }
  }
  __cc__(tagName, opts) {
    let tempOne = new vdom_pb.TemplateOne()
    let cc = new vdom_pb.TemplateUnit()

    cc.setType('__cc__')
    cc.setTag(tagName)
    tempOne.setId(++this.idCount)

    // 设置__opts__在数组中得索引
    // if (Object.keys(opts.__opts__).length > 0) {
    //   this.optArr.push(opts.__opts__) //save to js
    //   cc.setOptsid(this.optArr.length - 1)
    // } else {
    //   cc.setOptsid(-1)
    // }

    let styVaA_ = this.parseObj(opts.__opts__)
    cc.setOptsarrList(styVaA_.getVlarrList())

    tempOne.setCv(cc)
    this.expCell.addTemplatearr(tempOne)

    return { id_: this.idCount, aiot_: this }
  }
  __cb__(opts, renderFunc) {
    let tempOne = new vdom_pb.TemplateOne()
    let cb = new vdom_pb.TemplateUnit()

    cb.setType('__cb__')
    tempOne.setId(++this.idCount)

    let styVaA_ = this.parseObj(opts.__opts__)
    cb.setOptsarrList(styVaA_.getVlarrList())

    let aiotcb = new AIOT()
    aiotcb.optArr = this.optArr

    let $data
    let renderArr = renderFunc($data, aiotcb)
    let typ = Object.prototype.toString.call(renderArr)
    if (typ === '[object Object]') {
      cb.setTemplaterenderList(renderArr.aiot_.expCell.getTemplatearrList())
    } else if (typ === '[object Array]') {
      for (let key in renderArr) {
        cb.setTemplaterenderList(renderArr[key].aiot_.expCell.getTemplatearrList())
        break
      }
    }

    if (aiotcb.funcArr.length > 0) {
      const fun = function($data) {
        return aiotcb.aiotf.funcArr
      }
      fun.aiotcb = aiotcb
      this.funcArr.push(fun)
      cb.setFuncrender(this.funcArr.length - 1)
    } else {
      cb.setFuncrender(-1)
    }

    // this.funcArr = aiotcb.aiotf.funcArr

    tempOne.setCv(cb)
    this.expCell.addTemplatearr(tempOne)

    return { id_: this.idCount, aiot_: this }
  }
}

module.exports = AIOT
