// @aiot-toolkit/parser/lib/ux/translate/vela/protobuf/vdom_pb.js
/**
 * @fileoverview
 * @enhanceable
 * @suppress {messageConventions} JS Compiler reports an error if a variable or
 *     field starts with 'MSG_' and isn't a translatable message.
 * @public
 */
// GENERATED CODE -- DO NOT EDIT!

/* eslint-disable */

var jspb = require('google-protobuf')
var goog = jspb
var global = Function('return this')()

goog.exportSymbol('proto.CssCell', null, global)
goog.exportSymbol('proto.ExportArr', null, global)
goog.exportSymbol('proto.ExportCell', null, global)
goog.exportSymbol('proto.OneStyleValue', null, global)
goog.exportSymbol('proto.SelectorKey', null, global)
goog.exportSymbol('proto.StringArr', null, global)
goog.exportSymbol('proto.StyleKey', null, global)
goog.exportSymbol('proto.StyleValueArr', null, global)
goog.exportSymbol('proto.TemplateOne', null, global)
goog.exportSymbol('proto.TemplateUnit', null, global)
goog.exportSymbol('proto.Version', null, global)

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.StringArr = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.StringArr.repeatedFields_, null)
}
goog.inherits(proto.StringArr, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.StringArr.displayName = 'proto.StringArr'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.StringArr.repeatedFields_ = [1]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.StringArr.prototype.toObject = function(opt_includeInstance) {
    return proto.StringArr.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.StringArr} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.StringArr.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        stringvlarrList: jspb.Message.getRepeatedField(msg, 1)
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.StringArr}
 */
proto.StringArr.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.StringArr()
  return proto.StringArr.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.StringArr} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.StringArr}
 */
proto.StringArr.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {string} */ (reader.readString())
        msg.addStringvlarr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.StringArr.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.StringArr.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.StringArr} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.StringArr.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getStringvlarrList()
  if (f.length > 0) {
    writer.writeRepeatedString(1, f)
  }
}

/**
 * repeated string stringVlArr = 1;
 * @return {!Array<string>}
 */
proto.StringArr.prototype.getStringvlarrList = function() {
  return /** @type {!Array<string>} */ (jspb.Message.getRepeatedField(this, 1))
}

/** @param {!Array<string>} value */
proto.StringArr.prototype.setStringvlarrList = function(value) {
  jspb.Message.setField(this, 1, value || [])
}

/**
 * @param {!string} value
 * @param {number=} opt_index
 */
proto.StringArr.prototype.addStringvlarr = function(value, opt_index) {
  jspb.Message.addToRepeatedField(this, 1, value, opt_index)
}

proto.StringArr.prototype.clearStringvlarrList = function() {
  this.setStringvlarrList([])
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.StyleKey = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null)
}
goog.inherits(proto.StyleKey, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.StyleKey.displayName = 'proto.StyleKey'
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.StyleKey.prototype.toObject = function(opt_includeInstance) {
    return proto.StyleKey.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.StyleKey} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.StyleKey.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        index: jspb.Message.getFieldWithDefault(msg, 1, 0),
        value: jspb.Message.getFieldWithDefault(msg, 2, '')
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.StyleKey}
 */
proto.StyleKey.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.StyleKey()
  return proto.StyleKey.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.StyleKey} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.StyleKey}
 */
proto.StyleKey.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setIndex(value)
        break
      case 2:
        var value = /** @type {string} */ (reader.readString())
        msg.setValue(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.StyleKey.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.StyleKey.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.StyleKey} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.StyleKey.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getIndex()
  if (f !== 0) {
    writer.writeInt32(1, f)
  }
  f = message.getValue()
  if (f.length > 0) {
    writer.writeString(2, f)
  }
}

/**
 * optional int32 index = 1;
 * @return {number}
 */
proto.StyleKey.prototype.getIndex = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0))
}

/** @param {number} value */
proto.StyleKey.prototype.setIndex = function(value) {
  jspb.Message.setProto3IntField(this, 1, value)
}

/**
 * optional string value = 2;
 * @return {string}
 */
proto.StyleKey.prototype.getValue = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ''))
}

/** @param {string} value */
proto.StyleKey.prototype.setValue = function(value) {
  jspb.Message.setProto3StringField(this, 2, value)
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.OneStyleValue = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, proto.OneStyleValue.oneofGroups_)
}
goog.inherits(proto.OneStyleValue, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.OneStyleValue.displayName = 'proto.OneStyleValue'
}
/**
 * Oneof group definitions for this message. Each group defines the field
 * numbers belonging to that group. When of these fields' value is set, all
 * other fields in the group are cleared. During deserialization, if multiple
 * fields are encountered for a group, only the last value seen will be kept.
 * @private {!Array<!Array<number>>}
 * @const
 */
proto.OneStyleValue.oneofGroups_ = [[2, 3, 4, 5, 6, 7, 8]]

/**
 * @enum {number}
 */
proto.OneStyleValue.ValueCase = {
  VALUE_NOT_SET: 0,
  INT32VL: 2,
  BOOLVL: 3,
  FUNCID: 4,
  DOUBLEVL: 5,
  STRINGVL: 6,
  STRINGARR: 7,
  STYLEVLARR: 8
}

/**
 * @return {proto.OneStyleValue.ValueCase}
 */
proto.OneStyleValue.prototype.getValueCase = function() {
  return /** @type {proto.OneStyleValue.ValueCase} */ (jspb.Message.computeOneofCase(
    this,
    proto.OneStyleValue.oneofGroups_[0]
  ))
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.OneStyleValue.prototype.toObject = function(opt_includeInstance) {
    return proto.OneStyleValue.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.OneStyleValue} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.OneStyleValue.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        key: (f = msg.getKey()) && proto.StyleKey.toObject(includeInstance, f),
        int32vl: jspb.Message.getFieldWithDefault(msg, 2, 0),
        boolvl: jspb.Message.getFieldWithDefault(msg, 3, false),
        funcid: jspb.Message.getFieldWithDefault(msg, 4, 0),
        doublevl: +jspb.Message.getFieldWithDefault(msg, 5, 0.0),
        stringvl: jspb.Message.getFieldWithDefault(msg, 6, ''),
        stringarr: (f = msg.getStringarr()) && proto.StringArr.toObject(includeInstance, f),
        stylevlarr: (f = msg.getStylevlarr()) && proto.StyleValueArr.toObject(includeInstance, f)
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.OneStyleValue}
 */
proto.OneStyleValue.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.OneStyleValue()
  return proto.OneStyleValue.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.OneStyleValue} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.OneStyleValue}
 */
proto.OneStyleValue.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = new proto.StyleKey()
        reader.readMessage(value, proto.StyleKey.deserializeBinaryFromReader)
        msg.setKey(value)
        break
      case 2:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setInt32vl(value)
        break
      case 3:
        var value = /** @type {boolean} */ (reader.readBool())
        msg.setBoolvl(value)
        break
      case 4:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setFuncid(value)
        break
      case 5:
        var value = /** @type {number} */ (reader.readDouble())
        msg.setDoublevl(value)
        break
      case 6:
        var value = /** @type {string} */ (reader.readString())
        msg.setStringvl(value)
        break
      case 7:
        var value = new proto.StringArr()
        reader.readMessage(value, proto.StringArr.deserializeBinaryFromReader)
        msg.setStringarr(value)
        break
      case 8:
        var value = new proto.StyleValueArr()
        reader.readMessage(value, proto.StyleValueArr.deserializeBinaryFromReader)
        msg.setStylevlarr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.OneStyleValue.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.OneStyleValue.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.OneStyleValue} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.OneStyleValue.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getKey()
  if (f != null) {
    writer.writeMessage(1, f, proto.StyleKey.serializeBinaryToWriter)
  }
  f = /** @type {number} */ (jspb.Message.getField(message, 2))
  if (f != null) {
    writer.writeInt32(2, f)
  }
  f = /** @type {boolean} */ (jspb.Message.getField(message, 3))
  if (f != null) {
    writer.writeBool(3, f)
  }
  f = /** @type {number} */ (jspb.Message.getField(message, 4))
  if (f != null) {
    writer.writeInt32(4, f)
  }
  f = /** @type {number} */ (jspb.Message.getField(message, 5))
  if (f != null) {
    writer.writeDouble(5, f)
  }
  f = /** @type {string} */ (jspb.Message.getField(message, 6))
  if (f != null) {
    writer.writeString(6, f)
  }
  f = message.getStringarr()
  if (f != null) {
    writer.writeMessage(7, f, proto.StringArr.serializeBinaryToWriter)
  }
  f = message.getStylevlarr()
  if (f != null) {
    writer.writeMessage(8, f, proto.StyleValueArr.serializeBinaryToWriter)
  }
}

/**
 * optional StyleKey key = 1;
 * @return {?proto.StyleKey}
 */
proto.OneStyleValue.prototype.getKey = function() {
  return /** @type{?proto.StyleKey} */ (jspb.Message.getWrapperField(this, proto.StyleKey, 1))
}

/** @param {?proto.StyleKey|undefined} value */
proto.OneStyleValue.prototype.setKey = function(value) {
  jspb.Message.setWrapperField(this, 1, value)
}

proto.OneStyleValue.prototype.clearKey = function() {
  this.setKey(undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasKey = function() {
  return jspb.Message.getField(this, 1) != null
}

/**
 * optional int32 int32Vl = 2;
 * @return {number}
 */
proto.OneStyleValue.prototype.getInt32vl = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0))
}

/** @param {number} value */
proto.OneStyleValue.prototype.setInt32vl = function(value) {
  jspb.Message.setOneofField(this, 2, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearInt32vl = function() {
  jspb.Message.setOneofField(this, 2, proto.OneStyleValue.oneofGroups_[0], undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasInt32vl = function() {
  return jspb.Message.getField(this, 2) != null
}

/**
 * optional bool boolVl = 3;
 * Note that Boolean fields may be set to 0/1 when serialized from a Java server.
 * You should avoid comparisons like {@code val === true/false} in those cases.
 * @return {boolean}
 */
proto.OneStyleValue.prototype.getBoolvl = function() {
  return /** @type {boolean} */ (jspb.Message.getFieldWithDefault(this, 3, false))
}

/** @param {boolean} value */
proto.OneStyleValue.prototype.setBoolvl = function(value) {
  jspb.Message.setOneofField(this, 3, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearBoolvl = function() {
  jspb.Message.setOneofField(this, 3, proto.OneStyleValue.oneofGroups_[0], undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasBoolvl = function() {
  return jspb.Message.getField(this, 3) != null
}

/**
 * optional int32 funcId = 4;
 * @return {number}
 */
proto.OneStyleValue.prototype.getFuncid = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 4, 0))
}

/** @param {number} value */
proto.OneStyleValue.prototype.setFuncid = function(value) {
  jspb.Message.setOneofField(this, 4, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearFuncid = function() {
  jspb.Message.setOneofField(this, 4, proto.OneStyleValue.oneofGroups_[0], undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasFuncid = function() {
  return jspb.Message.getField(this, 4) != null
}

/**
 * optional double doubleVl = 5;
 * @return {number}
 */
proto.OneStyleValue.prototype.getDoublevl = function() {
  return /** @type {number} */ (+jspb.Message.getFieldWithDefault(this, 5, 0.0))
}

/** @param {number} value */
proto.OneStyleValue.prototype.setDoublevl = function(value) {
  jspb.Message.setOneofField(this, 5, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearDoublevl = function() {
  jspb.Message.setOneofField(this, 5, proto.OneStyleValue.oneofGroups_[0], undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasDoublevl = function() {
  return jspb.Message.getField(this, 5) != null
}

/**
 * optional string stringVl = 6;
 * @return {string}
 */
proto.OneStyleValue.prototype.getStringvl = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 6, ''))
}

/** @param {string} value */
proto.OneStyleValue.prototype.setStringvl = function(value) {
  jspb.Message.setOneofField(this, 6, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearStringvl = function() {
  jspb.Message.setOneofField(this, 6, proto.OneStyleValue.oneofGroups_[0], undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasStringvl = function() {
  return jspb.Message.getField(this, 6) != null
}

/**
 * optional StringArr stringArr = 7;
 * @return {?proto.StringArr}
 */
proto.OneStyleValue.prototype.getStringarr = function() {
  return /** @type{?proto.StringArr} */ (jspb.Message.getWrapperField(this, proto.StringArr, 7))
}

/** @param {?proto.StringArr|undefined} value */
proto.OneStyleValue.prototype.setStringarr = function(value) {
  jspb.Message.setOneofWrapperField(this, 7, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearStringarr = function() {
  this.setStringarr(undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasStringarr = function() {
  return jspb.Message.getField(this, 7) != null
}

/**
 * optional StyleValueArr styleVlArr = 8;
 * @return {?proto.StyleValueArr}
 */
proto.OneStyleValue.prototype.getStylevlarr = function() {
  return /** @type{?proto.StyleValueArr} */ (jspb.Message.getWrapperField(
    this,
    proto.StyleValueArr,
    8
  ))
}

/** @param {?proto.StyleValueArr|undefined} value */
proto.OneStyleValue.prototype.setStylevlarr = function(value) {
  jspb.Message.setOneofWrapperField(this, 8, proto.OneStyleValue.oneofGroups_[0], value)
}

proto.OneStyleValue.prototype.clearStylevlarr = function() {
  this.setStylevlarr(undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.OneStyleValue.prototype.hasStylevlarr = function() {
  return jspb.Message.getField(this, 8) != null
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.StyleValueArr = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.StyleValueArr.repeatedFields_, null)
}
goog.inherits(proto.StyleValueArr, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.StyleValueArr.displayName = 'proto.StyleValueArr'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.StyleValueArr.repeatedFields_ = [1]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.StyleValueArr.prototype.toObject = function(opt_includeInstance) {
    return proto.StyleValueArr.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.StyleValueArr} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.StyleValueArr.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        vlarrList: jspb.Message.toObjectList(
          msg.getVlarrList(),
          proto.OneStyleValue.toObject,
          includeInstance
        )
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.StyleValueArr}
 */
proto.StyleValueArr.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.StyleValueArr()
  return proto.StyleValueArr.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.StyleValueArr} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.StyleValueArr}
 */
proto.StyleValueArr.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = new proto.OneStyleValue()
        reader.readMessage(value, proto.OneStyleValue.deserializeBinaryFromReader)
        msg.addVlarr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.StyleValueArr.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.StyleValueArr.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.StyleValueArr} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.StyleValueArr.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getVlarrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(1, f, proto.OneStyleValue.serializeBinaryToWriter)
  }
}

/**
 * repeated OneStyleValue vlArr = 1;
 * @return {!Array<!proto.OneStyleValue>}
 */
proto.StyleValueArr.prototype.getVlarrList = function() {
  return /** @type{!Array<!proto.OneStyleValue>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.OneStyleValue,
    1
  ))
}

/** @param {!Array<!proto.OneStyleValue>} value */
proto.StyleValueArr.prototype.setVlarrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 1, value)
}

/**
 * @param {!proto.OneStyleValue=} opt_value
 * @param {number=} opt_index
 * @return {!proto.OneStyleValue}
 */
proto.StyleValueArr.prototype.addVlarr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 1, opt_value, proto.OneStyleValue, opt_index)
}

proto.StyleValueArr.prototype.clearVlarrList = function() {
  this.setVlarrList([])
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.SelectorKey = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null)
}
goog.inherits(proto.SelectorKey, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.SelectorKey.displayName = 'proto.SelectorKey'
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.SelectorKey.prototype.toObject = function(opt_includeInstance) {
    return proto.SelectorKey.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.SelectorKey} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.SelectorKey.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        id: jspb.Message.getFieldWithDefault(msg, 1, 0),
        name: jspb.Message.getFieldWithDefault(msg, 2, '')
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.SelectorKey}
 */
proto.SelectorKey.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.SelectorKey()
  return proto.SelectorKey.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.SelectorKey} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.SelectorKey}
 */
proto.SelectorKey.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setId(value)
        break
      case 2:
        var value = /** @type {string} */ (reader.readString())
        msg.setName(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.SelectorKey.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.SelectorKey.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.SelectorKey} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.SelectorKey.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getId()
  if (f !== 0) {
    writer.writeInt32(1, f)
  }
  f = message.getName()
  if (f.length > 0) {
    writer.writeString(2, f)
  }
}

/**
 * optional int32 id = 1;
 * @return {number}
 */
proto.SelectorKey.prototype.getId = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0))
}

/** @param {number} value */
proto.SelectorKey.prototype.setId = function(value) {
  jspb.Message.setProto3IntField(this, 1, value)
}

/**
 * optional string name = 2;
 * @return {string}
 */
proto.SelectorKey.prototype.getName = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ''))
}

/** @param {string} value */
proto.SelectorKey.prototype.setName = function(value) {
  jspb.Message.setProto3StringField(this, 2, value)
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.CssCell = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.CssCell.repeatedFields_, null)
}
goog.inherits(proto.CssCell, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.CssCell.displayName = 'proto.CssCell'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.CssCell.repeatedFields_ = [1, 2, 3]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.CssCell.prototype.toObject = function(opt_includeInstance) {
    return proto.CssCell.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.CssCell} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.CssCell.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        selectorarrList: jspb.Message.toObjectList(
          msg.getSelectorarrList(),
          proto.SelectorKey.toObject,
          includeInstance
        ),
        stylearrList: jspb.Message.toObjectList(
          msg.getStylearrList(),
          proto.OneStyleValue.toObject,
          includeInstance
        ),
        propertyarrList: jspb.Message.toObjectList(
          msg.getPropertyarrList(),
          proto.OneStyleValue.toObject,
          includeInstance
        )
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.CssCell}
 */
proto.CssCell.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.CssCell()
  return proto.CssCell.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.CssCell} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.CssCell}
 */
proto.CssCell.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = new proto.SelectorKey()
        reader.readMessage(value, proto.SelectorKey.deserializeBinaryFromReader)
        msg.addSelectorarr(value)
        break
      case 2:
        var value = new proto.OneStyleValue()
        reader.readMessage(value, proto.OneStyleValue.deserializeBinaryFromReader)
        msg.addStylearr(value)
        break
      case 3:
        var value = new proto.OneStyleValue()
        reader.readMessage(value, proto.OneStyleValue.deserializeBinaryFromReader)
        msg.addPropertyarr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.CssCell.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.CssCell.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.CssCell} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.CssCell.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getSelectorarrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(1, f, proto.SelectorKey.serializeBinaryToWriter)
  }
  f = message.getStylearrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(2, f, proto.OneStyleValue.serializeBinaryToWriter)
  }
  f = message.getPropertyarrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(3, f, proto.OneStyleValue.serializeBinaryToWriter)
  }
}

/**
 * repeated SelectorKey selectorArr = 1;
 * @return {!Array<!proto.SelectorKey>}
 */
proto.CssCell.prototype.getSelectorarrList = function() {
  return /** @type{!Array<!proto.SelectorKey>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.SelectorKey,
    1
  ))
}

/** @param {!Array<!proto.SelectorKey>} value */
proto.CssCell.prototype.setSelectorarrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 1, value)
}

/**
 * @param {!proto.SelectorKey=} opt_value
 * @param {number=} opt_index
 * @return {!proto.SelectorKey}
 */
proto.CssCell.prototype.addSelectorarr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 1, opt_value, proto.SelectorKey, opt_index)
}

proto.CssCell.prototype.clearSelectorarrList = function() {
  this.setSelectorarrList([])
}

/**
 * repeated OneStyleValue styleArr = 2;
 * @return {!Array<!proto.OneStyleValue>}
 */
proto.CssCell.prototype.getStylearrList = function() {
  return /** @type{!Array<!proto.OneStyleValue>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.OneStyleValue,
    2
  ))
}

/** @param {!Array<!proto.OneStyleValue>} value */
proto.CssCell.prototype.setStylearrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 2, value)
}

/**
 * @param {!proto.OneStyleValue=} opt_value
 * @param {number=} opt_index
 * @return {!proto.OneStyleValue}
 */
proto.CssCell.prototype.addStylearr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 2, opt_value, proto.OneStyleValue, opt_index)
}

proto.CssCell.prototype.clearStylearrList = function() {
  this.setStylearrList([])
}

/**
 * repeated OneStyleValue propertyArr = 3;
 * @return {!Array<!proto.OneStyleValue>}
 */
proto.CssCell.prototype.getPropertyarrList = function() {
  return /** @type{!Array<!proto.OneStyleValue>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.OneStyleValue,
    3
  ))
}

/** @param {!Array<!proto.OneStyleValue>} value */
proto.CssCell.prototype.setPropertyarrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 3, value)
}

/**
 * @param {!proto.OneStyleValue=} opt_value
 * @param {number=} opt_index
 * @return {!proto.OneStyleValue}
 */
proto.CssCell.prototype.addPropertyarr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 3, opt_value, proto.OneStyleValue, opt_index)
}

proto.CssCell.prototype.clearPropertyarrList = function() {
  this.setPropertyarrList([])
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.TemplateUnit = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.TemplateUnit.repeatedFields_, null)
}
goog.inherits(proto.TemplateUnit, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.TemplateUnit.displayName = 'proto.TemplateUnit'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.TemplateUnit.repeatedFields_ = [5, 6]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.TemplateUnit.prototype.toObject = function(opt_includeInstance) {
    return proto.TemplateUnit.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.TemplateUnit} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.TemplateUnit.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        type: jspb.Message.getFieldWithDefault(msg, 1, ''),
        tag: jspb.Message.getFieldWithDefault(msg, 2, ''),
        optsid: jspb.Message.getFieldWithDefault(msg, 3, 0),
        funcrender: jspb.Message.getFieldWithDefault(msg, 4, 0),
        optsarrList: jspb.Message.toObjectList(
          msg.getOptsarrList(),
          proto.OneStyleValue.toObject,
          includeInstance
        ),
        templaterenderList: jspb.Message.toObjectList(
          msg.getTemplaterenderList(),
          proto.TemplateOne.toObject,
          includeInstance
        )
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.TemplateUnit}
 */
proto.TemplateUnit.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.TemplateUnit()
  return proto.TemplateUnit.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.TemplateUnit} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.TemplateUnit}
 */
proto.TemplateUnit.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {string} */ (reader.readString())
        msg.setType(value)
        break
      case 2:
        var value = /** @type {string} */ (reader.readString())
        msg.setTag(value)
        break
      case 3:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setOptsid(value)
        break
      case 4:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setFuncrender(value)
        break
      case 5:
        var value = new proto.OneStyleValue()
        reader.readMessage(value, proto.OneStyleValue.deserializeBinaryFromReader)
        msg.addOptsarr(value)
        break
      case 6:
        var value = new proto.TemplateOne()
        reader.readMessage(value, proto.TemplateOne.deserializeBinaryFromReader)
        msg.addTemplaterender(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.TemplateUnit.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.TemplateUnit.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.TemplateUnit} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.TemplateUnit.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getType()
  if (f.length > 0) {
    writer.writeString(1, f)
  }
  f = message.getTag()
  if (f.length > 0) {
    writer.writeString(2, f)
  }
  f = message.getOptsid()
  if (f !== 0) {
    writer.writeInt32(3, f)
  }
  f = message.getFuncrender()
  if (f !== 0) {
    writer.writeInt32(4, f)
  }
  f = message.getOptsarrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(5, f, proto.OneStyleValue.serializeBinaryToWriter)
  }
  f = message.getTemplaterenderList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(6, f, proto.TemplateOne.serializeBinaryToWriter)
  }
}

/**
 * optional string type = 1;
 * @return {string}
 */
proto.TemplateUnit.prototype.getType = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ''))
}

/** @param {string} value */
proto.TemplateUnit.prototype.setType = function(value) {
  jspb.Message.setProto3StringField(this, 1, value)
}

/**
 * optional string tag = 2;
 * @return {string}
 */
proto.TemplateUnit.prototype.getTag = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 2, ''))
}

/** @param {string} value */
proto.TemplateUnit.prototype.setTag = function(value) {
  jspb.Message.setProto3StringField(this, 2, value)
}

/**
 * optional int32 optsId = 3;
 * @return {number}
 */
proto.TemplateUnit.prototype.getOptsid = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 3, 0))
}

/** @param {number} value */
proto.TemplateUnit.prototype.setOptsid = function(value) {
  jspb.Message.setProto3IntField(this, 3, value)
}

/**
 * optional int32 funcRender = 4;
 * @return {number}
 */
proto.TemplateUnit.prototype.getFuncrender = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 4, 0))
}

/** @param {number} value */
proto.TemplateUnit.prototype.setFuncrender = function(value) {
  jspb.Message.setProto3IntField(this, 4, value)
}

/**
 * repeated OneStyleValue optsArr = 5;
 * @return {!Array<!proto.OneStyleValue>}
 */
proto.TemplateUnit.prototype.getOptsarrList = function() {
  return /** @type{!Array<!proto.OneStyleValue>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.OneStyleValue,
    5
  ))
}

/** @param {!Array<!proto.OneStyleValue>} value */
proto.TemplateUnit.prototype.setOptsarrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 5, value)
}

/**
 * @param {!proto.OneStyleValue=} opt_value
 * @param {number=} opt_index
 * @return {!proto.OneStyleValue}
 */
proto.TemplateUnit.prototype.addOptsarr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 5, opt_value, proto.OneStyleValue, opt_index)
}

proto.TemplateUnit.prototype.clearOptsarrList = function() {
  this.setOptsarrList([])
}

/**
 * repeated TemplateOne templateRender = 6;
 * @return {!Array<!proto.TemplateOne>}
 */
proto.TemplateUnit.prototype.getTemplaterenderList = function() {
  return /** @type{!Array<!proto.TemplateOne>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.TemplateOne,
    6
  ))
}

/** @param {!Array<!proto.TemplateOne>} value */
proto.TemplateUnit.prototype.setTemplaterenderList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 6, value)
}

/**
 * @param {!proto.TemplateOne=} opt_value
 * @param {number=} opt_index
 * @return {!proto.TemplateOne}
 */
proto.TemplateUnit.prototype.addTemplaterender = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 6, opt_value, proto.TemplateOne, opt_index)
}

proto.TemplateUnit.prototype.clearTemplaterenderList = function() {
  this.setTemplaterenderList([])
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.TemplateOne = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.TemplateOne.repeatedFields_, null)
}
goog.inherits(proto.TemplateOne, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.TemplateOne.displayName = 'proto.TemplateOne'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.TemplateOne.repeatedFields_ = [2]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.TemplateOne.prototype.toObject = function(opt_includeInstance) {
    return proto.TemplateOne.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.TemplateOne} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.TemplateOne.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        id: jspb.Message.getFieldWithDefault(msg, 1, 0),
        chldsidList: jspb.Message.getRepeatedField(msg, 2),
        cv: (f = msg.getCv()) && proto.TemplateUnit.toObject(includeInstance, f)
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.TemplateOne}
 */
proto.TemplateOne.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.TemplateOne()
  return proto.TemplateOne.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.TemplateOne} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.TemplateOne}
 */
proto.TemplateOne.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setId(value)
        break
      case 2:
        var value = /** @type {!Array<number>} */ (reader.readPackedInt32())
        msg.setChldsidList(value)
        break
      case 3:
        var value = new proto.TemplateUnit()
        reader.readMessage(value, proto.TemplateUnit.deserializeBinaryFromReader)
        msg.setCv(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.TemplateOne.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.TemplateOne.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.TemplateOne} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.TemplateOne.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getId()
  if (f !== 0) {
    writer.writeInt32(1, f)
  }
  f = message.getChldsidList()
  if (f.length > 0) {
    writer.writePackedInt32(2, f)
  }
  f = message.getCv()
  if (f != null) {
    writer.writeMessage(3, f, proto.TemplateUnit.serializeBinaryToWriter)
  }
}

/**
 * optional int32 id = 1;
 * @return {number}
 */
proto.TemplateOne.prototype.getId = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 1, 0))
}

/** @param {number} value */
proto.TemplateOne.prototype.setId = function(value) {
  jspb.Message.setProto3IntField(this, 1, value)
}

/**
 * repeated int32 chldsId = 2;
 * @return {!Array<number>}
 */
proto.TemplateOne.prototype.getChldsidList = function() {
  return /** @type {!Array<number>} */ (jspb.Message.getRepeatedField(this, 2))
}

/** @param {!Array<number>} value */
proto.TemplateOne.prototype.setChldsidList = function(value) {
  jspb.Message.setField(this, 2, value || [])
}

/**
 * @param {!number} value
 * @param {number=} opt_index
 */
proto.TemplateOne.prototype.addChldsid = function(value, opt_index) {
  jspb.Message.addToRepeatedField(this, 2, value, opt_index)
}

proto.TemplateOne.prototype.clearChldsidList = function() {
  this.setChldsidList([])
}

/**
 * optional TemplateUnit __cv__ = 3;
 * @return {?proto.TemplateUnit}
 */
proto.TemplateOne.prototype.getCv = function() {
  return /** @type{?proto.TemplateUnit} */ (jspb.Message.getWrapperField(
    this,
    proto.TemplateUnit,
    3
  ))
}

/** @param {?proto.TemplateUnit|undefined} value */
proto.TemplateOne.prototype.setCv = function(value) {
  jspb.Message.setWrapperField(this, 3, value)
}

proto.TemplateOne.prototype.clearCv = function() {
  this.setCv(undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.TemplateOne.prototype.hasCv = function() {
  return jspb.Message.getField(this, 3) != null
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ExportCell = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.ExportCell.repeatedFields_, null)
}
goog.inherits(proto.ExportCell, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.ExportCell.displayName = 'proto.ExportCell'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.ExportCell.repeatedFields_ = [2, 3]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.ExportCell.prototype.toObject = function(opt_includeInstance) {
    return proto.ExportCell.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.ExportCell} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.ExportCell.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        tagname: jspb.Message.getFieldWithDefault(msg, 1, ''),
        cssarrList: jspb.Message.toObjectList(
          msg.getCssarrList(),
          proto.CssCell.toObject,
          includeInstance
        ),
        templatearrList: jspb.Message.toObjectList(
          msg.getTemplatearrList(),
          proto.TemplateOne.toObject,
          includeInstance
        )
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ExportCell}
 */
proto.ExportCell.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.ExportCell()
  return proto.ExportCell.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ExportCell} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ExportCell}
 */
proto.ExportCell.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {string} */ (reader.readString())
        msg.setTagname(value)
        break
      case 2:
        var value = new proto.CssCell()
        reader.readMessage(value, proto.CssCell.deserializeBinaryFromReader)
        msg.addCssarr(value)
        break
      case 3:
        var value = new proto.TemplateOne()
        reader.readMessage(value, proto.TemplateOne.deserializeBinaryFromReader)
        msg.addTemplatearr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ExportCell.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.ExportCell.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ExportCell} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ExportCell.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getTagname()
  if (f.length > 0) {
    writer.writeString(1, f)
  }
  f = message.getCssarrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(2, f, proto.CssCell.serializeBinaryToWriter)
  }
  f = message.getTemplatearrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(3, f, proto.TemplateOne.serializeBinaryToWriter)
  }
}

/**
 * optional string tagName = 1;
 * @return {string}
 */
proto.ExportCell.prototype.getTagname = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ''))
}

/** @param {string} value */
proto.ExportCell.prototype.setTagname = function(value) {
  jspb.Message.setProto3StringField(this, 1, value)
}

/**
 * repeated CssCell cssArr = 2;
 * @return {!Array<!proto.CssCell>}
 */
proto.ExportCell.prototype.getCssarrList = function() {
  return /** @type{!Array<!proto.CssCell>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.CssCell,
    2
  ))
}

/** @param {!Array<!proto.CssCell>} value */
proto.ExportCell.prototype.setCssarrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 2, value)
}

/**
 * @param {!proto.CssCell=} opt_value
 * @param {number=} opt_index
 * @return {!proto.CssCell}
 */
proto.ExportCell.prototype.addCssarr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 2, opt_value, proto.CssCell, opt_index)
}

proto.ExportCell.prototype.clearCssarrList = function() {
  this.setCssarrList([])
}

/**
 * repeated TemplateOne templateArr = 3;
 * @return {!Array<!proto.TemplateOne>}
 */
proto.ExportCell.prototype.getTemplatearrList = function() {
  return /** @type{!Array<!proto.TemplateOne>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.TemplateOne,
    3
  ))
}

/** @param {!Array<!proto.TemplateOne>} value */
proto.ExportCell.prototype.setTemplatearrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 3, value)
}

/**
 * @param {!proto.TemplateOne=} opt_value
 * @param {number=} opt_index
 * @return {!proto.TemplateOne}
 */
proto.ExportCell.prototype.addTemplatearr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 3, opt_value, proto.TemplateOne, opt_index)
}

proto.ExportCell.prototype.clearTemplatearrList = function() {
  this.setTemplatearrList([])
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.Version = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, null, null)
}
goog.inherits(proto.Version, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.Version.displayName = 'proto.Version'
}

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.Version.prototype.toObject = function(opt_includeInstance) {
    return proto.Version.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.Version} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.Version.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        versionname: jspb.Message.getFieldWithDefault(msg, 1, ''),
        versioncode: jspb.Message.getFieldWithDefault(msg, 2, 0)
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.Version}
 */
proto.Version.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.Version()
  return proto.Version.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.Version} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.Version}
 */
proto.Version.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = /** @type {string} */ (reader.readString())
        msg.setVersionname(value)
        break
      case 2:
        var value = /** @type {number} */ (reader.readInt32())
        msg.setVersioncode(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.Version.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.Version.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.Version} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.Version.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getVersionname()
  if (f.length > 0) {
    writer.writeString(1, f)
  }
  f = message.getVersioncode()
  if (f !== 0) {
    writer.writeInt32(2, f)
  }
}

/**
 * optional string versionName = 1;
 * @return {string}
 */
proto.Version.prototype.getVersionname = function() {
  return /** @type {string} */ (jspb.Message.getFieldWithDefault(this, 1, ''))
}

/** @param {string} value */
proto.Version.prototype.setVersionname = function(value) {
  jspb.Message.setProto3StringField(this, 1, value)
}

/**
 * optional int32 versionCode = 2;
 * @return {number}
 */
proto.Version.prototype.getVersioncode = function() {
  return /** @type {number} */ (jspb.Message.getFieldWithDefault(this, 2, 0))
}

/** @param {number} value */
proto.Version.prototype.setVersioncode = function(value) {
  jspb.Message.setProto3IntField(this, 2, value)
}

/**
 * Generated by JsPbCodeGenerator.
 * @param {Array=} opt_data Optional initial data array, typically from a
 * server response, or constructed directly in Javascript. The array is used
 * in place and becomes part of the constructed object. It is not cloned.
 * If no data is provided, the constructed object will be empty, but still
 * valid.
 * @extends {jspb.Message}
 * @constructor
 */
proto.ExportArr = function(opt_data) {
  jspb.Message.initialize(this, opt_data, 0, -1, proto.ExportArr.repeatedFields_, null)
}
goog.inherits(proto.ExportArr, jspb.Message)
if (goog.DEBUG && !COMPILED) {
  proto.ExportArr.displayName = 'proto.ExportArr'
}
/**
 * List of repeated fields within this message type.
 * @private {!Array<number>}
 * @const
 */
proto.ExportArr.repeatedFields_ = [2]

if (jspb.Message.GENERATE_TO_OBJECT) {
  /**
   * Creates an object representation of this proto suitable for use in Soy templates.
   * Field names that are reserved in JavaScript and will be renamed to pb_name.
   * To access a reserved field use, foo.pb_<name>, eg, foo.pb_default.
   * For the list of reserved names please see:
   *     com.google.apps.jspb.JsClassTemplate.JS_RESERVED_WORDS.
   * @param {boolean=} opt_includeInstance Whether to include the JSPB instance
   *     for transitional soy proto support: http://goto/soy-param-migration
   * @return {!Object}
   */
  proto.ExportArr.prototype.toObject = function(opt_includeInstance) {
    return proto.ExportArr.toObject(opt_includeInstance, this)
  }

  /**
   * Static version of the {@see toObject} method.
   * @param {boolean|undefined} includeInstance Whether to include the JSPB
   *     instance for transitional soy proto support:
   *     http://goto/soy-param-migration
   * @param {!proto.ExportArr} msg The msg instance to transform.
   * @return {!Object}
   * @suppress {unusedLocalVariables} f is only used for nested messages
   */
  proto.ExportArr.toObject = function(includeInstance, msg) {
    var f,
      obj = {
        ver: (f = msg.getVer()) && proto.Version.toObject(includeInstance, f),
        valuearrList: jspb.Message.toObjectList(
          msg.getValuearrList(),
          proto.ExportCell.toObject,
          includeInstance
        )
      }

    if (includeInstance) {
      obj.$jspbMessageInstance = msg
    }
    return obj
  }
}

/**
 * Deserializes binary data (in protobuf wire format).
 * @param {jspb.ByteSource} bytes The bytes to deserialize.
 * @return {!proto.ExportArr}
 */
proto.ExportArr.deserializeBinary = function(bytes) {
  var reader = new jspb.BinaryReader(bytes)
  var msg = new proto.ExportArr()
  return proto.ExportArr.deserializeBinaryFromReader(msg, reader)
}

/**
 * Deserializes binary data (in protobuf wire format) from the
 * given reader into the given message object.
 * @param {!proto.ExportArr} msg The message object to deserialize into.
 * @param {!jspb.BinaryReader} reader The BinaryReader to use.
 * @return {!proto.ExportArr}
 */
proto.ExportArr.deserializeBinaryFromReader = function(msg, reader) {
  while (reader.nextField()) {
    if (reader.isEndGroup()) {
      break
    }
    var field = reader.getFieldNumber()
    switch (field) {
      case 1:
        var value = new proto.Version()
        reader.readMessage(value, proto.Version.deserializeBinaryFromReader)
        msg.setVer(value)
        break
      case 2:
        var value = new proto.ExportCell()
        reader.readMessage(value, proto.ExportCell.deserializeBinaryFromReader)
        msg.addValuearr(value)
        break
      default:
        reader.skipField()
        break
    }
  }
  return msg
}

/**
 * Serializes the message to binary data (in protobuf wire format).
 * @return {!Uint8Array}
 */
proto.ExportArr.prototype.serializeBinary = function() {
  var writer = new jspb.BinaryWriter()
  proto.ExportArr.serializeBinaryToWriter(this, writer)
  return writer.getResultBuffer()
}

/**
 * Serializes the given message to binary data (in protobuf wire
 * format), writing to the given BinaryWriter.
 * @param {!proto.ExportArr} message
 * @param {!jspb.BinaryWriter} writer
 * @suppress {unusedLocalVariables} f is only used for nested messages
 */
proto.ExportArr.serializeBinaryToWriter = function(message, writer) {
  var f = undefined
  f = message.getVer()
  if (f != null) {
    writer.writeMessage(1, f, proto.Version.serializeBinaryToWriter)
  }
  f = message.getValuearrList()
  if (f.length > 0) {
    writer.writeRepeatedMessage(2, f, proto.ExportCell.serializeBinaryToWriter)
  }
}

/**
 * optional Version ver = 1;
 * @return {?proto.Version}
 */
proto.ExportArr.prototype.getVer = function() {
  return /** @type{?proto.Version} */ (jspb.Message.getWrapperField(this, proto.Version, 1))
}

/** @param {?proto.Version|undefined} value */
proto.ExportArr.prototype.setVer = function(value) {
  jspb.Message.setWrapperField(this, 1, value)
}

proto.ExportArr.prototype.clearVer = function() {
  this.setVer(undefined)
}

/**
 * Returns whether this field is set.
 * @return {!boolean}
 */
proto.ExportArr.prototype.hasVer = function() {
  return jspb.Message.getField(this, 1) != null
}

/**
 * repeated ExportCell valueArr = 2;
 * @return {!Array<!proto.ExportCell>}
 */
proto.ExportArr.prototype.getValuearrList = function() {
  return /** @type{!Array<!proto.ExportCell>} */ (jspb.Message.getRepeatedWrapperField(
    this,
    proto.ExportCell,
    2
  ))
}

/** @param {!Array<!proto.ExportCell>} value */
proto.ExportArr.prototype.setValuearrList = function(value) {
  jspb.Message.setRepeatedWrapperField(this, 2, value)
}

/**
 * @param {!proto.ExportCell=} opt_value
 * @param {number=} opt_index
 * @return {!proto.ExportCell}
 */
proto.ExportArr.prototype.addValuearr = function(opt_value, opt_index) {
  return jspb.Message.addToRepeatedWrapperField(this, 2, opt_value, proto.ExportCell, opt_index)
}

proto.ExportArr.prototype.clearValuearrList = function() {
  this.setValuearrList([])
}

goog.object.extend(exports, proto)
