// @aiot-toolkit/parser/lib/ux/enum/StyleSelectorType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Selector = exports.MediaType = exports.IRuleTypes = exports.AtruleSelector = void 0;
exports.findSelectorIndex = findSelectorIndex;
exports.isMeidaType = isMeidaType;
exports.isSupportMediaType = isSupportMediaType;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var IRuleTypes = exports.IRuleTypes = /*#__PURE__*/function (IRuleTypes) {
  IRuleTypes["RULE"] = "Rule";
  IRuleTypes["AT_RULE"] = "Atrule";
  return IRuleTypes;
}(IRuleTypes || {});
var Selector = exports.Selector = /*#__PURE__*/function (Selector) {
  Selector["CLASS_SELECTOR"] = "ClassSelector";
  Selector["ID_SELECTOR"] = "IdSelector";
  Selector["TYPE_SELECTOR"] = "TypeSelector";
  Selector["COMBINATOR"] = "Combinator";
  Selector["PSEUDO_CLASS_SELECTOR"] = "PseudoClassSelector";
  return Selector;
}(Selector || {});
var AtruleSelector = exports.AtruleSelector = /*#__PURE__*/function (AtruleSelector) {
  AtruleSelector["KEYFRAMES"] = "keyframes";
  AtruleSelector["MEDIA"] = "media";
  AtruleSelector["FONTFACE"] = "font-face";
  return AtruleSelector;
}(AtruleSelector || {});
var MediaType = exports.MediaType = /*#__PURE__*/function (MediaType) {
  MediaType["ALL"] = "all";
  MediaType["PRINT"] = "print";
  MediaType["SCREEN"] = "screen";
  MediaType["SPEECH"] = "speech";
  return MediaType;
}(MediaType || {});
function findSelectorIndex(type) {
  let index = -1;
  switch (type) {
    case Selector.CLASS_SELECTOR:
      index = 0;
      break;
    case Selector.ID_SELECTOR:
      index = 1;
      break;
    case Selector.TYPE_SELECTOR:
      index = 2;
      break;
    case AtruleSelector.KEYFRAMES:
      index = 3;
      break;
    case AtruleSelector.FONTFACE:
      index = 4;
      break;
  }
  return index;
}
function isMeidaType(type) {
  return Object.values(MediaType).includes(type);
}
/**
 * 安卓快应用支持所有的媒体类型，只需要判断媒体类型是否合法
 * vela快应用仅支持 MediaType.SCREEN
 * @param type
 * @param projectType
 * @returns
 */
function isSupportMediaType(type, projectType) {
  let supportedMediaList = [];
  switch (projectType) {
    case _sharedUtils.ProjectType.VELA_UX:
      supportedMediaList = [MediaType.SCREEN];
      break;
    case _sharedUtils.ProjectType.ANDDROID_UX:
      supportedMediaList = [MediaType.ALL, MediaType.PRINT, MediaType.SCREEN, MediaType.SPEECH];
      break;
    default:
      break;
  }
  return supportedMediaList.includes(type);
}
