// @aiot-toolkit/parser/lib/ux/enum/TemplateNodeType.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
/**
 * TemplateNode 类型枚举
 */
var TemplateNodeType = /*#__PURE__*/function (TemplateNodeType) {
  TemplateNodeType["TEMPLATE"] = "template";
  TemplateNodeType["COMMENT"] = "#comment";
  TemplateNodeType["TEXT"] = "#text";
  TemplateNodeType["DOCUMENT_TYPE"] = "#documentType";
  TemplateNodeType["DOCUMENT_FRAGMENT"] = "#document-fragment";
  return TemplateNodeType;
}(TemplateNodeType || {});
var _default = exports.default = TemplateNodeType;
(function (_TemplateNodeType) {
  const ALL_LIST = _TemplateNodeType.ALL_LIST = [TemplateNodeType.TEMPLATE, TemplateNodeType.COMMENT, TemplateNodeType.TEXT, TemplateNodeType.DOCUMENT_TYPE];
  function isElement(node) {
    return ![TemplateNodeType.COMMENT, TemplateNodeType.TEXT, TemplateNodeType.DOCUMENT_TYPE].includes(node.nodeName);
  }
  _TemplateNodeType.isElement = isElement;
  function isUnEmptyText(node) {
    return node.nodeName === '#text' && Boolean(node.value.trim());
  }
  _TemplateNodeType.isUnEmptyText = isUnEmptyText;
  function isEffectElement(node) {
    return isElement(node) || isUnEmptyText(node);
  }
  _TemplateNodeType.isEffectElement = isEffectElement;
})(TemplateNodeType || (TemplateNodeType = {}));
