// @aiot-toolkit/parser/lib/ux/translate/android/attributeTranslate/ForTranslate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _sharedUtils = require("@aiot-toolkit/shared-utils");
var _ElementConfig = _interopRequireDefault(require("../../../config/android/ElementConfig"));
var _ElementConfigUtil = _interopRequireDefault(require("../../../utils/ElementConfigUtil"));
var _TemplateUtil = _interopRequireDefault(require("../../vela/utils/TemplateUtil"));
var _CfTranslate = _interopRequireDefault(require("../../vela/wrap/CfTranslate"));
var _AndroidContext = _interopRequireDefault(require("../AndroidContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * ForTranslate
 */
class ForTranslate {
  KEY = 'for';
  constructor(options, compilerOption) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.templateUtil = new _TemplateUtil.default(_AndroidContext.default, options, compilerOption, new _ElementConfigUtil.default(_ElementConfig.default));
  }
  match(node) {
    return Boolean(_TemplateUtil.default.getAttribute(node, this.KEY));
  }

  /**
   * 转换 for 节点
   * 1. 只有 list，生成
   * ```
   * { repeat:function }
   * ```
   * 2. 有index、 item中的任意一项，生成
   * ```json
   * repeat: {
   *    exp: function,
   *    key: index,
   *    value: item
   * }
   * ```
   * @param node
   */
  async translate(node) {
    const {
      onLog,
      filePath
    } = this.options;
    const forAttribute = _TemplateUtil.default.getAttribute(node, this.KEY);
    if (forAttribute) {
      const {
        listNode,
        itemNode,
        indexNode
      } = _CfTranslate.default.extractForNodes(forAttribute, this.options);
      if (!listNode) {
        onLog({
          filePath,
          position: _TemplateUtil.default.parse5LocationToPosition(node.sourceCodeLocation?.attrs?.[this.KEY]),
          level: _sharedUtils.Loglevel.THROW,
          message: [{
            word: 'list'
          }, `value Lost, example: for="list"`]
        });
        return null;
      }
      const result = {};
      const exp = await this.templateUtil.translateAttributeValue({
        name: forAttribute.name,
        nameLocation: forAttribute.nameLocation,
        value: _TemplateUtil.default.toDynamicValue(listNode.getText())
      }, []);
      const itemName = itemNode.escapedText === '$item' ? undefined : itemNode.escapedText;
      const indexName = indexNode.escapedText === '$idx' ? undefined : indexNode.escapedText;
      if (itemName || indexName) {
        result.repeat = {
          exp,
          key: indexName ? `'${indexName}'` : undefined,
          value: itemName ? `'${itemName}'` : undefined
        };
      } else {
        result.repeat = exp;
      }
      return result;
    }
    return null;
  }
}
var _default = exports.default = ForTranslate;
