// @aiot-toolkit/parser/lib/ux/translate/android/attributeTranslate/IfTranslate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _ElementConfig = _interopRequireDefault(require("../../../config/android/ElementConfig"));
var _ElementConfigUtil = _interopRequireDefault(require("../../../utils/ElementConfigUtil"));
var _TemplateUtil = _interopRequireDefault(require("../../vela/utils/TemplateUtil"));
var _AndroidContext = _interopRequireDefault(require("../AndroidContext"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
/**
 * IfTranslate
 */
class IfTranslate {
  KEY_LIST = ['if', 'elif', 'else'];
  constructor(options, compilerOption) {
    this.options = options;
    this.compilerOption = compilerOption;
    this.templateUtil = new _TemplateUtil.default(_AndroidContext.default, options, compilerOption, new _ElementConfigUtil.default(_ElementConfig.default));
  }
  match(node) {
    return this.KEY_LIST.some(item => _TemplateUtil.default.getAttribute(node, item));
  }

  /**
   *
   *
   * 1. 获取条件列表，节点的条件 = 前置条件取反 && 当前节点自身条件
   * 2. 返回 {show:Function}
   * @param node
   * @param identifiers
   */
  async translate(node, identifiers) {
    const conditions = this.templateUtil.getIfCoditionList(node);
    if (conditions) {
      const value = this.templateUtil.genIfConditionList(conditions, identifiers);
      return {
        shown: `function () {
          return ${value};
        }`
      };
    }
    return null;
  }
}
var _default = exports.default = IfTranslate;
