// @aiot-toolkit/commander/lib/persistentCommand/PersistentCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _chalk = _interopRequireDefault(require("chalk"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const CTRL_C = '\u0003';
const CTRL_L = '\u000C';
const BLT = '\u203A';
class PersistentCommand {
  running = false;
  actionRunning = false;
  constructor(data) {
    this.data = data;
  }
  start() {
    if (this.running) {
      return;
    }
    const {
      stdin
    } = process;
    stdin.setRawMode(true);
    stdin.resume();
    stdin.setEncoding('utf8');
    stdin.on('data', this.dataHandler);
    this.running = true;
  }
  dataHandler = async data => {
    switch (data) {
      case CTRL_C:
        process.exit();
        break;
      case CTRL_L:
        this.clearLog();
        break;
      default:
        if (this.actionRunning) {
          console.log(_chalk.default.yellow(`Wait for the task to complete`));
          return;
        }
        const item = this.data.options.find(item => item.key === data);
        if (item) {
          this.actionRunning = true;
          try {
            await item.action();
          } catch (error) {
            console.error(error);
          }
          this.actionRunning = false;
        }
        break;
    }
  };
  stop() {
    const {
      stdin
    } = process;
    stdin.removeListener('data', this.dataHandler);
    stdin.setRawMode(false);
    stdin.resume();
    this.running = false;
  }
  writeLine() {
    const str = new Array(30).fill(0).map(() => '-').join('');
    console.log(_chalk.default.dim(str));
  }
  describe() {
    const {
      description,
      options
    } = this.data;
    this.writeLine();
    // 输出描述
    console.log(_chalk.default.green(description));
    console.log('');

    // 输出命令列表
    options.forEach(item => {
      const {
        key,
        description
      } = item;
      console.log(`${BLT} ${_chalk.default.bold(key)} | ${description}`);
    });

    // 输出退出的方法
    console.log();
    console.log(_chalk.default.dim(`Press Ctrl+C to exit.`));
    this.writeLine();
  }
  clearLog() {
    process.stdout.write(process.platform === 'win32' ? '\x1B[2J\x1B[0f' : '\x1B[2J\x1B[3J\x1B[H');
  }
}
var _default = exports.default = PersistentCommand;
